<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Hugo 0.80.0" />


<title>Modern Statistics for Modern Biology-1 - A Hugo website</title>
<meta property="og:title" content="Modern Statistics for Modern Biology-1 - A Hugo website">


  <link href='../../../../favicon.ico' rel='icon' type='image/x-icon'/>



  








<link href='//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css' rel='stylesheet' type='text/css' />



<link rel="stylesheet" href="../../../../css/fonts.css" media="all">
<link rel="stylesheet" href="../../../../css/main.css" media="all">



  </head>
  <body>
    <div class="wrapper">
      <header class="header">
        <nav class="nav">
  <a href="../../../../" class="nav-logo">
    <img src="../../../../images/logo.png"
         width="50"
         height="50"
         alt="Logo">
  </a>

  <ul class="nav-links">
    
    <li><a href="../../../../about/">about</a></li>
    
    <li><a href="../../../../vitae/">CV</a></li>
    
    <li><a href="https://github.com/danli349">GitHub</a></li>
    
    <li><a href="https://scholar.google.com/citations?user=mNjLK8EAAAAJ&amp;hl=en">Googlr Scholar</a></li>
    
  </ul>
</nav>

      </header>


<main class="content" role="main">

  <article class="article">
    
    <span class="article-duration">134 min read</span>
    

    <h1 class="article-title">Modern Statistics for Modern Biology-1</h1>

    
    <span class="article-date">2023-04-28</span>
    

    <div class="article-content">
      
<script src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/htmlwidgets/htmlwidgets.js"></script>
<script src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/plotly-binding/plotly.js"></script>
<script src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/typedarray/typedarray.min.js"></script>
<script src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/jquery/jquery.min.js"></script>
<link href="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/crosstalk/css/crosstalk.min.css" rel="stylesheet" />
<script src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/crosstalk/js/crosstalk.min.js"></script>
<link href="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/plotly-htmlwidgets-css/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/plotly-main/plotly-latest.min.js"></script>

<div id="TOC">
<ul>
<li><a href="#generative-models-for-discrete-data" id="toc-generative-models-for-discrete-data">1 Generative Models for Discrete Data</a></li>
<li><a href="#statistical-modeling" id="toc-statistical-modeling">2 Statistical Modeling</a></li>
<li><a href="#high-quality-graphics-in-r" id="toc-high-quality-graphics-in-r">3 High Quality Graphics in R</a></li>
<li><a href="#mixture-models" id="toc-mixture-models">4 Mixture Models</a></li>
</ul>
</div>

<div id="generative-models-for-discrete-data" class="section level2">
<h2>1 Generative Models for Discrete Data</h2>
<pre class="r"><code>## -----------------------------------------------------------------------------
dpois(x = 3, lambda = 5)</code></pre>
<pre><code>## [1] 0.1403739</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
.oldopt = options(digits = 2)
0:12</code></pre>
<pre><code>##  [1]  0  1  2  3  4  5  6  7  8  9 10 11 12</code></pre>
<pre class="r"><code>dpois(x = 0:12, lambda = 5)</code></pre>
<pre><code>##  [1] 0.0067 0.0337 0.0842 0.1404 0.1755 0.1755 0.1462 0.1044 0.0653 0.0363
## [11] 0.0181 0.0082 0.0034</code></pre>
<pre class="r"><code>barplot(dpois(0:12, 5), names.arg = 0:12, col = &quot;red&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-2-1.svg" width="576" /></p>
<pre class="r"><code>options(.oldopt)</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
genotype = c(&quot;AA&quot;,&quot;AO&quot;,&quot;BB&quot;,&quot;AO&quot;,&quot;OO&quot;,&quot;AO&quot;,&quot;AA&quot;,&quot;BO&quot;,&quot;BO&quot;,
             &quot;AO&quot;,&quot;BB&quot;,&quot;AO&quot;,&quot;BO&quot;,&quot;AB&quot;,&quot;OO&quot;,&quot;AB&quot;,&quot;BB&quot;,&quot;AO&quot;,&quot;AO&quot;)
table(genotype)</code></pre>
<pre><code>## genotype
## AA AB AO BB BO OO 
##  2  2  7  3  3  2</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
genotypeF = factor(genotype)
levels(genotypeF)</code></pre>
<pre><code>## [1] &quot;AA&quot; &quot;AB&quot; &quot;AO&quot; &quot;BB&quot; &quot;BO&quot; &quot;OO&quot;</code></pre>
<pre class="r"><code>table(genotypeF)</code></pre>
<pre><code>## genotypeF
## AA AB AO BB BO OO 
##  2  2  7  3  3  2</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
rbinom(15, prob = 0.5, size = 1)</code></pre>
<pre><code>##  [1] 1 1 1 1 1 1 1 1 0 1 0 0 0 0 0</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
rbinom(12, prob = 2/3, size = 1)</code></pre>
<pre><code>##  [1] 1 1 1 0 0 0 1 1 0 0 1 1</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
rbinom(1, prob = 2/3, size = 12)</code></pre>
<pre><code>## [1] 9</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
set.seed(235569515)
rbinom(1, prob = 0.3, size = 15)</code></pre>
<pre><code>## [1] 5</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
probabilities = dbinom(0:15, prob = 0.3, size = 15)
round(probabilities, 2)</code></pre>
<pre><code>##  [1] 0.00 0.03 0.09 0.17 0.22 0.21 0.15 0.08 0.03 0.01 0.00 0.00 0.00 0.00 0.00
## [16] 0.00</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
barplot(probabilities, names.arg = 0:15, col = &quot;red&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-8-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
plot(dbinom(0:12, prob = 5e-4, size = 1e4), dpois(0:12, lambda = 5), asp = 1)
abline(a = 0, b = 1, col = &quot;blue&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-8-2.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
5^3 * exp(-5) / factorial(3)</code></pre>
<pre><code>## [1] 0.1403739</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
rbinom(1, prob = 5e-4, size = 10000)</code></pre>
<pre><code>## [1] 6</code></pre>
<pre class="r"><code>simulations = rbinom(n = 300000, prob = 5e-4, size = 10000)
barplot(table(simulations), col = &quot;lavender&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-10-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
`[&lt;-`(rep(0, 100), 22, 1)</code></pre>
<pre><code>##   [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
##  [38] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
##  [75] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
s100 = rpois(100, lambda=0.5)
s100</code></pre>
<pre><code>##   [1] 0 0 1 0 0 0 0 0 1 0 0 1 0 0 0 1 0 0 1 0 0 1 1 0 0 0 0 0 0 0 0 1 1 1 2 0 0
##  [38] 0 1 2 0 1 0 1 2 0 2 0 1 2 1 0 1 0 0 0 0 0 1 0 0 1 2 0 0 0 1 0 1 2 0 0 0 0
##  [75] 0 0 0 0 1 0 1 1 1 0 0 0 0 0 0 0 0 1 0 0 0 1 0 4 3 1</code></pre>
<pre class="r"><code>barplot(s100, ylim = c(0, 7), width = 0.7, xlim = c(-0.5,100.5),
  names.arg = seq(along = s100), col=&quot;lavender&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-12-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
## set.seed(8969311)
## e100 = rpois(100,lambda = 0.5)
## e100[42] = 7
## save(e100, file = &quot;../data/e100.RData&quot;)

## -----------------------------------------------------------------------------
setwd(&quot;D:/books/R/blogdown/hugo&quot;)
load(&quot;../data/e100.RData&quot;)
barplot(e100, ylim = c(0, 7), width = 0.7, xlim = c(-0.5, 100.5),
  names.arg = seq(along = e100), col = &quot;darkolivegreen&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-13-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
barplot(e100, ylim = c(0, 7), width = 0.7, xlim = c(-0.5, 100.5),
  names.arg = seq(along = e100), col = &quot;darkolivegreen&quot;)
text(35, 7, adj = c(-0.05, 0.5), labels = &quot;?&quot;, xpd = NA, col = &quot;red&quot;,
  cex = 1.25, font = 2)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-14-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
1 - ppois(6, 0.5)</code></pre>
<pre><code>## [1] 1.00238e-06</code></pre>
<pre class="r"><code>ppois(6, 0.5, lower.tail = FALSE)</code></pre>
<pre><code>## [1] 1.00238e-06</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
maxes = replicate(100000, {
  max(rpois(100, 0.5))
})
table(maxes)</code></pre>
<pre><code>## maxes
##     1     2     3     4     5     6     7     9 
##     7 23028 60840 14364  1604   141    15     1</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
mean( maxes &gt;= 7 )</code></pre>
<pre><code>## [1] 0.00016</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
dmultinom(c(4, 2, 0, 0), prob = rep(1/4, 4))</code></pre>
<pre><code>## [1] 0.003662109</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
pvec = rep(1/4, 4)
t(rmultinom(1, prob = pvec, size = 8))</code></pre>
<pre><code>##      [,1] [,2] [,3] [,4]
## [1,]    1    3    1    3</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
pvec = rep(1/4, 4)
t(rmultinom(1, prob = pvec, size = 80))</code></pre>
<pre><code>##      [,1] [,2] [,3] [,4]
## [1,]   23   22   15   20</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
pvec = rep(1/4, 4)
t(rmultinom(8, prob = pvec, size = 1))</code></pre>
<pre><code>##      [,1] [,2] [,3] [,4]
## [1,]    0    0    0    1
## [2,]    1    0    0    0
## [3,]    0    0    1    0
## [4,]    1    0    0    0
## [5,]    0    0    0    1
## [6,]    1    0    0    0
## [7,]    1    0    0    0
## [8,]    0    0    0    1</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
obsunder0 = rmultinom(1000, prob = pvec, size = 20)
dim(obsunder0)</code></pre>
<pre><code>## [1]    4 1000</code></pre>
<pre class="r"><code>obsunder0[, 1:11]</code></pre>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11]
## [1,]    7    4    3    3    4    5    1    2    7     4     5
## [2,]    4    9    3    6    6    2    6    8    5     5     4
## [3,]    6    4    6    5    3    9    5    5    4     3     4
## [4,]    3    3    8    6    7    4    8    5    4     8     7</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
thep = unique(pvec); stopifnot(length(thep)==1, thep == 0.25)
thep</code></pre>
<pre><code>## [1] 0.25</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
expected0 = pvec * 20
sum((obsunder0[, 1] - expected0)^2 / expected0)</code></pre>
<pre><code>## [1] 2</code></pre>
<pre class="r"><code>sum((obsunder0[, 2] - expected0)^2 / expected0)</code></pre>
<pre><code>## [1] 4.4</code></pre>
<pre class="r"><code>sum((obsunder0[, 3] - expected0)^2 / expected0)</code></pre>
<pre><code>## [1] 3.6</code></pre>
<pre class="r"><code>(5-5)^2/5+(4-5)^2/5+(6-5)^2/5+(5-5)^2/5</code></pre>
<pre><code>## [1] 0.4</code></pre>
<pre class="r"><code>(4-5)^2/5+(6-5)^2/5+(6-5)^2/5+(4-5)^2/5</code></pre>
<pre><code>## [1] 0.8</code></pre>
<pre class="r"><code>(6-5)^2/5+(4-5)^2/5+(4-5)^2/5+(6-5)^2/5</code></pre>
<pre><code>## [1] 0.8</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
stat = function(obsvd, exptd = 20 * pvec) {
  sum((obsvd - exptd)^2 / exptd)
}
stat(obsunder0[, 1])</code></pre>
<pre><code>## [1] 2</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
S0 = apply(obsunder0, 2, stat)
summary(S0)</code></pre>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##   0.000   1.200   2.800   3.086   4.000  16.400</code></pre>
<pre class="r"><code>hist(S0, breaks = 25, col = &quot;lavender&quot;, main = &quot;&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-28-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
q95 = quantile(S0, probs = 0.95)
q95</code></pre>
<pre><code>## 95% 
## 7.6</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
## ## This was done to save this object for its reuse in Chapter 2.
## save(S0, file = &quot;../data/S0.RData&quot;)

## -----------------------------------------------------------------------------
pvecA = c(3/8, 1/4, 1/4, 1/8)
observed = rmultinom(1000, prob = pvecA, size = 20)
dim(observed)</code></pre>
<pre><code>## [1]    4 1000</code></pre>
<pre class="r"><code>observed[, 1:7]</code></pre>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7]
## [1,]    3    6    7    9   15   10    9
## [2,]    9    3    5    3    2    5    6
## [3,]    7    9    4    4    3    3    2
## [4,]    1    2    4    4    0    2    3</code></pre>
<pre class="r"><code>apply(observed, 1, mean)</code></pre>
<pre><code>## [1] 7.459 4.926 5.100 2.515</code></pre>
<pre class="r"><code>expectedA = pvecA * 20
expectedA</code></pre>
<pre><code>## [1] 7.5 5.0 5.0 2.5</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
stat(observed[, 1])</code></pre>
<pre><code>## [1] 8</code></pre>
<pre class="r"><code>S1 = apply(observed, 2, stat)
q95 # can&#39;t reject the first observation, because the statistic is within the 95% percentile</code></pre>
<pre><code>## 95% 
## 7.6</code></pre>
<pre class="r"><code>sum(S1 &gt; q95)</code></pre>
<pre><code>## [1] 197</code></pre>
<pre class="r"><code>power = mean(S1 &gt; q95)
power</code></pre>
<pre><code>## [1] 0.197</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
#stopifnot(stat(observed[, 1]) &lt; q95)</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
dbinom(2, size = 10, prob = 0.3) # run size = 10, prob = 0.3 binomial distribution, the probability at 2</code></pre>
<pre><code>## [1] 0.2334744</code></pre>
<pre class="r"><code>pbinom(2, size = 10, prob = 0.3) # run size = 10, prob = 0.3 binomial distribution, the probability &lt;= 2</code></pre>
<pre><code>## [1] 0.3827828</code></pre>
<pre class="r"><code>sum(dbinom(0:2, size = 10, prob = 0.3)) </code></pre>
<pre><code>## [1] 0.3827828</code></pre>
<p>Whenever we note that we keep needing a certain sequence of commands, it’s good to put them into a function. The function body contains the instructions that we want to do over and over again, the function arguments take those things that we may want to vary. Write a function to compute the probability of having a maximum as big as <code>m</code> when looking across <code>n</code> Poisson variables with rate <code>lambda</code>.</p>
<pre class="r"><code>## -----------------------------------------------------------------------------
poismax = function(lambda, n, m) {
  epsilon = 1 - ppois(m - 1, lambda)
  1 - exp( -n * epsilon)
}
poismax(lambda = 0.5, n = 100, m = 7)</code></pre>
<pre><code>## [1] 0.0001002329</code></pre>
<pre class="r"><code>poismax(lambda = mean(e100), n = 100, m = 7)</code></pre>
<pre><code>## [1] 0.0001870183</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
poismax = function(lambda, n = 100, m = 7) {
  1 - exp( -n * (1 - ppois(m - 1, lambda)))
}
poismax(0.5)</code></pre>
<pre><code>## [1] 0.0001002329</code></pre>
<pre class="r"><code>poismax(0.5, m = 9)</code></pre>
<pre><code>## [1] 3.43549e-07</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
## if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
##     install.packages(&quot;BiocManager&quot;)
## BiocManager::install(c(&quot;Biostrings&quot;, &quot;BSgenome.Celegans.UCSC.ce2&quot;))

## -----------------------------------------------------------------------------
library(&quot;BSgenome.Celegans.UCSC.ce2&quot;)</code></pre>
<pre><code>## Loading required package: BSgenome</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which.max, which.min</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## 
## Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     expand.grid, I, unname</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## 
## Attaching package: &#39;IRanges&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:grDevices&#39;:
## 
##     windows</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: Biostrings</code></pre>
<pre><code>## Loading required package: XVector</code></pre>
<pre><code>## 
## Attaching package: &#39;Biostrings&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     strsplit</code></pre>
<pre><code>## Loading required package: rtracklayer</code></pre>
<pre class="r"><code>Celegans</code></pre>
<pre><code>## Worm genome:
## # organism: Caenorhabditis elegans (Worm)
## # genome: ce2
## # provider: UCSC
## # release date: Mar. 2004
## # 7 sequences:
## #   chrI   chrII  chrIII chrIV  chrV   chrX   chrM                              
## # (use &#39;seqnames()&#39; to see all the sequence names, use the &#39;$&#39; or &#39;[[&#39; operator
## # to access a given sequence)</code></pre>
<pre class="r"><code>seqnames(Celegans)</code></pre>
<pre><code>## [1] &quot;chrI&quot;   &quot;chrII&quot;  &quot;chrIII&quot; &quot;chrIV&quot;  &quot;chrV&quot;   &quot;chrX&quot;   &quot;chrM&quot;</code></pre>
<pre class="r"><code>Celegans$chrM</code></pre>
<pre><code>## 13794-letter DNAString object
## seq: CAGTAAATAGTTTAATAAAAATATAGCATTTGGGTT...TATTTATAGATATATACTTTGTATATATCTATATTA</code></pre>
<pre class="r"><code>class(Celegans$chrM)</code></pre>
<pre><code>## [1] &quot;DNAString&quot;
## attr(,&quot;package&quot;)
## [1] &quot;Biostrings&quot;</code></pre>
<pre class="r"><code>length(Celegans$chrM)</code></pre>
<pre><code>## [1] 13794</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;Biostrings&quot;)
lfM = letterFrequency(Celegans$chrM, letters=c(&quot;A&quot;, &quot;C&quot;, &quot;G&quot;, &quot;T&quot;))
lfM</code></pre>
<pre><code>##    A    C    G    T 
## 4335 1225 2055 6179</code></pre>
<pre class="r"><code>sum(lfM)</code></pre>
<pre><code>## [1] 13794</code></pre>
<pre class="r"><code>lfM / sum(lfM)</code></pre>
<pre><code>##          A          C          G          T 
## 0.31426707 0.08880673 0.14897782 0.44794838</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
t(rmultinom(1, length(Celegans$chrM), p = rep(1/4, 4)))</code></pre>
<pre><code>##      [,1] [,2] [,3] [,4]
## [1,] 3463 3397 3464 3470</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
length(Celegans$chrM) / 4</code></pre>
<pre><code>## [1] 3448.5</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
oestat = function(o, e) {
  sum((o-e)^2 / e)
}
oe = oestat(o = lfM, e = length(Celegans$chrM) / 4)
oe</code></pre>
<pre><code>## [1] 4386.634</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
B = 10000
n = length(Celegans$chrM)
expected = rep(n / 4, 4)
oenull = replicate(B,
  oestat(e = expected, o = rmultinom(1, n, p = rep(1/4, 4))))


## -----------------------------------------------------------------------------
hist(oenull, breaks = 100, col = &quot;skyblue&quot;, main = &quot;&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-44-1.svg" width="576" /></p>
<pre class="r"><code>q95 = quantile(oenull, probs = 0.95)
q95</code></pre>
<pre><code>##     95% 
## 7.84544</code></pre>
<pre class="r"><code>mean(oenull &gt; 10)</code></pre>
<pre><code>## [1] 0.0178</code></pre>
<pre class="r"><code>?Distributions</code></pre>
<pre><code>## starting httpd help server ... done</code></pre>
<p>For the Cauchy distribution see <code>dcauchy</code>.</p>
<p>For the chi-squared distribution see <code>dchisq</code>.</p>
<p>For the exponential distribution see <code>dexp</code>.</p>
<p>For the F distribution see <code>df</code>.</p>
<p>For the gamma distribution see <code>dgamma</code>.</p>
<p>For the geometric distribution see <code>dgeom</code>. (This is also a special case of the negative binomial.)</p>
<p>For the hypergeometric distribution see <code>dhyper</code>.</p>
<p>For the log-normal distribution see <code>dlnorm</code>.</p>
<p>For the multinomial distribution see <code>dmultinom</code>.</p>
<p>For the negative binomial distribution see <code>dnbinom</code>.</p>
<p>For the normal distribution see <code>dnorm</code>.</p>
<p>For the Poisson distribution see <code>dpois</code>.</p>
<p>For the Student’s t distribution see <code>dt</code>.</p>
<p>For the uniform distribution see <code>dunif</code>.</p>
<p>For the Weibull distribution see <code>dweibull</code>.</p>
<pre class="r"><code>posiv = rpois(n=1000, lambda=3)
mean(posiv)</code></pre>
<pre><code>## [1] 3.037</code></pre>
<pre class="r"><code>var(posiv)</code></pre>
<pre><code>## [1] 3.056688</code></pre>
</div>
<div id="statistical-modeling" class="section level2">
<h2>2 Statistical Modeling</h2>
<pre class="r"><code>rpois(100, 0.5)</code></pre>
<pre><code>##   [1] 1 0 0 1 1 0 1 2 0 0 0 0 2 0 1 1 1 1 1 0 1 0 0 1 1 0 0 1 3 1 1 1 1 1 1 1 1
##  [38] 0 0 1 0 1 0 0 1 0 0 0 0 0 0 2 1 0 0 1 0 1 0 2 0 0 0 0 0 0 1 0 0 1 0 0 0 0
##  [75] 1 0 2 1 2 0 4 0 0 2 1 2 1 1 0 1 1 1 1 0 0 0 0 0 0 0</code></pre>
<pre class="r"><code>poismax = function(lambda, n, m) {
  epsilon = 1 - ppois(m - 1, lambda)
  1 - exp( -n * epsilon)
}
poismax(lambda = 0.5, n = 100, m = 7)</code></pre>
<pre><code>## [1] 0.0001002329</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
setwd(&quot;D:/books/R/blogdown/hugo&quot;)
load(&quot;../data/e100.RData&quot;)
e99 = e100[-which.max(e100)]
e100</code></pre>
<pre><code>##   [1] 2 0 1 0 0 0 2 1 0 0 0 0 0 0 0 0 0 1 1 1 0 0 1 0 1 0 0 0 1 0 0 0 0 1 0 1 1
##  [38] 0 1 2 2 7 1 0 2 0 1 0 1 1 1 1 0 1 0 0 0 0 1 2 2 1 0 0 0 0 0 1 0 0 0 1 1 0
##  [75] 0 1 1 0 0 0 0 1 0 0 0 1 0 1 0 0 1 1 0 0 1 1 0 0 1 0</code></pre>
<pre class="r"><code>e99</code></pre>
<pre><code>##  [1] 2 0 1 0 0 0 2 1 0 0 0 0 0 0 0 0 0 1 1 1 0 0 1 0 1 0 0 0 1 0 0 0 0 1 0 1 1 0
## [39] 1 2 2 1 0 2 0 1 0 1 1 1 1 0 1 0 0 0 0 1 2 2 1 0 0 0 0 0 1 0 0 0 1 1 0 0 1 1
## [77] 0 0 0 0 1 0 0 0 1 0 1 0 0 1 1 0 0 1 1 0 0 1 0</code></pre>
<pre class="r"><code>which.max(e100)</code></pre>
<pre><code>## [1] 42</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
barplot(table(e99), space = 0.8, col = &quot;chartreuse4&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-53-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;vcd&quot;)</code></pre>
<pre><code>## Loading required package: grid</code></pre>
<pre><code>## 
## Attaching package: &#39;grid&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:Biostrings&#39;:
## 
##     pattern</code></pre>
<pre><code>## 
## Attaching package: &#39;vcd&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:GenomicRanges&#39;:
## 
##     tile</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     tile</code></pre>
<pre class="r"><code>gf1 = goodfit( e99, &quot;poisson&quot;)
rootogram(gf1, xlab = &quot;&quot;, rect_gp = gpar(fill = &quot;chartreuse4&quot;))</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-54-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
simp = rpois(100, lambda = 0.05)
gf2 = goodfit(simp, &quot;poisson&quot;)
rootogram(gf2, xlab = &quot;&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-55-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
table(e100)</code></pre>
<pre><code>## e100
##  0  1  2  7 
## 58 34  7  1</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
table(rpois(100, 3))</code></pre>
<pre><code>## 
##  0  1  2  3  4  5  6  8  9 
##  2 19 24 17 19 10  4  4  1</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
counts  =  table(e100)
stopifnot(identical(names(counts), c(&quot;0&quot;, &quot;1&quot;, &quot;2&quot;, &quot;7&quot;)), all(counts==c(58, 34, 7, 1)))</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
prod(dpois(c(0, 1, 2, 7), lambda = 3) ^ (c(58, 34, 7, 1)))</code></pre>
<pre><code>## [1] 1.392143e-110</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
prod(dpois(c(0, 1, 2, 7), lambda = 0.4) ^ (c(58, 34, 7, 1)))</code></pre>
<pre><code>## [1] 8.5483e-46</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
prod(dpois(c(0, 1, 2, 7), lambda = 0) ^ (c(58, 34, 7, 1)))</code></pre>
<pre><code>## [1] 0</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
loglikelihood  =  function(lambda, data = e100) {
  sum(log(dpois(data, lambda)))
}

## -----------------------------------------------------------------------------
lambdas = seq(0.05, 0.95, length = 100)
loglik = vapply(lambdas, loglikelihood, numeric(1))
plot(lambdas, loglik, type = &quot;l&quot;, col = &quot;red&quot;, ylab = &quot;&quot;, lwd = 2,
     xlab = expression(lambda))
m0 = mean(e100)
abline(v = m0, col = &quot;blue&quot;, lwd = 2)
abline(h = loglikelihood(m0), col = &quot;purple&quot;, lwd = 2)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-62-1.svg" width="576" /></p>
<pre class="r"><code>m0</code></pre>
<pre><code>## [1] 0.55</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
gf  =  goodfit(e100, &quot;poisson&quot;)
names(gf)</code></pre>
<pre><code>## [1] &quot;observed&quot; &quot;count&quot;    &quot;fitted&quot;   &quot;type&quot;     &quot;method&quot;   &quot;df&quot;       &quot;par&quot;</code></pre>
<pre class="r"><code>gf$par</code></pre>
<pre><code>## $lambda
## [1] 0.55</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
cb  =  c(rep(0, 110), rep(1, 10))

## -----------------------------------------------------------------------------
table(cb)</code></pre>
<pre><code>## cb
##   0   1 
## 110  10</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
mean(cb)</code></pre>
<pre><code>## [1] 0.08333333</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
probs  =  seq(0, 0.3, by = 0.001)
likelihood = dbinom(sum(cb), prob = probs, size = length(cb))
plot(probs, likelihood, pch = 16, xlab = &quot;probability of success&quot;,
       ylab = &quot;likelihood&quot;, cex=0.6)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-67-1.svg" width="576" /></p>
<pre class="r"><code>probs[which.max(likelihood)]</code></pre>
<pre><code>## [1] 0.083</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
stopifnot(abs(probs[which.max(likelihood)]-1/12) &lt; diff(probs[1:2]))</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
loglikelihood = function(p, n = 300, y = 40) {
  log(choose(n, y)) + y * log(p) + (n - y) * log(1 - p)
}

## -----------------------------------------------------------------------------
p_seq = seq(0, 1, by = 0.001)
plot(p_seq, loglikelihood(p_seq), xlab = &quot;p&quot;, ylab = &quot;log f(p|y)&quot;, type = &quot;l&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-70-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;Biostrings&quot;)
setwd(&quot;D:/books/R/blogdown/hugo&quot;)
staph = readDNAStringSet(&quot;../data/staphsequence.ffn.txt&quot;, &quot;fasta&quot;)

## -----------------------------------------------------------------------------
staph[1]</code></pre>
<pre><code>## DNAStringSet object of length 1:
##     width seq                                               names               
## [1]  1362 ATGTCGGAAAAAGAAATTTGGGA...AAAAAGAAATAAGAAATGTATAA lcl|NC_002952.2_c...</code></pre>
<p>The double square brackets <code>[[i]]</code> extract the sequence of the <code>i</code>-th gene as a DNAString, as opposed to the pair of single brackets <code>[i]</code>, which return a DNAStringSet with just a single DNAString in it. If you look at the length of <code>staph[1]</code>, it is <code>1</code>, whereas <code>staph[[1]]</code> has length <code>1362</code>.</p>
<pre class="r"><code>letterFrequency(staph[[1]], letters = &quot;ACGT&quot;, OR = 0)</code></pre>
<pre><code>##   A   C   G   T 
## 522 219 229 392</code></pre>
<pre class="r"><code>length(staph[1])</code></pre>
<pre><code>## [1] 1</code></pre>
<pre class="r"><code>length(staph[[1]])</code></pre>
<pre><code>## [1] 1362</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
letterFrq = vapply(staph, letterFrequency, FUN.VALUE = numeric(4),
         letters = &quot;ACGT&quot;, OR = 0)
colnames(letterFrq) = paste0(&quot;gene&quot;, seq(along = staph))
tab10 = letterFrq[, 1:10]
tab10</code></pre>
<pre><code>##   gene1 gene2 gene3 gene4 gene5 gene6 gene7 gene8 gene9 gene10
## A   522   413    85   411   685   887   275   510   487    191
## C   219   176    31   168   293   395   137   244   180    111
## G   229   193    56   207   423   586   169   316   263    142
## T   392   352    74   327   531   793   250   445   357    252</code></pre>
<pre class="r"><code>computeProportions = function(x) { x/sum(x) }
prop10 = apply(tab10, 2, computeProportions) # nucleotides Proportions in column
round(prop10, digits = 2)</code></pre>
<pre><code>##   gene1 gene2 gene3 gene4 gene5 gene6 gene7 gene8 gene9 gene10
## A  0.38  0.36  0.35  0.37  0.35  0.33  0.33  0.34  0.38   0.27
## C  0.16  0.16  0.13  0.15  0.15  0.15  0.16  0.16  0.14   0.16
## G  0.17  0.17  0.23  0.19  0.22  0.22  0.20  0.21  0.20   0.20
## T  0.29  0.31  0.30  0.29  0.27  0.30  0.30  0.29  0.28   0.36</code></pre>
<pre class="r"><code>p0 = rowMeans(prop10) # nucleotides Proportions in all 10 genes
p0</code></pre>
<pre><code>##         A         C         G         T 
## 0.3470531 0.1518313 0.2011442 0.2999714</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
cs = colSums(tab10) # 10 genes length
cs</code></pre>
<pre><code>##  gene1  gene2  gene3  gene4  gene5  gene6  gene7  gene8  gene9 gene10 
##   1362   1134    246   1113   1932   2661    831   1515   1287    696</code></pre>
<pre class="r"><code>expectedtab10 = outer(p0, cs, FUN = &quot;*&quot;) # expected nucleotides Proportions if every gene has the same Proportion
round(expectedtab10)</code></pre>
<pre><code>##   gene1 gene2 gene3 gene4 gene5 gene6 gene7 gene8 gene9 gene10
## A   473   394    85   386   671   924   288   526   447    242
## C   207   172    37   169   293   404   126   230   195    106
## G   274   228    49   224   389   535   167   305   259    140
## T   409   340    74   334   580   798   249   454   386    209</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
randomtab10 = sapply(cs, function(s) { rmultinom(1, s, p0) } )
randomtab10</code></pre>
<pre><code>##      gene1 gene2 gene3 gene4 gene5 gene6 gene7 gene8 gene9 gene10
## [1,]   464   388    82   384   674   930   322   494   413    226
## [2,]   231   175    48   166   313   425   118   243   179    116
## [3,]   260   234    51   230   380   555   160   341   296    137
## [4,]   407   337    65   333   565   751   231   437   399    217</code></pre>
<pre class="r"><code>all(colSums(randomtab10) == cs)</code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
stopifnot(all(colSums(randomtab10) == cs))</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
stat = function(obsvd, exptd) {
   sum((obsvd - exptd)^2 / exptd)
}
B = 1000
simulstat = replicate(B, {
  randomtab10 = sapply(cs, function(s) { rmultinom(1, s, p0) })
  stat(randomtab10, expectedtab10)
})
S1 = stat(tab10, expectedtab10)
sum(simulstat &gt;= S1)</code></pre>
<pre><code>## [1] 0</code></pre>
<pre class="r"><code>length(simulstat)</code></pre>
<pre><code>## [1] 1000</code></pre>
<pre class="r"><code>head(simulstat)</code></pre>
<pre><code>## [1] 40.43308 25.31234 27.95283 17.37396 34.72065 29.15233</code></pre>
<pre class="r"><code>expectedtab10</code></pre>
<pre><code>##      gene1    gene2    gene3    gene4    gene5    gene6    gene7    gene8
## A 472.6864 393.5582 85.37507 386.2701 670.5066 923.5084 288.4011 525.7855
## C 206.7942 172.1767 37.35049 168.9882 293.3380 404.0230 126.1718 230.0244
## G 273.9584 228.0975 49.48147 223.8735 388.6105 535.2446 167.1508 304.7334
## T 408.5611 340.1676 73.79297 333.8682 579.5448 798.2239 249.2762 454.4567
##      gene9   gene10
## A 446.6574 241.5490
## C 195.4069 105.6746
## G 258.8726 139.9963
## T 386.0632 208.7801</code></pre>
<pre class="r"><code>randomtab10</code></pre>
<pre><code>##      gene1 gene2 gene3 gene4 gene5 gene6 gene7 gene8 gene9 gene10
## [1,]   464   388    82   384   674   930   322   494   413    226
## [2,]   231   175    48   166   313   425   118   243   179    116
## [3,]   260   234    51   230   380   555   160   341   296    137
## [4,]   407   337    65   333   565   751   231   437   399    217</code></pre>
<pre class="r"><code>hist(simulstat, col = &quot;lavender&quot;, breaks = seq(0, 75, length.out=50))
abline(v = S1, col = &quot;red&quot;)
abline(v = quantile(simulstat, probs = c(0.95, 0.99)),
       col = c(&quot;darkgreen&quot;, &quot;blue&quot;), lty = 2)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-85-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
stopifnot(max(simulstat)&lt;75, S1&lt;75)</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
qs = ppoints(100)  #Generates the sequence of probability points 0.005 to 0.995
quantile(simulstat, qs)</code></pre>
<pre><code>##     0.5%     1.5%     2.5%     3.5%     4.5%     5.5%     6.5%     7.5% 
## 14.98844 16.28320 17.50259 18.24076 18.62344 19.07341 19.66079 19.91208 
##     8.5%     9.5%    10.5%    11.5%    12.5%    13.5%    14.5%    15.5% 
## 20.29684 20.62778 20.93192 21.10932 21.40516 21.64190 21.85037 22.13960 
##    16.5%    17.5%    18.5%    19.5%    20.5%    21.5%    22.5%    23.5% 
## 22.53136 22.93293 23.14039 23.41548 23.60116 23.95936 24.17024 24.42954 
##    24.5%    25.5%    26.5%    27.5%    28.5%    29.5%    30.5%    31.5% 
## 24.69359 24.92622 24.99886 25.16888 25.33169 25.50434 25.62677 25.83369 
##    32.5%    33.5%    34.5%    35.5%    36.5%    37.5%    38.5%    39.5% 
## 25.98940 26.28855 26.63070 26.81300 26.98819 27.12084 27.30601 27.47563 
##    40.5%    41.5%    42.5%    43.5%    44.5%    45.5%    46.5%    47.5% 
## 27.59761 27.78386 28.02836 28.18457 28.51322 28.66197 28.88965 29.00322 
##    48.5%    49.5%    50.5%    51.5%    52.5%    53.5%    54.5%    55.5% 
## 29.19831 29.43226 29.75325 29.87609 30.04020 30.18108 30.37550 30.63155 
##    56.5%    57.5%    58.5%    59.5%    60.5%    61.5%    62.5%    63.5% 
## 30.83079 31.01395 31.22237 31.45265 31.74654 31.90807 32.30407 32.55708 
##    64.5%    65.5%    66.5%    67.5%    68.5%    69.5%    70.5%    71.5% 
## 32.80945 32.95201 33.29111 33.50298 33.80705 33.94406 34.16076 34.33740 
##    72.5%    73.5%    74.5%    75.5%    76.5%    77.5%    78.5%    79.5% 
## 34.55167 34.79766 35.08447 35.45335 35.71938 35.85952 36.19083 36.42843 
##    80.5%    81.5%    82.5%    83.5%    84.5%    85.5%    86.5%    87.5% 
## 36.75272 37.05766 37.49101 37.94178 38.20867 38.60341 38.92721 39.44580 
##    88.5%    89.5%    90.5%    91.5%    92.5%    93.5%    94.5%    95.5% 
## 39.92408 40.43715 41.02038 41.56662 42.35350 43.56930 44.39798 44.93912 
##    96.5%    97.5%    98.5%    99.5% 
## 45.83205 48.84615 50.06561 54.04509</code></pre>
<pre class="r"><code>qchisq(qs, df = 30)</code></pre>
<pre><code>##   [1] 13.78672 15.71876 16.79077 17.57614 18.21194 18.75446 19.23275 19.66390
##   [9] 20.05886 20.42512 20.76806 21.09166 21.39894 21.69231 21.97367 22.24457
##  [17] 22.50629 22.75990 23.00632 23.24631 23.48056 23.70963 23.93405 24.15428
##  [25] 24.37070 24.58370 24.79358 25.00066 25.20519 25.40743 25.60760 25.80592
##  [33] 26.00258 26.19776 26.39164 26.58439 26.77614 26.96707 27.15729 27.34696
##  [41] 27.53620 27.72513 27.91389 28.10259 28.29136 28.48031 28.66957 28.85924
##  [49] 29.04945 29.24032 29.43196 29.62450 29.81806 30.01277 30.20876 30.40616
##  [57] 30.60512 30.80577 31.00828 31.21280 31.41949 31.62854 31.84012 32.05444
##  [65] 32.27170 32.49213 32.71597 32.94349 33.17495 33.41067 33.65098 33.89624
##  [73] 34.14685 34.40324 34.66590 34.93536 35.21223 35.49719 35.79099 36.09449
##  [81] 36.40869 36.73473 37.07391 37.42780 37.79820 38.18728 38.59768 39.03260
##  [89] 39.49603 39.99300 40.53003 41.11573 41.76190 42.48522 43.31057 44.27745
##  [97] 45.45461 46.97924 49.19886 53.67196</code></pre>
<pre class="r"><code>hist(rchisq(1000, df = 30), col = &quot;lavender&quot;, breaks = 50)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-89-1.svg" width="576" /></p>
<pre class="r"><code>quantile(qchisq(qs, df = 30), qs)</code></pre>
<pre><code>##     0.5%     1.5%     2.5%     3.5%     4.5%     5.5%     6.5%     7.5% 
## 14.74308 16.23869 17.16382 17.87179 18.45879 18.96730 19.42030 19.83175 
##     8.5%     9.5%    10.5%    11.5%    12.5%    13.5%    14.5%    15.5% 
## 20.21086 20.56401 20.89588 21.20996 21.50896 21.79501 22.06984 22.33486 
##    16.5%    17.5%    18.5%    19.5%    20.5%    21.5%    22.5%    23.5% 
## 22.59125 22.83999 23.08192 23.31776 23.54813 23.77359 23.99461 24.21163 
##    24.5%    25.5%    26.5%    27.5%    28.5%    29.5%    30.5%    31.5% 
## 24.42502 24.63512 24.84224 25.04668 25.24867 25.44846 25.64627 25.84230 
##    32.5%    33.5%    34.5%    35.5%    36.5%    37.5%    38.5%    39.5% 
## 26.03673 26.22975 26.42152 26.61219 26.80192 26.99084 27.17910 27.36683 
##    40.5%    41.5%    42.5%    43.5%    44.5%    45.5%    46.5%    47.5% 
## 27.55414 27.74118 27.92804 28.11486 28.30175 28.48883 28.67621 28.86400 
##    48.5%    49.5%    50.5%    51.5%    52.5%    53.5%    54.5%    55.5% 
## 29.05231 29.24127 29.43100 29.62161 29.81322 30.00595 30.19994 30.39530 
##    56.5%    57.5%    58.5%    59.5%    60.5%    61.5%    62.5%    63.5% 
## 30.59219 30.79073 30.99107 31.19337 31.39779 31.60450 31.81367 32.02550 
##    64.5%    65.5%    66.5%    67.5%    68.5%    69.5%    70.5%    71.5% 
## 32.24019 32.45796 32.67904 32.90367 33.13213 33.36471 33.60172 33.84351 
##    72.5%    73.5%    74.5%    75.5%    76.5%    77.5%    78.5%    79.5% 
## 34.09046 34.34299 34.60155 34.86665 35.13886 35.41883 35.70725 36.00496 
##    80.5%    81.5%    82.5%    83.5%    84.5%    85.5%    86.5%    87.5% 
## 36.31286 36.63203 36.96368 37.30925 37.67041 38.04916 38.44789 38.86951 
##    88.5%    89.5%    90.5%    91.5%    92.5%    93.5%    94.5%    95.5% 
## 39.31761 39.79669 40.31253 40.87266 41.48728 42.17057 42.94329 43.83752 
##    96.5%    97.5%    98.5%    99.5% 
## 44.90723 46.25504 48.12235 51.45778</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
qqplot(qchisq(ppoints(B), df = 30), simulstat, main = &quot;&quot;,
  xlab = expression(chi[nu==30]^2), asp = 1, cex = 0.5, pch = 16)
abline(a = 0, b = 1, col = &quot;red&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-91-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
1 - pchisq(S1, df = 30)</code></pre>
<pre><code>## [1] 4.74342e-05</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
setwd(&quot;D:/books/R/blogdown/hugo&quot;)
load(&quot;../data/ChargaffTable.RData&quot;)
ChargaffTable</code></pre>
<pre><code>##                   A    T    C    G
## Human-Thymus   30.9 29.4 19.9 19.8
## Mycobac.Tuber  15.1 14.6 34.9 35.4
## Chicken-Eryth. 28.8 29.2 20.5 21.5
## Sheep-liver    29.3 29.3 20.5 20.7
## Sea Urchin     32.8 32.1 17.7 17.3
## Wheat          27.3 27.1 22.7 22.8
## Yeast          31.3 32.9 18.7 17.1
## E.coli         24.7 23.6 26.0 25.7</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
stopifnot(nrow(ChargaffTable) == 8)
mycolors = c(&quot;chocolate&quot;, &quot;aquamarine4&quot;, &quot;cadetblue4&quot;, &quot;coral3&quot;,
            &quot;chartreuse4&quot;,&quot;darkgoldenrod4&quot;,&quot;darkcyan&quot;,&quot;brown4&quot;)
par(mfrow=c(2, 4), mai = c(0, 0.7, 0.7, 0))
for (i in 1:8) {
  cbp = barplot(ChargaffTable[i, ], horiz = TRUE, axes = FALSE, axisnames = FALSE, col = mycolors[i])
  ax = axis(3, las = 2, labels = FALSE, col = mycolors[i], cex = 0.5, at = c(0, 10, 20))
  mtext(side = 3, at = ax,  text = paste(ax), col = mycolors[i], line = 0, las = 1, cex = 0.9)
  mtext(side = 2, at = cbp, text = colnames(ChargaffTable), col = mycolors[i], line = 0, las = 2, cex = 1)
  title(paste(rownames(ChargaffTable)[i]), col = mycolors[i], cex = 1.1)
}</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-94-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
statChf = function(x){
  sum((x[, &quot;C&quot;] - x[, &quot;G&quot;])^2 + (x[, &quot;A&quot;] - x[, &quot;T&quot;])^2)
}
chfstat = statChf(ChargaffTable)
chfstat</code></pre>
<pre><code>## [1] 11.08</code></pre>
<pre class="r"><code>permstat = replicate(100000, {
     permuted = t(apply(ChargaffTable, 1, sample))
     colnames(permuted) = colnames(ChargaffTable)
     statChf(permuted)
})
pChf = mean(permstat &lt;= chfstat)
pChf</code></pre>
<pre><code>## [1] 0.00012</code></pre>
<pre class="r"><code>hist(permstat, breaks = 100, main = &quot;&quot;, col = &quot;lavender&quot;)
abline(v = chfstat, lwd = 2, col = &quot;red&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-97-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
HairEyeColor[,, &quot;Female&quot;]</code></pre>
<pre><code>##        Eye
## Hair    Brown Blue Hazel Green
##   Black    36    9     5     2
##   Brown    66   34    29    14
##   Red      16    7     7     7
##   Blond     4   64     5     8</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
str(HairEyeColor)</code></pre>
<pre><code>##  &#39;table&#39; num [1:4, 1:4, 1:2] 32 53 10 3 11 50 10 30 10 25 ...
##  - attr(*, &quot;dimnames&quot;)=List of 3
##   ..$ Hair: chr [1:4] &quot;Black&quot; &quot;Brown&quot; &quot;Red&quot; &quot;Blond&quot;
##   ..$ Eye : chr [1:4] &quot;Brown&quot; &quot;Blue&quot; &quot;Hazel&quot; &quot;Green&quot;
##   ..$ Sex : chr [1:2] &quot;Male&quot; &quot;Female&quot;</code></pre>
<pre class="r"><code>?HairEyeColor</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
setwd(&quot;D:/books/R/blogdown/hugo&quot;)
load(&quot;../data/Deuteranopia.RData&quot;)
Deuteranopia</code></pre>
<pre><code>##           Men Women
## Deute      19     2
## NonDeute 1981  1998</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
test &lt;- chisq.test(Deuteranopia)
test</code></pre>
<pre><code>## 
##  Pearson&#39;s Chi-squared test with Yates&#39; continuity correction
## 
## data:  Deuteranopia
## X-squared = 12.255, df = 1, p-value = 0.0004641</code></pre>
<pre class="r"><code>test$statistic # test statistic</code></pre>
<pre><code>## X-squared 
##  12.25481</code></pre>
<pre class="r"><code>test$p.value # p-value</code></pre>
<pre><code>## [1] 0.0004640594</code></pre>
<pre class="r"><code>test$observed</code></pre>
<pre><code>##           Men Women
## Deute      19     2
## NonDeute 1981  1998</code></pre>
<pre class="r"><code>test$expected</code></pre>
<pre><code>##             Men  Women
## Deute      10.5   10.5
## NonDeute 1989.5 1989.5</code></pre>
<pre class="r"><code>(19+2)/2</code></pre>
<pre><code>## [1] 10.5</code></pre>
<pre class="r"><code>(1981+1998)/2</code></pre>
<pre><code>## [1] 1989.5</code></pre>
<pre class="r"><code># Yates&#39; continuity correction is subtract 0.5 from the absolute value 
(19-10.5-0.5)^2/10.5+(abs(2-10.5)-0.5)^2/10.5+(abs(1981-1989.5)-0.5)^2/1989.5+(1998-1989.5-0.5)^2/1989.5</code></pre>
<pre><code>## [1] 12.25481</code></pre>
<pre class="r"><code>test$method</code></pre>
<pre><code>## [1] &quot;Pearson&#39;s Chi-squared test with Yates&#39; continuity correction&quot;</code></pre>
<pre class="r"><code>test$residuals</code></pre>
<pre><code>##                 Men      Women
## Deute     2.6231569 -2.6231569
## NonDeute -0.1905667  0.1905667</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;HardyWeinberg&quot;)</code></pre>
<pre><code>## Loading required package: mice</code></pre>
<pre><code>## 
## Attaching package: &#39;mice&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:IRanges&#39;:
## 
##     cbind, rbind</code></pre>
<pre><code>## The following objects are masked from &#39;package:S4Vectors&#39;:
## 
##     cbind, rbind</code></pre>
<pre><code>## The following objects are masked from &#39;package:BiocGenerics&#39;:
## 
##     cbind, rbind</code></pre>
<pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     filter</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     cbind, rbind</code></pre>
<pre><code>## Loading required package: Rsolnp</code></pre>
<pre><code>## Loading required package: nnet</code></pre>
<pre><code>## 
## Attaching package: &#39;HardyWeinberg&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:S4Vectors&#39;:
## 
##     fold</code></pre>
<pre class="r"><code>data(&quot;Mourant&quot;)
Mourant[214:216,]</code></pre>
<pre><code>##     Population    Country Total  MM  MN  NN
## 214    Oceania Micronesia   962 228 436 298
## 215    Oceania Micronesia   678  36 229 413
## 216    Oceania     Tahiti   580 188 296  96</code></pre>
<pre class="r"><code>nMM = Mourant$MM[216]
nMN = Mourant$MN[216]
nNN = Mourant$NN[216]
loglik = function(p, q = 1 - p) {
  2 * nMM * log(p) + nMN * log(2*p*q) + 2 * nNN * log(q)
}
xv = seq(0.01, 0.99, by = 0.01)
yv = loglik(xv)
plot(x = xv, y = yv, type = &quot;l&quot;, lwd = 2,
     xlab = &quot;p&quot;, ylab = &quot;log-likelihood&quot;)
imax = which.max(yv)
abline(v = xv[imax], h = yv[imax], lwd = 1.5, col = &quot;blue&quot;)
abline(h = yv[imax], lwd = 1.5, col = &quot;purple&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-107-1.svg" width="576" /></p>
<pre class="r"><code>xv[imax]</code></pre>
<pre><code>## [1] 0.58</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
phat  =  af(c(nMM, nMN, nNN))
phat</code></pre>
<pre><code>## [1] 0.5793103</code></pre>
<pre class="r"><code>pMM   =  phat^2
qhat  =  1 - phat

## -----------------------------------------------------------------------------
pHW = c(MM = phat^2, MN = 2*phat*qhat, NN = qhat^2)
sum(c(nMM, nMN, nNN)) * pHW</code></pre>
<pre><code>##       MM       MN       NN 
## 194.6483 282.7034 102.6483</code></pre>
<pre class="r"><code>Mourant[, c(&quot;MM&quot;, &quot;MN&quot;, &quot;NN&quot;)][c(1, 69, 128, 148, 192),]</code></pre>
<pre><code>##       MM   MN  NN
## 1   1743 3222 763
## 69   254  736 174
## 128  207  916 142
## 148  181  563  68
## 192   90  473  56</code></pre>
<pre class="r"><code>as.matrix(Mourant[, c(&quot;MM&quot;, &quot;MN&quot;, &quot;NN&quot;)])[c(1, 69, 128, 148, 192),]</code></pre>
<pre><code>##        MM   MN  NN
## [1,] 1743 3222 763
## [2,]  254  736 174
## [3,]  207  916 142
## [4,]  181  563  68
## [5,]   90  473  56</code></pre>
<pre class="r"><code>c(&quot;red&quot;, rep(&quot;purple&quot;, 4))</code></pre>
<pre><code>## [1] &quot;red&quot;    &quot;purple&quot; &quot;purple&quot; &quot;purple&quot; &quot;purple&quot;</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
par(mai = rep(0.1, 4))
pops = c(1, 69, 128, 148, 192)
genotypeFrequencies = as.matrix(Mourant[, c(&quot;MM&quot;, &quot;MN&quot;, &quot;NN&quot;)])
HWTernaryPlot(genotypeFrequencies[pops, ],
        markerlab = Mourant$Country[pops],
        alpha = 0.0001, curvecols = c(&quot;red&quot;, rep(&quot;purple&quot;, 4)),
        mcex = 0.75, vertex.cex = 1)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-113-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
HWTernaryPlot(genotypeFrequencies[pops, ],
              markerlab = Mourant$Country[pops],
              curvecols = c(&quot;red&quot;, rep(&quot;purple&quot;, 4)),
              alpha = 0.0001, mcex = 0.75, vertex.cex = 1)

HWTernaryPlot(genotypeFrequencies[-pops, ], 
              newframe = FALSE, alpha = 0.0001, cex = 0.5)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-114-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
newgf = round(genotypeFrequencies / 50)
HWTernaryPlot(newgf[pops, ],
              markerlab = Mourant$Country[pops],
              curvecols = c(&quot;red&quot;, rep(&quot;purple&quot;, 4)),
              alpha = 0.0001, mcex = 0.75, vertex.cex = 1)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-115-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;seqLogo&quot;)</code></pre>
<pre><code>## 
## Attaching package: &#39;seqLogo&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:mice&#39;:
## 
##     ic</code></pre>
<pre class="r"><code>setwd(&quot;D:/books/R/blogdown/hugo&quot;)
load(&quot;../data/kozak.RData&quot;)
kozak</code></pre>
<pre><code>##   [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9]
## A 0.33 0.25  0.4 0.15 0.20    1    0    0 0.05
## C 0.12 0.25  0.1 0.40 0.40    0    0    0 0.05
## G 0.33 0.25  0.4 0.20 0.25    0    0    1 0.90
## T 0.22 0.25  0.1 0.25 0.15    0    1    0 0.00</code></pre>
<pre class="r"><code>pwm = makePWM(kozak)
seqLogo(pwm, ic.scale = FALSE)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-116-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;markovchain&quot;)</code></pre>
<pre><code>## Package:  markovchain
## Version:  0.9.1
## Date:     2023-01-20
## BugReport: https://github.com/spedygiorgio/markovchain/issues</code></pre>
<pre class="r"><code>library(&quot;igraph&quot;)</code></pre>
<pre><code>## 
## Attaching package: &#39;igraph&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:rtracklayer&#39;:
## 
##     blocks, path</code></pre>
<pre><code>## The following object is masked from &#39;package:Biostrings&#39;:
## 
##     union</code></pre>
<pre><code>## The following object is masked from &#39;package:XVector&#39;:
## 
##     path</code></pre>
<pre><code>## The following object is masked from &#39;package:GenomicRanges&#39;:
## 
##     union</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     union</code></pre>
<pre><code>## The following object is masked from &#39;package:S4Vectors&#39;:
## 
##     union</code></pre>
<pre><code>## The following objects are masked from &#39;package:BiocGenerics&#39;:
## 
##     normalize, path, union</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     decompose, spectrum</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     union</code></pre>
<pre class="r"><code>sequence = toupper(c(&quot;a&quot;, &quot;c&quot;, &quot;a&quot;, &quot;c&quot;, &quot;g&quot;, &quot;t&quot;, &quot;t&quot;, &quot;t&quot;, &quot;t&quot;, &quot;c&quot;, &quot;c&quot;,
                     &quot;a&quot;, &quot;c&quot;, &quot;g&quot;, &quot;t&quot;, &quot;a&quot;, &quot;c&quot;,&quot;c&quot;,&quot;c&quot;,&quot;a&quot;,&quot;a&quot;,&quot;a&quot;,&quot;t&quot;,&quot;a&quot;,
                     &quot;c&quot;,&quot;g&quot;,&quot;g&quot;,&quot;c&quot;,&quot;a&quot;,&quot;t&quot;,&quot;g&quot;,&quot;t&quot;,&quot;g&quot;,&quot;t&quot;,&quot;g&quot;,&quot;a&quot;,&quot;g&quot;,&quot;c&quot;,&quot;t&quot;,&quot;g&quot;))
mcFit   =  markovchainFit(data = sequence)
mcFit</code></pre>
<pre><code>## $estimate
## MLE Fit 
##  A  4 - dimensional discrete Markov Chain defined by the following states: 
##  A, C, G, T 
##  The transition matrix  (by rows)  is defined as follows: 
##           A         C         G          T
## A 0.2000000 0.5000000 0.1000000 0.20000000
## C 0.3636364 0.2727273 0.2727273 0.09090909
## G 0.1250000 0.2500000 0.1250000 0.50000000
## T 0.2000000 0.1000000 0.4000000 0.30000000
## 
## 
## $standardError
##           A         C         G          T
## A 0.1414214 0.2236068 0.1000000 0.14142136
## C 0.1818182 0.1574592 0.1574592 0.09090909
## G 0.1250000 0.1767767 0.1250000 0.25000000
## T 0.1414214 0.1000000 0.2000000 0.17320508
## 
## $confidenceLevel
## [1] 0.95
## 
## $lowerEndpointMatrix
##             A          C           G         T
## A 0.000000000 0.06173864 0.000000000 0.0000000
## C 0.007279201 0.00000000 0.000000000 0.0000000
## G 0.000000000 0.00000000 0.000000000 0.0100089
## T 0.000000000 0.00000000 0.008007122 0.0000000
## 
## $upperEndpointMatrix
##           A         C         G         T
## A 0.4771808 0.9382614 0.2959964 0.4771808
## C 0.7199935 0.5813416 0.5813416 0.2690877
## G 0.3699955 0.5964760 0.3699955 0.9899911
## T 0.4771808 0.2959964 0.7919929 0.6394758
## 
## $logLikelihood
## [1] -48.94867</code></pre>
<pre class="r"><code>MCgraph =  markovchain:::.getNet(mcFit$estimate, round = TRUE)
MCgraph</code></pre>
<pre><code>## IGRAPH 7e22123 DNW- 4 16 -- 
## + attr: name (v/c), weight (e/n)
## + edges from 7e22123 (vertex names):
##  [1] A-&gt;A A-&gt;C A-&gt;G A-&gt;T C-&gt;A C-&gt;C C-&gt;G C-&gt;T G-&gt;A G-&gt;C G-&gt;G G-&gt;T T-&gt;A T-&gt;C T-&gt;G
## [16] T-&gt;T</code></pre>
<pre class="r"><code>E(MCgraph)</code></pre>
<pre><code>## + 16/16 edges from 7e22123 (vertex names):
##  [1] A-&gt;A A-&gt;C A-&gt;G A-&gt;T C-&gt;A C-&gt;C C-&gt;G C-&gt;T G-&gt;A G-&gt;C G-&gt;G G-&gt;T T-&gt;A T-&gt;C T-&gt;G
## [16] T-&gt;T</code></pre>
<pre class="r"><code>E(MCgraph)$weight</code></pre>
<pre><code>##  [1] 20.00 50.00 10.00 20.00 36.36 27.27 27.27  9.09 12.50 25.00 12.50 50.00
## [13] 20.00 10.00 40.00 30.00</code></pre>
<pre class="r"><code>edgelab =  round(E(MCgraph)$weight / 100, 2)
edgelab</code></pre>
<pre><code>##  [1] 0.20 0.50 0.10 0.20 0.36 0.27 0.27 0.09 0.12 0.25 0.12 0.50 0.20 0.10 0.40
## [16] 0.30</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
par(mai=c(0,0,0,0))
plot.igraph(MCgraph, edge.label = edgelab,
       vertex.size = 40, xlim = c(-1, 1.25))</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-121-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
par(mai=c(0,0,0,0))
plot.igraph(MCgraph, edge.label = edgelab,
       vertex.size = 40, xlim = c(-1, 1.25))</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-122-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
setwd(&quot;D:/books/R/blogdown/hugo&quot;)
haplo6 = read.table(&quot;../data/haplotype6.txt&quot;, header = TRUE)
haplo6</code></pre>
<pre><code>##   Individual DYS19 DXYS156Y DYS389m DYS389n DYS389p
## 1         H1    14       12       4      12       3
## 2         H3    15       13       4      13       3
## 3         H4    15       11       5      11       3
## 4         H5    17       13       4      11       3
## 5         H7    13       12       5      12       3
## 6         H8    16       11       5      12       3</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
with(haplo6, stopifnot(Individual[1] == &quot;H1&quot;, DYS19[1] == 14, DXYS156Y[1] == 12))</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
dfbetas = data.frame(
  p = rep(p_seq, 3),
  dbeta = c(dbeta(p_seq,  10,  30),
            dbeta(p_seq,  20,  60), 
            dbeta(p_seq,  50, 150)),
  pars = rep(c(&quot;Beta(10,30)&quot;, &quot;Beta(20,60)&quot;, &quot;Beta(50,150)&quot;), each = length(p_seq)))
library(&quot;ggplot2&quot;)</code></pre>
<pre><code>## 
## Attaching package: &#39;ggplot2&#39;</code></pre>
<pre><code>## The following object is masked _by_ &#39;.GlobalEnv&#39;:
## 
##     stat</code></pre>
<pre class="r"><code>ggplot(dfbetas) +
  geom_line(aes(x = p, y = dbeta, colour = pars)) +
  theme(legend.title = element_blank()) +
  geom_vline(aes(xintercept = 0.25), colour = &quot;#990000&quot;, linetype = &quot;dashed&quot;) + # mean
  geom_vline(aes(xintercept = (10-1)/(10+30-2)), colour = &quot;#888800&quot;) # mode</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-125-1.svg" width="576" /></p>
<pre class="r"><code>data &lt;- rbeta(100000, 50, 350)
summary(data)</code></pre>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.06693 0.11347 0.12434 0.12496 0.13572 0.21112</code></pre>
<pre class="r"><code>hist(data, breaks = 100, col = &quot;skyblue&quot;, main = &quot;&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-126-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
rp = rbeta(100000, 50, 350)
y = vapply(rp, 
           function(x) rbinom(1, prob = x, size = 300), 
           integer(1))
hist(y, breaks = 50, col = &quot;orange&quot;, main = &quot;&quot;, xlab = &quot;&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-127-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
set.seed(0xbebe)
y1 = vapply(rp, 
            function(x) rbinom(1, prob = x, size = 300), 
            integer(1))
set.seed(0xbebe)
y2 = rbinom(length(rp), rp, size = 300)
stopifnot(identical(y1, y2))
hist(y2, breaks = 50, col = &quot;orange&quot;, main = &quot;&quot;, xlab = &quot;&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-128-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
pPostEmp = rp[ y == 40 ]
hist(pPostEmp, breaks = 40, col = &quot;chartreuse4&quot;, main = &quot;&quot;,
  probability = TRUE, xlab = &quot;posterior p&quot;)

p_seq = seq(0, 1, by = 0.001)
densPostTheory = dbeta(p_seq, 50 + 40, 350 + 260)
lines(p_seq, densPostTheory, type = &quot;l&quot;, lwd = 3)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-129-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
mean(pPostEmp)</code></pre>
<pre><code>## [1] 0.1283359</code></pre>
<pre class="r"><code>dp = p_seq[2] - p_seq[1]
dp</code></pre>
<pre><code>## [1] 0.001</code></pre>
<pre class="r"><code>sum(p_seq * densPostTheory * dp)</code></pre>
<pre><code>## [1] 0.1285714</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
stopifnot(abs(mean(pPostEmp) - sum(p_seq * densPostTheory * dp)) &lt; 1e-3)</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
pPostMC = rbeta(n = 100000, 90, 610)
mean(pPostMC)</code></pre>
<pre><code>## [1] 0.128571</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
qqplot(pPostMC, pPostEmp, type = &quot;l&quot;, asp = 1)
abline(a = 0, b = 1, col = &quot;blue&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-133-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
densPost2 = dbeta(p_seq, 115, 735)
mcPost2   = rbeta(1e6, 115, 735)
sum(p_seq * densPost2 * dp)   # mean, by numeric integration</code></pre>
<pre><code>## [1] 0.1352941</code></pre>
<pre class="r"><code>mean(mcPost2)                 # mean by MC</code></pre>
<pre><code>## [1] 0.1352657</code></pre>
<pre class="r"><code>p_seq[which.max(densPost2)]   # MAP estimate</code></pre>
<pre><code>## [1] 0.134</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
quantile(mcPost2, c(0.025, 0.975))</code></pre>
<pre><code>##      2.5%     97.5% 
## 0.1131087 0.1590205</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;Biostrings&quot;)

## -----------------------------------------------------------------------------
## GENETIC_CODE
## IUPAC_CODE_MAP
## vignette(package = &quot;Biostrings&quot;)
## vignette(&quot;BiostringsQuickOverview&quot;, package = &quot;Biostrings&quot;)

## -----------------------------------------------------------------------------
GENETIC_CODE</code></pre>
<pre><code>## TTT TTC TTA TTG TCT TCC TCA TCG TAT TAC TAA TAG TGT TGC TGA TGG CTT CTC CTA CTG 
## &quot;F&quot; &quot;F&quot; &quot;L&quot; &quot;L&quot; &quot;S&quot; &quot;S&quot; &quot;S&quot; &quot;S&quot; &quot;Y&quot; &quot;Y&quot; &quot;*&quot; &quot;*&quot; &quot;C&quot; &quot;C&quot; &quot;*&quot; &quot;W&quot; &quot;L&quot; &quot;L&quot; &quot;L&quot; &quot;L&quot; 
## CCT CCC CCA CCG CAT CAC CAA CAG CGT CGC CGA CGG ATT ATC ATA ATG ACT ACC ACA ACG 
## &quot;P&quot; &quot;P&quot; &quot;P&quot; &quot;P&quot; &quot;H&quot; &quot;H&quot; &quot;Q&quot; &quot;Q&quot; &quot;R&quot; &quot;R&quot; &quot;R&quot; &quot;R&quot; &quot;I&quot; &quot;I&quot; &quot;I&quot; &quot;M&quot; &quot;T&quot; &quot;T&quot; &quot;T&quot; &quot;T&quot; 
## AAT AAC AAA AAG AGT AGC AGA AGG GTT GTC GTA GTG GCT GCC GCA GCG GAT GAC GAA GAG 
## &quot;N&quot; &quot;N&quot; &quot;K&quot; &quot;K&quot; &quot;S&quot; &quot;S&quot; &quot;R&quot; &quot;R&quot; &quot;V&quot; &quot;V&quot; &quot;V&quot; &quot;V&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;D&quot; &quot;D&quot; &quot;E&quot; &quot;E&quot; 
## GGT GGC GGA GGG 
## &quot;G&quot; &quot;G&quot; &quot;G&quot; &quot;G&quot; 
## attr(,&quot;alt_init_codons&quot;)
## [1] &quot;TTG&quot; &quot;CTG&quot;</code></pre>
<pre class="r"><code>IUPAC_CODE_MAP</code></pre>
<pre><code>##      A      C      G      T      M      R      W      S      Y      K      V 
##    &quot;A&quot;    &quot;C&quot;    &quot;G&quot;    &quot;T&quot;   &quot;AC&quot;   &quot;AG&quot;   &quot;AT&quot;   &quot;CG&quot;   &quot;CT&quot;   &quot;GT&quot;  &quot;ACG&quot; 
##      H      D      B      N 
##  &quot;ACT&quot;  &quot;AGT&quot;  &quot;CGT&quot; &quot;ACGT&quot;</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;BSgenome&quot;)
ag = available.genomes()
length(ag)</code></pre>
<pre><code>## [1] 112</code></pre>
<pre class="r"><code>ag[1:2]</code></pre>
<pre><code>## [1] &quot;BSgenome.Alyrata.JGI.v1&quot;              
## [2] &quot;BSgenome.Amellifera.BeeBase.assembly4&quot;</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;BSgenome.Ecoli.NCBI.20080805&quot;)
Ecoli</code></pre>
<pre><code>## E. coli genome:
## # organism: Escherichia coli (E. coli)
## # genome: 2008/08/05
## # provider: NCBI
## # release date: NA
## # 13 sequences:
## #   NC_008253 NC_008563 NC_010468 NC_004431 NC_009801 NC_009800 NC_002655
## #   NC_002695 NC_010498 NC_007946 NC_010473 NC_000913 AC_000091          
## # (use &#39;seqnames()&#39; to see all the sequence names, use the &#39;$&#39; or &#39;[[&#39; operator
## # to access a given sequence)</code></pre>
<pre class="r"><code>shineDalgarno = &quot;AGGAGGT&quot;
ecoli = Ecoli$NC_010473
ecoli</code></pre>
<pre><code>## 4686137-letter DNAString object
## seq: AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTC...ATCACCAAATAAAAAACGCCTTAGTAAGTATTTTTC</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
window = 50000
starts = seq(1, length(ecoli) - window, by = window)
length(ecoli) - window</code></pre>
<pre><code>## [1] 4636137</code></pre>
<pre class="r"><code>head(starts)</code></pre>
<pre><code>## [1]      1  50001 100001 150001 200001 250001</code></pre>
<pre class="r"><code>seq_along(starts)</code></pre>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
## [26] 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50
## [51] 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75
## [76] 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93</code></pre>
<pre class="r"><code>starts[93]</code></pre>
<pre><code>## [1] 4600001</code></pre>
<pre class="r"><code>ends   = starts + window - 1
head(ends)</code></pre>
<pre><code>## [1]  50000 100000 150000 200000 250000 300000</code></pre>
<pre class="r"><code>seq_along(ends)</code></pre>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
## [26] 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50
## [51] 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75
## [76] 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93</code></pre>
<pre class="r"><code>ends[93]</code></pre>
<pre><code>## [1] 4650000</code></pre>
<pre class="r"><code>numMatches = vapply(seq_along(starts), function(i) {
  countPattern(shineDalgarno, ecoli[starts[i]:ends[i]],
               max.mismatch = 0)
  }, numeric(1))
table(numMatches)</code></pre>
<pre><code>## numMatches
##  0  1  2  3  4 
## 48 32  8  3  2</code></pre>
<pre class="r"><code>mean(numMatches) # the maximum likelihood estimate of poisson lambda is the mean</code></pre>
<pre><code>## [1] 0.6989247</code></pre>
<pre class="r"><code>dpois(c(0, 1, 2, 3, 4), lambda = 0.6989247)*length(numMatches) # the prediction of poisson distribution</code></pre>
<pre><code>## [1] 46.2321199 32.3127706 11.2920967  2.6307751  0.4596784</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;vcd&quot;)
gf = goodfit(numMatches, &quot;poisson&quot;)
summary(gf)</code></pre>
<pre><code>## 
##   Goodness-of-fit test for poisson distribution
## 
##                       X^2 df  P(&gt; X^2)
## Likelihood Ratio 4.134932  3 0.2472577</code></pre>
<pre class="r"><code>gf$observed</code></pre>
<pre><code>## [1] 48 32  8  3  2</code></pre>
<pre class="r"><code>gf$count</code></pre>
<pre><code>## [1] 0 1 2 3 4</code></pre>
<pre class="r"><code>gf$df</code></pre>
<pre><code>## [1] 3</code></pre>
<pre class="r"><code>gf$method</code></pre>
<pre><code>## [1] &quot;ML&quot;</code></pre>
<pre class="r"><code>gf$fitted</code></pre>
<pre><code>## [1] 46.2321185 32.3127710 11.2920974  2.6307754  0.4596785</code></pre>
<pre class="r"><code>gf$par</code></pre>
<pre><code>## $lambda
## [1] 0.6989247</code></pre>
<pre class="r"><code>distplot(numMatches, type = &quot;poisson&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-151-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
sdMatches = matchPattern(shineDalgarno, ecoli, max.mismatch = 0)
head(sdMatches)</code></pre>
<pre><code>## Views on a 4686137-letter DNAString subject
## subject: AGCTTTTCATTCTGACTGCAACGGGCAATATGTC...CACCAAATAAAAAACGCCTTAGTAAGTATTTTTC
## views:
##        start    end width
##   [1]  56593  56599     7 [AGGAGGT]
##   [2] 199644 199650     7 [AGGAGGT]
##   [3] 202176 202182     7 [AGGAGGT]
##   [4] 214433 214439     7 [AGGAGGT]
##   [5] 217429 217435     7 [AGGAGGT]
##   [6] 227331 227337     7 [AGGAGGT]</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
betweenmotifs = gaps(sdMatches)
head(betweenmotifs)</code></pre>
<pre><code>## Views on a 4686137-letter DNAString subject
## subject: AGCTTTTCATTCTGACTGCAACGGGCAATATGTC...CACCAAATAAAAAACGCCTTAGTAAGTATTTTTC
## views:
##        start    end  width
##   [1]      1  56592  56592 [AGCTTTTCATTCTGACTGCAACGG...TCCAGGTGTCAGAACCCGGCAGAC]
##   [2]  56600 199643 143044 [AAAGCTACCGTTATCCAGAATGGT...TGGGAGAGCGCCTGCTTTGCACGC]
##   [3] 199651 202175   2525 [CTGCGGTTCGATCCCGCATAGCTC...GTTGGCTAATCCTGGTCGGACATC]
##   [4] 202183 214432  12250 [TAGTGCAATGGCATAAGCCAGCTT...ATTATCGTGTTATCGCCAGGCTTT]
##   [5] 214440 217428   2989 [TAATAACATGGGCAGGATAAGCTC...CATAACGAAAAGCCCCTTACTTGT]
##   [6] 217436 227330   9895 [CTGACCACTTGTGATGATATGGTT...CGTTTTCTGATTCCACAAACTGCA]</code></pre>
<pre class="r"><code>length(betweenmotifs)</code></pre>
<pre><code>## [1] 66</code></pre>
<pre class="r"><code>head(width(betweenmotifs))</code></pre>
<pre><code>## [1]  56592 143044   2525  12250   2989   9895</code></pre>
<pre class="r"><code>mean(width(betweenmotifs))</code></pre>
<pre><code>## [1] 70995.18</code></pre>
<pre class="r"><code>table(width(betweenmotifs))</code></pre>
<pre><code>## 
##    431    921   2525   2526   2989   4731   5029   5509   6066   9373   9895 
##      1      1      1      1      1      1      1      1      1      1      1 
##  10156  10395  11659  11825  12250  18141  18402  21195  22018  23136  23923 
##      1      1      1      1      1      1      1      1      1      1      1 
##  27523  27604  27756  27949  29936  38566  39295  41394  41401  44557  46875 
##      1      1      1      1      1      1      1      1      1      1      1 
##  47744  51681  53830  56080  56592  59234  59292  62938  63513  64798  69573 
##      1      1      1      1      1      1      1      1      1      1      1 
##  74505  79547  86205  86243  87327  90295  90941  94206  97259  97557 113253 
##      1      1      1      1      1      1      1      1      1      1      1 
## 119872 134895 143044 153621 179327 189903 241430 251600 262848 311563 429015 
##      1      1      1      1      1      1      1      1      1      1      1</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;Renext&quot;)</code></pre>
<pre><code>## Loading required package: evd</code></pre>
<pre><code>## 
## Attaching package: &#39;evd&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:igraph&#39;:
## 
##     clusters</code></pre>
<pre class="r"><code>expplot(width(betweenmotifs), rate = 1/mean(width(betweenmotifs)),
        labels = &quot;fit&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-155-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
## gofExp.test(width(betweenmotifs))

## -----------------------------------------------------------------------------
library(&quot;BSgenome.Hsapiens.UCSC.hg19&quot;)
setwd(&quot;D:/books/R/blogdown/hugo&quot;)
chr8  =  Hsapiens$chr8
CpGtab = read.table(&quot;../data/model-based-cpg-islands-hg19.txt&quot;,
                    header = TRUE)
nrow(CpGtab)</code></pre>
<pre><code>## [1] 65699</code></pre>
<pre class="r"><code>head(CpGtab)</code></pre>
<pre><code>##     chr  start    end length CpGcount GCcontent pctGC obsExp
## 1 chr10  93098  93818    721       32       403 0.559  0.572
## 2 chr10  94002  94165    164       12        97 0.591  0.841
## 3 chr10  94527  95302    776       65       538 0.693  0.702
## 4 chr10 119652 120193    542       53       369 0.681  0.866
## 5 chr10 122133 122621    489       51       339 0.693  0.880
## 6 chr10 180265 180720    456       32       256 0.561  0.893</code></pre>
<pre class="r"><code>irCpG = with(dplyr::filter(CpGtab, chr == &quot;chr8&quot;),
         IRanges(start = start, end = end))
head(irCpG)</code></pre>
<pre><code>## IRanges object with 6 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]     32437     33708      1272
##   [2]     40354     40656       303
##   [3]     44536     46203      1668
##   [4]     99457    100721      1265
##   [5]    155954    156761       808
##   [6]    179033    179756       724</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
grCpG = GRanges(ranges = irCpG, seqnames = &quot;chr8&quot;, strand = &quot;+&quot;)
genome(grCpG) = &quot;hg19&quot;
head(grCpG)</code></pre>
<pre><code>## GRanges object with 6 ranges and 0 metadata columns:
##       seqnames        ranges strand
##          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr8   32437-33708      +
##   [2]     chr8   40354-40656      +
##   [3]     chr8   44536-46203      +
##   [4]     chr8  99457-100721      +
##   [5]     chr8 155954-156761      +
##   [6]     chr8 179033-179756      +
##   -------
##   seqinfo: 1 sequence from hg19 genome; no seqlengths</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;Gviz&quot;)
ideo = IdeogramTrack(genome = &quot;hg19&quot;, chromosome = &quot;chr8&quot;)
plotTracks(
  list(GenomeAxisTrack(),
    AnnotationTrack(grCpG, name = &quot;CpG&quot;), ideo),
    from = 2200000, to = 5800000,
    shape = &quot;box&quot;, fill = &quot;#006400&quot;, stacking = &quot;dense&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-160-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
CGIview    = Views(unmasked(Hsapiens$chr8), irCpG)
head(CGIview)</code></pre>
<pre><code>## Views on a 146364022-letter DNAString subject
## subject: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
## views:
##        start    end width
##   [1]  32437  33708  1272 [CGGGTGCCATCTCAGCAGCTCACGG...AGAACACGGCGGGGGGGGCGGCGC]
##   [2]  40354  40656   303 [CGATTAGCGGGAGCGCAAAGCGAAG...GCGGTCGGCTCGGCGAGGAGCCGG]
##   [3]  44536  46203  1668 [CGGCGCTGCAGGAGAGGAGATGCCC...GAGTTCATGGGTGTGACGGGGCGT]
##   [4]  99457 100721  1265 [CGGCTGGCCGGGCAGGGGGGCTGAC...ATCCTCCCGGCGGTCGGGCGGCGG]
##   [5] 155954 156761   808 [CGGGAAAGATTTTATTCACCGTCGA...CGCAGTATACTGGCGGCACGCCGC]
##   [6] 179033 179756   724 [CGTGTTCCGGGGGTTATATGAGTGT...CCCTAGAGCTCAAGGCACTGTCGG]</code></pre>
<pre class="r"><code>NonCGIview = Views(unmasked(Hsapiens$chr8), gaps(irCpG))
head(NonCGIview)</code></pre>
<pre><code>## Views on a 146364022-letter DNAString subject
## subject: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
## views:
##        start    end width
##   [1]  33709  40353  6645 [TCTTCTCATCACGTGCTCTCAGGGG...CGCGGAGCAGCAGGCACTCATTCC]
##   [2]  40657  44535  3879 [TCTCCAAGTGCCGCCAGCTGCGGGA...CGCGGGTTCTCTGTGGCCAGCAGG]
##   [3]  46204  99456 53253 [GTGCTGTGTGAGAACATGTGTGTAG...GCGCCCCTCACCTCCCGGACGGGG]
##   [4] 100722 155953 55232 [CGGCTGCCCTAAGTGATCTTTTTAA...ACCTTGATTATTAACGGCTGCAAC]
##   [5] 156762 179032 22271 [CTGCTGGCAGCTAGGGACATTGCAG...TATGTACGCTCTGAGAGATACATG]
##   [6] 179757 181776  2020 [AAGCTGAGCGCCCTCTGCTACCCCT...CATCCACGCCTGTCCCAAAGGTAC]</code></pre>
<pre class="r"><code>head(DNAStringSet(CGIview))</code></pre>
<pre><code>## DNAStringSet object of length 6:
##     width seq
## [1]  1272 CGGGTGCCATCTCAGCAGCTCACGGTGTGGAAAC...GTTTGCGGAAGAACACGGCGGGGGGGGCGGCGC
## [2]   303 CGATTAGCGGGAGCGCAAAGCGAAGGGGCGGCCT...CGCGCACGCGCGGTCGGCTCGGCGAGGAGCCGG
## [3]  1668 CGGCGCTGCAGGAGAGGAGATGCCCAGGCCAGGC...TCTCGGTGTGAGTTCATGGGTGTGACGGGGCGT
## [4]  1265 CGGCTGGCCGGGCAGGGGGGCTGACCCCCCCCAC...CTGAACTCCATCCTCCCGGCGGTCGGGCGGCGG
## [5]   808 CGGGAAAGATTTTATTCACCGTCGATGCGGCCCC...TCTCTTGCTCGCAGTATACTGGCGGCACGCCGC
## [6]   724 CGTGTTCCGGGGGTTATATGAGTGTGACGGGTGT...ACTGCAGCGCCCTAGAGCTCAAGGCACTGTCGG</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
seqCGI      = as(CGIview, &quot;DNAStringSet&quot;)
head(seqCGI)</code></pre>
<pre><code>## DNAStringSet object of length 6:
##     width seq
## [1]  1272 CGGGTGCCATCTCAGCAGCTCACGGTGTGGAAAC...GTTTGCGGAAGAACACGGCGGGGGGGGCGGCGC
## [2]   303 CGATTAGCGGGAGCGCAAAGCGAAGGGGCGGCCT...CGCGCACGCGCGGTCGGCTCGGCGAGGAGCCGG
## [3]  1668 CGGCGCTGCAGGAGAGGAGATGCCCAGGCCAGGC...TCTCGGTGTGAGTTCATGGGTGTGACGGGGCGT
## [4]  1265 CGGCTGGCCGGGCAGGGGGGCTGACCCCCCCCAC...CTGAACTCCATCCTCCCGGCGGTCGGGCGGCGG
## [5]   808 CGGGAAAGATTTTATTCACCGTCGATGCGGCCCC...TCTCTTGCTCGCAGTATACTGGCGGCACGCCGC
## [6]   724 CGTGTTCCGGGGGTTATATGAGTGTGACGGGTGT...ACTGCAGCGCCCTAGAGCTCAAGGCACTGTCGG</code></pre>
<pre class="r"><code>length(seqCGI)</code></pre>
<pre><code>## [1] 2855</code></pre>
<pre class="r"><code>seqNonCGI   = as(NonCGIview, &quot;DNAStringSet&quot;)
head(seqNonCGI)</code></pre>
<pre><code>## DNAStringSet object of length 6:
##     width seq
## [1]  6645 TCTTCTCATCACGTGCTCTCAGGGGCCACGATGT...CCGCAAGAACGCGGAGCAGCAGGCACTCATTCC
## [2]  3879 TCTCCAAGTGCCGCCAGCTGCGGGATTTCCTCTG...GCCGGCGCACGCGGGTTCTCTGTGGCCAGCAGG
## [3] 53253 GTGCTGTGTGAGAACATGTGTGTAGTGTTCACAT...CGGGCAGAGGCGCCCCTCACCTCCCGGACGGGG
## [4] 55232 CGGCTGCCCTAAGTGATCTTTTTAAGTAAAGGAG...TGGTCTCTGACCTTGATTATTAACGGCTGCAAC
## [5] 22271 CTGCTGGCAGCTAGGGACATTGCAGGGCCCTCTT...CCCGGGCGGTATGTACGCTCTGAGAGATACATG
## [6]  2020 AAGCTGAGCGCCCTCTGCTACCCCTCCTGCTGCA...ATGGGATGTCATCCACGCCTGTCCCAAAGGTAC</code></pre>
<pre class="r"><code>length(seqNonCGI)</code></pre>
<pre><code>## [1] 2854</code></pre>
<pre class="r"><code>dinucCpG    = sapply(seqCGI, dinucleotideFrequency)
dim(dinucCpG)</code></pre>
<pre><code>## [1]   16 2855</code></pre>
<pre class="r"><code>dinucCpG[,1]</code></pre>
<pre><code>##  AA  AC  AG  AT  CA  CC  CG  CT  GA  GC  GG  GT  TA  TC  TG  TT 
##  49 127  55   8 132  44 116 112  54 129 107 100   4 104 112  18</code></pre>
<pre class="r"><code>dinucNonCpG = sapply(seqNonCGI, dinucleotideFrequency)
dim(dinucNonCpG)</code></pre>
<pre><code>## [1]   16 2854</code></pre>
<pre class="r"><code>dinucNonCpG[, 1]</code></pre>
<pre><code>##  AA  AC  AG  AT  CA  CC  CG  CT  GA  GC  GG  GT  TA  TC  TG  TT 
## 389 351 400 436 498 560 112 603 359 336 403 336 330 527 519 485</code></pre>
<pre class="r"><code>NonICounts = rowSums(dinucNonCpG)
NonICounts</code></pre>
<pre><code>##       AA       AC       AG       AT       CA       CC       CG       CT 
## 14223322  7129981  9795572 11291315 10241505  6931732  1174927  9779692 
##       GA       GC       GG       GT       TA       TC       TG       TT 
##  8372468  5706958  6934755  7112531  9602902  8359398 10221420 14197986</code></pre>
<pre class="r"><code>IslCounts  = rowSums(dinucCpG)
IslCounts</code></pre>
<pre><code>##     AA     AC     AG     AT     CA     CC     CG     CT     GA     GC     GG 
##  64103  83109 122131  44000 113346 193847 133549 122377 105186 177326 194517 
##     GT     TA     TC     TG     TT 
##  86752  31210 106738 114592  65861</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
TI  = matrix(IslCounts, ncol = 4, byrow = TRUE)
TnI = matrix(NonICounts, ncol = 4, byrow = TRUE)
dimnames(TI) = dimnames(TnI) =
  list(c(&quot;A&quot;, &quot;C&quot;, &quot;G&quot;, &quot;T&quot;), c(&quot;A&quot;, &quot;C&quot;, &quot;G&quot;, &quot;T&quot;))
head(TI)</code></pre>
<pre><code>##        A      C      G      T
## A  64103  83109 122131  44000
## C 113346 193847 133549 122377
## G 105186 177326 194517  86752
## T  31210 106738 114592  65861</code></pre>
<pre class="r"><code>head(TnI)</code></pre>
<pre><code>##          A       C        G        T
## A 14223322 7129981  9795572 11291315
## C 10241505 6931732  1174927  9779692
## G  8372468 5706958  6934755  7112531
## T  9602902 8359398 10221420 14197986</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
MI = TI /rowSums(TI)
MI</code></pre>
<pre><code>##            A         C         G         T
## A 0.20457773 0.2652333 0.3897678 0.1404212
## C 0.20128250 0.3442381 0.2371595 0.2173200
## G 0.18657245 0.3145299 0.3450223 0.1538754
## T 0.09802105 0.3352314 0.3598984 0.2068492</code></pre>
<pre class="r"><code>MN = TnI / rowSums(TnI)
MN</code></pre>
<pre><code>##           A         C          G         T
## A 0.3351380 0.1680007 0.23080886 0.2660524
## C 0.3641054 0.2464366 0.04177094 0.3476871
## G 0.2976696 0.2029017 0.24655406 0.2528746
## T 0.2265813 0.1972407 0.24117528 0.3350027</code></pre>
<p>The transitions are different. For instance, the transitions from C to A and T to A for in the islands (MI) transition matrix seem very different (0.201 versus 0.098).</p>
<pre class="r"><code>alphabetFrequency(seqCGI, baseOnly = TRUE, collapse = TRUE)</code></pre>
<pre><code>##      A      C      G      T  other 
## 313845 563875 564789 318990      0</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
freqIsl = alphabetFrequency(seqCGI, baseOnly = TRUE, collapse = TRUE)[1:4]
freqIsl</code></pre>
<pre><code>##      A      C      G      T 
## 313845 563875 564789 318990</code></pre>
<pre class="r"><code>freqIsl / sum(freqIsl)</code></pre>
<pre><code>##         A         C         G         T 
## 0.1781693 0.3201109 0.3206298 0.1810901</code></pre>
<pre class="r"><code>freqNon = alphabetFrequency(seqNonCGI, baseOnly = TRUE, collapse = TRUE)[1:4]
freqNon</code></pre>
<pre><code>##        A        C        G        T 
## 42440774 28128852 28127513 42382186</code></pre>
<pre class="r"><code>freqNon / sum(freqNon)</code></pre>
<pre><code>##         A         C         G         T 
## 0.3008292 0.1993832 0.1993737 0.3004139</code></pre>
<p>This shows an inverse pattern: in the CpG islands, C and G have frequencies around 0.32, whereas in the non-CpG islands, we have A and T that have frequencies around 0.30.</p>
<pre class="r"><code>## -----------------------------------------------------------------------------
alpha = log((freqIsl/sum(freqIsl)) / (freqNon/sum(freqNon)))
alpha</code></pre>
<pre><code>##          A          C          G          T 
## -0.5238084  0.4734387  0.4751059 -0.5061665</code></pre>
<pre class="r"><code>beta  = log(MI / MN)
beta</code></pre>
<pre><code>##            A         C         G          T
## A -0.4935945 0.4566418 0.5239611 -0.6390469
## C -0.5927341 0.3342289 1.7365319 -0.4699321
## G -0.4671646 0.4383576 0.3360277 -0.4967508
## T -0.8379216 0.5303960 0.4002977 -0.4821485</code></pre>
<pre class="r"><code>s = unlist(strsplit(&quot;ACGTTATACTACG&quot;, &quot;&quot;))
s</code></pre>
<pre><code>##  [1] &quot;A&quot; &quot;C&quot; &quot;G&quot; &quot;T&quot; &quot;T&quot; &quot;A&quot; &quot;T&quot; &quot;A&quot; &quot;C&quot; &quot;T&quot; &quot;A&quot; &quot;C&quot; &quot;G&quot;</code></pre>
<pre class="r"><code>beta[s[2-1], s[2]]</code></pre>
<pre><code>## [1] 0.4566418</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
x = &quot;ACGTTATACTACG&quot;
scorefun = function(x) {
  s = unlist(strsplit(x, &quot;&quot;))
  score = alpha[s[1]]
  if (length(s) &gt;= 2)
    for (j in 2:length(s))
      score = score + beta[s[j-1], s[j]]
  score
}
scorefun(x)</code></pre>
<pre><code>##          A 
## -0.2824623</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
generateRandomScores = function(s, len = 100, B = 1000) {
  alphFreq = alphabetFrequency(s)
  isGoodSeq = rowSums(alphFreq[, 5:ncol(alphFreq)]) == 0
  s = s[isGoodSeq]
  slen = sapply(s, length)
  prob = pmax(slen - len, 0)
  prob = prob / sum(prob)
  idx  = sample(length(s), B, replace = TRUE, prob = prob)
  ssmp = s[idx]
  start = sapply(ssmp, function(x) sample(length(x) - len, 1))
  scores = sapply(seq_len(B), function(i)
    scorefun(as.character(ssmp[[i]][start[i]+(1:len)]))
  )
  scores / len
}
scoresCGI    = generateRandomScores(seqCGI)
scoresNonCGI = generateRandomScores(seqNonCGI)</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
rgs = range(c(scoresCGI, scoresNonCGI))
rgs</code></pre>
<pre><code>## [1] -0.6191153  0.6016392</code></pre>
<pre class="r"><code>br = seq(rgs[1], rgs[2], length.out = 50)
h1 = hist(scoresCGI,    breaks = br, plot = FALSE)
h2 = hist(scoresNonCGI, breaks = br, plot = FALSE)
plot(h1, col = rgb(0, 0, 1, 1/4), xlim = c(-0.5, 0.5), ylim=c(0,120))
plot(h2, col = rgb(1, 0, 0, 1/4), add = TRUE)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-178-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
## ###This is for provenance reasons, keep track of how the data
## ###were generated for the EM exercise in Chapter 4.
## Mdata=c(scoresCGI,scoresNonCGI)
## MM1=sample(Mdata[1:1000],800)
## MM2=sample(Mdata[1001:2000],1000)
## Myst=c(MM1,MM2);names(Myst)=NULL
## saveRDS(c(MM1,MM2),&quot;../data/Myst.rds&quot;)
## ###True value of m1,m2,s1 and s2
## ###

## -----------------------------------------------------------------------------
stopifnot(max(h1$counts) &lt; 120, max(h2$counts) &lt; 120,
          h1$breaks[1] &gt;= br[1], h1$breaks[length(h1$breaks)] &lt;= br[length(br)],
          h2$breaks[1] &gt;= br[1], h2$breaks[length(h2$breaks)] &lt;= br[length(br)])</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
setwd(&quot;D:/books/R/blogdown/hugo&quot;)
mtb = read.table(&quot;../data/M_tuberculosis.txt&quot;, header = TRUE)
head(mtb, n = 4)</code></pre>
<pre><code>##   AmAcid Codon Number PerThous
## 1    Gly   GGG  25874    19.25
## 2    Gly   GGA  13306     9.90
## 3    Gly   GGT  25320    18.84
## 4    Gly   GGC  68310    50.82</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
Gly  =  mtb[ mtb$AmAcid == &quot;Gly&quot;, &quot;Number&quot;]
Gly</code></pre>
<pre><code>## [1] 25874 13306 25320 68310</code></pre>
<pre class="r"><code>Gly/sum(Gly)</code></pre>
<pre><code>## [1] 0.1948197 0.1001882 0.1906483 0.5143438</code></pre>
<pre class="r"><code>dim(mtb)</code></pre>
<pre><code>## [1] 64  4</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
pro  =  mtb[ mtb$AmAcid == &quot;Pro&quot;, &quot;Number&quot;]
pro</code></pre>
<pre><code>## [1] 42424  8229  4578 22895</code></pre>
<pre class="r"><code>pro/sum(pro)</code></pre>
<pre><code>## [1] 0.54302025 0.10532985 0.05859765 0.29305225</code></pre>
<pre class="r"><code>table(mtb$AmAcid)</code></pre>
<pre><code>## 
## Ala Arg Asn Asp Cys End Gln Glu Gly His Ile Leu Lys Met Phe Pro Ser Thr Trp Tyr 
##   4   6   2   2   2   3   2   2   4   2   3   6   2   1   2   4   6   4   1   2 
## Val 
##   4</code></pre>
<pre class="r"><code>table(mtb$Codon)</code></pre>
<pre><code>## 
## AAA AAC AAG AAT ACA ACC ACG ACT AGA AGC AGG AGT ATA ATC ATG ATT CAA CAC CAG CAT 
##   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 
## CCA CCC CCG CCT CGA CGC CGG CGT CTA CTC CTG CTT GAA GAC GAG GAT GCA GCC GCG GCT 
##   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 
## GGA GGC GGG GGT GTA GTC GTG GTT TAA TAC TAG TAT TCA TCC TCG TCT TGA TGC TGG TGT 
##   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 
## TTA TTC TTG TTT 
##   1   1   1   1</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
setwd(&quot;D:/books/R/blogdown/hugo&quot;)
staph = readDNAStringSet(&quot;../data/staphsequence.ffn.txt&quot;, &quot;fasta&quot;)

## -----------------------------------------------------------------------------
staph[1:3, ]</code></pre>
<pre><code>## DNAStringSet object of length 3:
##     width seq                                               names               
## [1]  1362 ATGTCGGAAAAAGAAATTTGGGA...AAAAAGAAATAAGAAATGTATAA lcl|NC_002952.2_c...
## [2]  1134 ATGATGGAATTCACTATTAAAAG...TTTTACCAATCAGAACTTACTAA lcl|NC_002952.2_c...
## [3]   246 GTGATTATTTTGGTTCAAGAAGT...TCATTCATCAAGGTGAACAATGA lcl|NC_002952.2_c...</code></pre>
<pre class="r"><code>staph</code></pre>
<pre><code>## DNAStringSet object of length 2650:
##        width seq                                            names               
##    [1]  1362 ATGTCGGAAAAAGAAATTTGGG...AAAGAAATAAGAAATGTATAA lcl|NC_002952.2_c...
##    [2]  1134 ATGATGGAATTCACTATTAAAA...TTACCAATCAGAACTTACTAA lcl|NC_002952.2_c...
##    [3]   246 GTGATTATTTTGGTTCAAGAAG...ATTCATCAAGGTGAACAATGA lcl|NC_002952.2_c...
##    [4]  1113 ATGAAGTTAAATACACTCCAAT...CAAGGTGAAATTATAAAGTAA lcl|NC_002952.2_c...
##    [5]  1932 GTGACTGCATTGTCAGATGTAA...TATGCAAACTTAGACTTCTAA lcl|NC_002952.2_c...
##    ...   ... ...
## [2646]   720 ATGACTGTAGAATGGTTAGCAG...ACTCCTTTACTTGAAAAATAA lcl|NC_002952.2_c...
## [2647]  1878 GTGGTTCAAGAATATGATGTAA...CTCCAAAGGGTGAGTGACTAA lcl|NC_002952.2_c...
## [2648]  1380 ATGGATTTAGATACAATTACGA...CAATTCTGCTTAGGTAAATAG lcl|NC_002952.2_c...
## [2649]   348 TTGGAAAAAGCTTACCGAATTA...TTTAATAAAAAGATTAAGTAA lcl|NC_002952.2_c...
## [2650]   138 ATGGTAAAACGTACTTATCAAC...CGTAAAGTTTTATCTGCATAA lcl|NC_002952.2_c...</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
letterFrequency(staph[[1]], letters = &quot;ACGT&quot;, OR = 0)</code></pre>
<pre><code>##   A   C   G   T 
## 522 219 229 392</code></pre>
<pre class="r"><code>GCstaph = data.frame(
  ID = names(staph),
  GC = rowSums(alphabetFrequency(staph)[, 2:3] / width(staph)) * 100
)</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
window = 100
gc = rowSums( letterFrequencyInSlidingView(staph[[364]], window,
      c(&quot;G&quot;,&quot;C&quot;)))/window
plot(x = seq(along = gc), y = gc, type = &quot;l&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-188-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
plot(x = seq(along = gc), y = gc, type = &quot;l&quot;)
lines(lowess(x = seq(along = gc), y = gc, f = 0.2), col = 2)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-189-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
dfbetas = data.frame(
  p = rep(p_seq, 5),
  dbeta = c(dbeta(p_seq, 0.5, 0.5), 
            dbeta(p_seq,   1,   1), 
            dbeta(p_seq,  10,  30),
            dbeta(p_seq,  20,  60), 
            dbeta(p_seq,  50, 150)),
  pars = rep(c(&quot;Beta(0.5,0.5)&quot;, &quot;U(0,1)=Beta(1,1)&quot;, 
               &quot;Beta(10,30)&quot;, &quot;Beta(20,60)&quot;, 
               &quot;Beta(50,150)&quot;), each = length(p_seq)))
ggplot(dfbetas) +
  geom_line(aes(x = p, y = dbeta, colour = pars)) +
  theme(legend.title = element_blank()) +
  geom_vline(aes(xintercept = 0.25), colour = &quot;#990000&quot;, linetype = &quot;dashed&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-190-1.svg" width="576" /></p>
</div>
<div id="high-quality-graphics-in-r" class="section level2">
<h2>3 High Quality Graphics in R</h2>
<pre class="r"><code>## -----------------------------------------------------------------------------
## ## produces the xkcd-plotting image used below
library(&quot;xkcd&quot;)</code></pre>
<pre><code>## Loading required package: extrafont</code></pre>
<pre><code>## Registering fonts with R</code></pre>
<pre class="r"><code>library(&quot;showtext&quot;)</code></pre>
<pre><code>## Loading required package: sysfonts</code></pre>
<pre><code>## Loading required package: showtextdb</code></pre>
<pre><code>## 
## Attaching package: &#39;showtextdb&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:extrafont&#39;:
## 
##     font_install</code></pre>
<pre class="r"><code>library(&quot;sysfonts&quot;)
library(&quot;tibble&quot;)</code></pre>
<pre><code>## 
## Attaching package: &#39;tibble&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:igraph&#39;:
## 
##     as_data_frame</code></pre>
<pre class="r"><code>## 
introplotdata = tibble(
   y = c(seq(-8, 1, length=25)^2, rep(1, 5), seq(1, 5,length=25)^2)^2,
   x = seq(1, 55, length.out = length(y)))

 dataman = tibble(
   x = 30,
   y = 400,
   scale = 100,
   ratioxy = 0.1,
   angleofspine =  -pi/2 ,
   anglerighthumerus = -pi/6,
   anglelefthumerus  = pi * 7/6,
   anglerightradius = 0,
   angleleftradius = 0,
   angleleftleg  = 19*pi/12,
   anglerightleg = 17*pi/12,
   angleofneck   = 1.4*pi)
 
 mapping = do.call(aes_string, colnames(dataman) %&gt;% (function(x) setNames(as.list(x), x)))</code></pre>
<pre><code>## Warning: `aes_string()` was deprecated in ggplot2 3.0.0.
## ℹ Please use tidy evaluation idioms with `aes()`.
## ℹ See also `vignette(&quot;ggplot2-in-packages&quot;)` for more information.
## This warning is displayed once every 8 hours.
## Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
## generated.</code></pre>
<pre class="r"><code>ggplot(introplotdata) + geom_line(aes(x = x, y = y), size = 2) +
    xkcdaxis(c(0, 50), c(0, 1000)) + xlab(&quot;Time to make plot in minutes&quot;) +
    ylab(&quot;Time to understand plot in minutes&quot;) + xkcdman(mapping, dataman) +
    theme(axis.title.x = element_text(margin = margin(15, 0, 0, 0)))</code></pre>
<pre><code>## Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
## ℹ Please use `linewidth` instead.
## This warning is displayed once every 8 hours.
## Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
## generated.</code></pre>
<pre><code>## Warning in theme_xkcd(): Not xkcd fonts installed! See vignette(&quot;xkcd-intro&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-191-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
head(DNase)</code></pre>
<pre><code>##   Run       conc density
## 1   1 0.04882812   0.017
## 2   1 0.04882812   0.018
## 3   1 0.19531250   0.121
## 4   1 0.19531250   0.124
## 5   1 0.39062500   0.206
## 6   1 0.39062500   0.215</code></pre>
<pre class="r"><code>plot(DNase$conc, DNase$density)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-193-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
plot(DNase$conc, DNase$density,
  ylab = attr(DNase, &quot;labels&quot;)$y,
  xlab = paste(attr(DNase, &quot;labels&quot;)$x, attr(DNase, &quot;units&quot;)$x),
  pch = 3,
  col = &quot;blue&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-194-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
hist(DNase$density, breaks=25, main = &quot;&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-195-1.svg" width="576" /></p>
<pre class="r"><code>boxplot(density ~ Run, data = DNase)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-195-2.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;Hiiragi2013&quot;)</code></pre>
<pre><code>## Loading required package: affy</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>## Loading required package: boot</code></pre>
<pre><code>## Loading required package: clue</code></pre>
<pre><code>## Loading required package: cluster</code></pre>
<pre><code>## Loading required package: genefilter</code></pre>
<pre><code>## Loading required package: geneplotter</code></pre>
<pre><code>## Loading required package: lattice</code></pre>
<pre><code>## 
## Attaching package: &#39;lattice&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:boot&#39;:
## 
##     melanoma</code></pre>
<pre><code>## The following object is masked from &#39;package:evd&#39;:
## 
##     qq</code></pre>
<pre><code>## Loading required package: annotate</code></pre>
<pre><code>## Loading required package: AnnotationDbi</code></pre>
<pre><code>## Loading required package: XML</code></pre>
<pre><code>## 
## Attaching package: &#39;geneplotter&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:Gviz&#39;:
## 
##     imageMap</code></pre>
<pre><code>## Loading required package: gplots</code></pre>
<pre><code>## 
## Attaching package: &#39;gplots&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:rtracklayer&#39;:
## 
##     space</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     space</code></pre>
<pre><code>## The following object is masked from &#39;package:S4Vectors&#39;:
## 
##     space</code></pre>
<pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     lowess</code></pre>
<pre><code>## Loading required package: gtools</code></pre>
<pre><code>## 
## Attaching package: &#39;gtools&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:boot&#39;:
## 
##     inv.logit, logit</code></pre>
<pre><code>## The following object is masked from &#39;package:igraph&#39;:
## 
##     permute</code></pre>
<pre><code>## Loading required package: KEGGREST</code></pre>
<pre><code>## Loading required package: MASS</code></pre>
<pre><code>## 
## Attaching package: &#39;MASS&#39;</code></pre>
<pre><code>## The following object is masked _by_ &#39;.GlobalEnv&#39;:
## 
##     genotype</code></pre>
<pre><code>## The following object is masked from &#39;package:AnnotationDbi&#39;:
## 
##     select</code></pre>
<pre><code>## The following object is masked from &#39;package:genefilter&#39;:
## 
##     area</code></pre>
<pre><code>## Loading required package: mouse4302.db</code></pre>
<pre><code>## Loading required package: org.Mm.eg.db</code></pre>
<pre><code>## </code></pre>
<pre><code>## </code></pre>
<pre><code>## Loading required package: RColorBrewer</code></pre>
<pre><code>## Loading required package: xtable</code></pre>
<pre class="r"><code>data(&quot;x&quot;)
dim(Biobase::exprs(x))</code></pre>
<pre><code>## [1] 45101   101</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
head(pData(x), n = 2)</code></pre>
<pre><code>##         File.name Embryonic.day Total.number.of.cells lineage genotype
## 1 E3.25  1_C32_IN         E3.25                    32               WT
## 2 E3.25  2_C32_IN         E3.25                    32               WT
##           ScanDate sampleGroup sampleColour
## 1 E3.25 2011-03-16       E3.25      #CAB2D6
## 2 E3.25 2011-03-16       E3.25      #CAB2D6</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;dplyr&quot;)</code></pre>
<pre><code>## 
## Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:MASS&#39;:
## 
##     select</code></pre>
<pre><code>## The following object is masked from &#39;package:AnnotationDbi&#39;:
## 
##     select</code></pre>
<pre><code>## The following object is masked from &#39;package:Biobase&#39;:
## 
##     combine</code></pre>
<pre><code>## The following objects are masked from &#39;package:igraph&#39;:
## 
##     as_data_frame, groups, union</code></pre>
<pre><code>## The following object is masked from &#39;package:HardyWeinberg&#39;:
## 
##     recode</code></pre>
<pre><code>## The following objects are masked from &#39;package:Biostrings&#39;:
## 
##     collapse, intersect, setdiff, setequal, union</code></pre>
<pre><code>## The following object is masked from &#39;package:XVector&#39;:
## 
##     slice</code></pre>
<pre><code>## The following objects are masked from &#39;package:GenomicRanges&#39;:
## 
##     intersect, setdiff, union</code></pre>
<pre><code>## The following object is masked from &#39;package:GenomeInfoDb&#39;:
## 
##     intersect</code></pre>
<pre><code>## The following objects are masked from &#39;package:IRanges&#39;:
## 
##     collapse, desc, intersect, setdiff, slice, union</code></pre>
<pre><code>## The following objects are masked from &#39;package:S4Vectors&#39;:
## 
##     first, intersect, rename, setdiff, setequal, union</code></pre>
<pre><code>## The following objects are masked from &#39;package:BiocGenerics&#39;:
## 
##     combine, intersect, setdiff, union</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre class="r"><code>groups = group_by(pData(x), sampleGroup) %&gt;%
  summarise(n = n(), color = unique(sampleColour))
groups</code></pre>
<pre><code>## # A tibble: 8 × 3
##   sampleGroup         n color  
##   &lt;chr&gt;           &lt;int&gt; &lt;chr&gt;  
## 1 E3.25              36 #CAB2D6
## 2 E3.25 (FGF4-KO)    17 #FDBF6F
## 3 E3.5 (EPI)         11 #A6CEE3
## 4 E3.5 (FGF4-KO)      8 #FF7F00
## 5 E3.5 (PE)          11 #B2DF8A
## 6 E4.5 (EPI)          4 #1F78B4
## 7 E4.5 (FGF4-KO)     10 #E31A1C
## 8 E4.5 (PE)           4 #33A02C</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
## f(x) %&gt;% g(y) %&gt;% h
## h(g(f(x), y))

## -----------------------------------------------------------------------------
library(&quot;ggplot2&quot;)
ggplot(DNase, aes(x = conc, y = density)) + geom_point()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-199-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
ggplot(groups, aes(x = sampleGroup, y = n)) +
  geom_bar(stat = &quot;identity&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-200-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
## check an assertion made in the text above
stopifnot(formals(ggplot2::geom_bar)$stat==&quot;count&quot;)</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
groupColor = setNames(groups$color, groups$sampleGroup)

## -----------------------------------------------------------------------------
ggplot(groups, aes(x = sampleGroup, y = n, fill = sampleGroup)) +
  geom_bar(stat = &quot;identity&quot;) +
  scale_fill_manual(values = groupColor, name = &quot;Groups&quot;) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-202-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
gg = ggplot(DNase, aes(x = conc, y = density)) + geom_point()

## -----------------------------------------------------------------------------
## gg
print(gg)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-203-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
ggplot2::ggsave(&quot;DNAse-histogram-demo.pdf&quot;, plot = gg)</code></pre>
<pre><code>## Saving 6 x 5 in image</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
file.remove(&quot;DNAse-histogram-demo.pdf&quot;)</code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;mouse4302.db&quot;)</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
## # I used this code to find the below two probes
idx = order(rowVars(Biobase::exprs(x)), decreasing=TRUE)[seq_len(2000)]
cc  = cor(t(Biobase::exprs(x)[idx,]))
cco = order(cc)[seq(1, 1001, by=2) ]
jj2 = rownames(Biobase::exprs(x))[ idx[ (cco-1) %/% length(idx) + 1 ] ]
jj1 = rownames(Biobase::exprs(x))[ idx[ (cco-1) %%  length(idx) + 1 ] ]
dftx = as.data.frame(t(Biobase::exprs(x)))
#par(ask=TRUE)
for(i in seq(along = cco)) {
  df = AnnotationDbi::select(mouse4302.db,
                             keys = c(jj1[i], jj2[i]), keytype = &quot;PROBEID&quot;,
    columns = c(&quot;SYMBOL&quot;, &quot;GENENAME&quot;))
   print(ggplot(dftx, aes( x = get(jj1[i]), y = get(jj2[i]))) +
   geom_point(shape = 1) +
   xlab(paste(jj1[i], df$SYMBOL[1])) +
   ylab(paste(jj2[i], df$SYMBOL[2])) +
   ggtitle(round(cc[jj1[i], jj2[i]], 3)) + geom_smooth(method = &quot;loess&quot;))
}</code></pre>
<pre><code>## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-1.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-2.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-3.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-4.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-5.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-6.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-7.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-8.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-9.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-10.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-11.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-12.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-13.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-14.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-15.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-16.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-17.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-18.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-19.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-20.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-21.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-22.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-23.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-24.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-25.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-26.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-27.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-28.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-29.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-30.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-31.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-32.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-33.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-34.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-35.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-36.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-37.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-38.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-39.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-40.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-41.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-42.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-43.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-44.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-45.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-46.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-47.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-48.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-49.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-50.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-51.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-52.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-53.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-54.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-55.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-56.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-57.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-58.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-59.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-60.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-61.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-62.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-63.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-64.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-65.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-66.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-67.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-68.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-69.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-70.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-71.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-72.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-73.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-74.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-75.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-76.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-77.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-78.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-79.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-80.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-81.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-82.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-83.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-84.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-85.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-86.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-87.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-88.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-89.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-90.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-91.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-92.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-93.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-94.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-95.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-96.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-97.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-98.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-99.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-100.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-101.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-102.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-103.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-104.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-105.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-106.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-107.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-108.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-109.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-110.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-111.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-112.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-113.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-114.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-115.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-116.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-117.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-118.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-119.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-120.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-121.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-122.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-123.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-124.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-125.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-126.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-127.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-128.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-129.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-130.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-131.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-132.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-133.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-134.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-135.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-136.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-137.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-138.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-139.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-140.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-141.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-142.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-143.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-144.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-145.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-146.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-147.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-148.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-149.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-150.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-151.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-152.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-153.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-154.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-155.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-156.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-157.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-158.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-159.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-160.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-161.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-162.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-163.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-164.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-165.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-166.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-167.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-168.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-169.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-170.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-171.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-172.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-173.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-174.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-175.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-176.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-177.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-178.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-179.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-180.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-181.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-182.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-183.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-184.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-185.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-186.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-187.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-188.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-189.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-190.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-191.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-192.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-193.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-194.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-195.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-196.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-197.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-198.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-199.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-200.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-201.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-202.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-203.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-204.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-205.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-206.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-207.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-208.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-209.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-210.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-211.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-212.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-213.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-214.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-215.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-216.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-217.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-218.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-219.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-220.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-221.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-222.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-223.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-224.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-225.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-226.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-227.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-228.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-229.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-230.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-231.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-232.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-233.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-234.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-235.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-236.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-237.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-238.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-239.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-240.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-241.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-242.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-243.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-244.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-245.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-246.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-247.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-248.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-249.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-250.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-251.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-252.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-253.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-254.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-255.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-256.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-257.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-258.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-259.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-260.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-261.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-262.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-263.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-264.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-265.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-266.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-267.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-268.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-269.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-270.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-271.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-272.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-273.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-274.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-275.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-276.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-277.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-278.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-279.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-280.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-281.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-282.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-283.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-284.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-285.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-286.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-287.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-288.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-289.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-290.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-291.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-292.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-293.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-294.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-295.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-296.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-297.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-298.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-299.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-300.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-301.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-302.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-303.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-304.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-305.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-306.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-307.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-308.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-309.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-310.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-311.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-312.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-313.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-314.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-315.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-316.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-317.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-318.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-319.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-320.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-321.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-322.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-323.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-324.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-325.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-326.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-327.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-328.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-329.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-330.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-331.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-332.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-333.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-334.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-335.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-336.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-337.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-338.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-339.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-340.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-341.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-342.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-343.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-344.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-345.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-346.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-347.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-348.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-349.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-350.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-351.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-352.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-353.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-354.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-355.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-356.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-357.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-358.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-359.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-360.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-361.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-362.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-363.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-364.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-365.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-366.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-367.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-368.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-369.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-370.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-371.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-372.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-373.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-374.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-375.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-376.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-377.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-378.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-379.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-380.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-381.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-382.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-383.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-384.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-385.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-386.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-387.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-388.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-389.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-390.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-391.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-392.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-393.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-394.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-395.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-396.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-397.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-398.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-399.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-400.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-401.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-402.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-403.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-404.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-405.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-406.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-407.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-408.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-409.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-410.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-411.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-412.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-413.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-414.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-415.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-416.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-417.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-418.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-419.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-420.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-421.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-422.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-423.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-424.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-425.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-426.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-427.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-428.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-429.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-430.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-431.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-432.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-433.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-434.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-435.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-436.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-437.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-438.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-439.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-440.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-441.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-442.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-443.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-444.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-445.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-446.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-447.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-448.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-449.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-450.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-451.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-452.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-453.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-454.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-455.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-456.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-457.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-458.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-459.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-460.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-461.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-462.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-463.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-464.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-465.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-466.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-467.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-468.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-469.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-470.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-471.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-472.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-473.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-474.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-475.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-476.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-477.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-478.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-479.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-480.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-481.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-482.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-483.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-484.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-485.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-486.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-487.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-488.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-489.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-490.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-491.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-492.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-493.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-494.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-495.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-496.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-497.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-498.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-499.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;
## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-500.svg" width="576" /></p>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-206-501.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
dftx = data.frame(t(Biobase::exprs(x)), pData(x))
ggplot( dftx, aes( x = X1426642_at, y = X1418765_at )) +
  geom_point( shape = 1 ) +
  geom_smooth( method = &quot;loess&quot; )</code></pre>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-207-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
stopifnot(is(dftx, &quot;data.frame&quot;))</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
.one &lt;- AnnotationDbi::select(mouse4302.db, 
                              keys = &quot;1418765_at&quot;, 
                              keytype = &quot;PROBEID&quot;, 
                              columns = &quot;SYMBOL&quot;)$SYMBOL</code></pre>
<pre><code>## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<pre class="r"><code>.two &lt;- AnnotationDbi::select(mouse4302.db, 
                              keys = &quot;1426642_at&quot;, 
                              keytype = &quot;PROBEID&quot;, 
                              columns = &quot;SYMBOL&quot;)$SYMBOL</code></pre>
<pre><code>## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
ggplot( dftx, aes( x = X1426642_at, y = X1418765_at ))  +
  geom_point( aes( color = sampleColour), shape = 19 ) +
  geom_smooth( method = &quot;loess&quot; ) +
  scale_color_discrete( guide = &quot;none&quot; )</code></pre>
<pre><code>## `geom_smooth()` using formula = &#39;y ~ x&#39;</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-209-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;mouse4302.db&quot;)



## -----------------------------------------------------------------------------
AnnotationDbi::select(mouse4302.db,
   keys = c(&quot;1426642_at&quot;, &quot;1418765_at&quot;), keytype = &quot;PROBEID&quot;,
   columns = c(&quot;SYMBOL&quot;, &quot;GENENAME&quot;))</code></pre>
<pre><code>## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<pre><code>##      PROBEID SYMBOL                                            GENENAME
## 1 1426642_at    Fn1                                       fibronectin 1
## 2 1418765_at  Timd2 T cell immunoglobulin and mucin domain containing 2</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
dfx = as.data.frame(Biobase::exprs(x))
ggplot(dfx, aes(x = `20 E3.25`)) + geom_histogram(binwidth = 0.2)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-211-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
pb = ggplot(groups, aes(x = sampleGroup, y = n))

## -----------------------------------------------------------------------------
class(pb)</code></pre>
<pre><code>## [1] &quot;gg&quot;     &quot;ggplot&quot;</code></pre>
<pre class="r"><code>pb</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-213-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
pb = pb + geom_bar(stat = &quot;identity&quot;)
pb = pb + aes(fill = sampleGroup)
pb = pb + theme(axis.text.x = element_text(angle = 90, hjust = 1))
pb = pb + scale_fill_manual(values = groupColor, name = &quot;Groups&quot;)
pb</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-214-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
pb.polar = pb + coord_polar() +
  theme(axis.text.x = element_text(angle = 0, hjust = 1),
        axis.text.y = element_blank(),
        axis.ticks = element_blank()) +
  xlab(&quot;&quot;) + ylab(&quot;&quot;)
pb.polar</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-215-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
selectedProbes = c( Fgf4 = &quot;1420085_at&quot;, Gata4 = &quot;1418863_at&quot;,
                   Gata6 = &quot;1425463_at&quot;,  Sox2 = &quot;1416967_at&quot;)

## -----------------------------------------------------------------------------
## # How I found the selectedProbes:
## AnnotationDbi::select(mouse4302.db,
##    keys = c(&quot;Fgf4&quot;, &quot;Sox2&quot;, &quot;Gata6&quot;, &quot;Gata4&quot;), keytype = &quot;SYMBOL&quot;,
##    columns = c(&quot;PROBEID&quot;))

## -----------------------------------------------------------------------------
selectedProbes2 = AnnotationDbi::select(mouse4302.db,
   keys = selectedProbes, keytype = &quot;PROBEID&quot;, columns = c(&quot;SYMBOL&quot;))</code></pre>
<pre><code>## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<pre class="r"><code>stopifnot(identical(sort(selectedProbes2$SYMBOL), sort(names(selectedProbes))),
          all(selectedProbes[selectedProbes2$SYMBOL] == selectedProbes2$PROBEID))</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;reshape2&quot;)
genes = melt(Biobase::exprs(x)[selectedProbes, ],
             varnames = c(&quot;probe&quot;, &quot;sample&quot;))

## -----------------------------------------------------------------------------
genes$gene =
  names(selectedProbes)[match(genes$probe, selectedProbes)]
head(genes)</code></pre>
<pre><code>##        probe  sample    value  gene
## 1 1420085_at 1 E3.25 3.027715  Fgf4
## 2 1418863_at 1 E3.25 4.843137 Gata4
## 3 1425463_at 1 E3.25 5.500618 Gata6
## 4 1416967_at 1 E3.25 1.731217  Sox2
## 5 1420085_at 2 E3.25 9.293016  Fgf4
## 6 1418863_at 2 E3.25 5.530016 Gata4</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
ggplot(genes, aes(x = gene, y = value)) +
  stat_summary(fun = mean, geom = &quot;bar&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-218-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;Hmisc&quot;)</code></pre>
<pre><code>## Loading required package: survival</code></pre>
<pre><code>## 
## Attaching package: &#39;survival&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:boot&#39;:
## 
##     aml</code></pre>
<pre><code>## Loading required package: Formula</code></pre>
<pre><code>## 
## Attaching package: &#39;Hmisc&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:dplyr&#39;:
## 
##     src, summarize</code></pre>
<pre><code>## The following objects are masked from &#39;package:xtable&#39;:
## 
##     label, label&lt;-</code></pre>
<pre><code>## The following object is masked from &#39;package:AnnotationDbi&#39;:
## 
##     contents</code></pre>
<pre><code>## The following object is masked from &#39;package:Biobase&#39;:
## 
##     contents</code></pre>
<pre><code>## The following objects are masked from &#39;package:Biostrings&#39;:
## 
##     mask, translate</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     format.pval, units</code></pre>
<pre class="r"><code>ggplot(genes, aes( x = gene, y = value, fill = gene)) +
  stat_summary(fun = mean, geom = &quot;bar&quot;) +
  stat_summary(fun.data = mean_cl_normal, geom = &quot;errorbar&quot;,
               width = 0.25)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-219-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
p = ggplot(genes, aes( x = gene, y = value, fill = gene))
p + geom_boxplot()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-220-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
p + geom_dotplot(binaxis = &quot;y&quot;, binwidth = 1/6,
       stackdir = &quot;center&quot;, stackratio = 0.75,
       aes(color = gene))</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-221-1.svg" width="576" /></p>
<pre class="r"><code>library(&quot;ggbeeswarm&quot;)
p + geom_beeswarm(aes(color = gene))</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-222-1.svg" width="576" /></p>
<pre class="r"><code>## ------------</code></pre>
<pre class="r"><code>#-----------------------------------------------------------------
ggplot(genes, aes( x = value, color = gene)) + geom_density()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-223-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
p + geom_violin()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-224-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;ggridges&quot;)
ggplot(genes, aes(x = value, y = gene, fill = gene)) + 
  geom_density_ridges()</code></pre>
<pre><code>## Picking joint bandwidth of 0.729</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-225-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
top42 = order(rowMeans(Biobase::exprs(x)), decreasing = TRUE)[1:42]
g42 = melt(Biobase::exprs(x)[rev(top42), ], varnames = c(&quot;probe&quot;, &quot;sample&quot;))
ggplot(g42, aes(x = value, y = probe)) </code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-226-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
ggplot(g42, aes(x = value, y = probe)) + 
  geom_density_ridges() + theme(legend.position = &quot;none&quot;,
    axis.title.y = element_blank(), axis.text.y = element_blank(),
    axis.ticks.y = element_blank()) + xlim(13, 15) </code></pre>
<pre><code>## Picking joint bandwidth of 0.0683</code></pre>
<pre><code>## Warning: Removed 35 rows containing non-finite values (`stat_density_ridges()`).</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-227-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
simdata = rnorm(70)
tibble(index = seq(along = simdata),
          sx = sort(simdata)) %&gt;%
ggplot(aes(x = sx, y = index)) + geom_step()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-228-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
ggplot(genes, aes( x = value, color = gene)) + stat_ecdf()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-229-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
## # I used the functon bimodality_coefficient from the modes package to identify the most
## # bimodal looking array, number 64
## j0 = which.max(vapply(seq_len(ncol(x)), function(j){
##        modes  ::    bimodality_coefficient(Biobase::exprs(x)[, j])
##     }, numeric(1)))

## -----------------------------------------------------------------------------
ggplot(dfx, aes(x = `64 E4.5 (EPI)`)) + geom_histogram(bins = 100)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-230-1.svg" width="576" /></p>
<pre class="r"><code>ggplot(dfx, aes(x = 2 ^ `64 E4.5 (EPI)`)) + 
  geom_histogram(binwidth = 20) + xlim(0, 1500)</code></pre>
<pre><code>## Warning: Removed 1457 rows containing non-finite values (`stat_bin()`).</code></pre>
<pre><code>## Warning: Removed 2 rows containing missing values (`geom_bar()`).</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-231-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
scp = ggplot(dfx, aes(x = `59 E4.5 (PE)` ,
                      y = `92 E4.5 (FGF4-KO)`))
scp + geom_point()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-232-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
scp  + geom_point(alpha = 0.1)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-233-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
scp + geom_density2d()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-234-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
scp + geom_density2d(h = 0.5, bins = 60)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-235-1.svg" width="576" /></p>
<pre class="r"><code>library(&quot;RColorBrewer&quot;)
colorscale = scale_fill_gradientn(
    colors = rev(brewer.pal(9, &quot;YlGnBu&quot;)),
    values = c(0, exp(seq(-5, 0, length.out = 100))))

scp + stat_density2d(h = 0.5, bins = 60,
          aes( fill = after_stat(level)), geom = &quot;polygon&quot;) +
  colorscale + coord_fixed()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-236-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
scp + geom_hex() + coord_fixed()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-237-1.svg" width="576" /></p>
<pre class="r"><code>scp + geom_hex(binwidth = c(0.2, 0.2)) + colorscale +
  coord_fixed()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-238-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;ggthemes&quot;)
sunsp = tibble(year   = time(sunspot.year),
               number = as.numeric(sunspot.year))
sp = ggplot(sunsp, aes(x = year, y = number)) + geom_line()
sp</code></pre>
<pre><code>## Don&#39;t know how to automatically pick scale for object of type &lt;ts&gt;. Defaulting
## to continuous.</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-239-1.svg" width="576" /></p>
<pre class="r"><code>ratio = with(sunsp, bank_slopes(year, number))
sp + coord_fixed(ratio = ratio)</code></pre>
<pre><code>## Don&#39;t know how to automatically pick scale for object of type &lt;ts&gt;. Defaulting
## to continuous.</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-240-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;magrittr&quot;)
dftx$lineage %&lt;&gt;% sub(&quot;^$&quot;, &quot;no&quot;, .)
dftx$lineage %&lt;&gt;% factor(levels = c(&quot;no&quot;, &quot;EPI&quot;, &quot;PE&quot;, &quot;FGF4-KO&quot;))

ggplot(dftx, aes(x = X1426642_at, y = X1418765_at)) +
  geom_point() + facet_grid( . ~ lineage )</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-241-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
ggplot(dftx,
  aes(x = X1426642_at, y = X1418765_at)) + geom_point() +
   facet_grid( Embryonic.day ~ lineage )</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-242-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
ggplot(mutate(dftx, Tdgf1 = cut(X1450989_at, breaks = 4)),
   aes(x = X1426642_at, y = X1418765_at)) + geom_point() +
   facet_wrap( ~ Tdgf1, ncol = 2 )</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-243-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;plotly&quot;)</code></pre>
<pre><code>## 
## Attaching package: &#39;plotly&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:Hmisc&#39;:
## 
##     subplot</code></pre>
<pre><code>## The following object is masked from &#39;package:MASS&#39;:
## 
##     select</code></pre>
<pre><code>## The following object is masked from &#39;package:AnnotationDbi&#39;:
## 
##     select</code></pre>
<pre><code>## The following object is masked from &#39;package:ggplot2&#39;:
## 
##     last_plot</code></pre>
<pre><code>## The following object is masked from &#39;package:igraph&#39;:
## 
##     groups</code></pre>
<pre><code>## The following object is masked from &#39;package:BSgenome&#39;:
## 
##     export</code></pre>
<pre><code>## The following object is masked from &#39;package:rtracklayer&#39;:
## 
##     export</code></pre>
<pre><code>## The following object is masked from &#39;package:XVector&#39;:
## 
##     slice</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     slice</code></pre>
<pre><code>## The following object is masked from &#39;package:S4Vectors&#39;:
## 
##     rename</code></pre>
<pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     filter</code></pre>
<pre><code>## The following object is masked from &#39;package:graphics&#39;:
## 
##     layout</code></pre>
<pre class="r"><code>plot_ly(economics, x = ~ date, y = ~ unemploy / pop)</code></pre>
<pre><code>## No trace type specified:
##   Based on info supplied, a &#39;scatter&#39; trace seems appropriate.
##   Read more about this trace type -&gt; https://plotly.com/r/reference/#scatter</code></pre>
<pre><code>## No scatter mode specifed:
##   Setting the mode to markers
##   Read more about this attribute -&gt; https://plotly.com/r/reference/#scatter-mode</code></pre>
<div class="plotly html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-1" style="width:576px;height:480px;"></div>
<script type="application/json" data-for="htmlwidget-1">{"x":{"visdat":{"3bc862db77ad":["function () ","plotlyVisDat"]},"cur_data":"3bc862db77ad","attrs":{"3bc862db77ad":{"x":{},"y":{},"alpha_stroke":1,"sizes":[10,100],"spans":[1,20]}},"layout":{"margin":{"b":40,"l":60,"t":25,"r":10},"xaxis":{"domain":[0,1],"automargin":true,"title":"date"},"yaxis":{"domain":[0,1],"automargin":true,"title":"unemploy/pop"},"hovermode":"closest","showlegend":false},"source":"A","config":{"modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"data":[{"x":["1967-07-01","1967-08-01","1967-09-01","1967-10-01","1967-11-01","1967-12-01","1968-01-01","1968-02-01","1968-03-01","1968-04-01","1968-05-01","1968-06-01","1968-07-01","1968-08-01","1968-09-01","1968-10-01","1968-11-01","1968-12-01","1969-01-01","1969-02-01","1969-03-01","1969-04-01","1969-05-01","1969-06-01","1969-07-01","1969-08-01","1969-09-01","1969-10-01","1969-11-01","1969-12-01","1970-01-01","1970-02-01","1970-03-01","1970-04-01","1970-05-01","1970-06-01","1970-07-01","1970-08-01","1970-09-01","1970-10-01","1970-11-01","1970-12-01","1971-01-01","1971-02-01","1971-03-01","1971-04-01","1971-05-01","1971-06-01","1971-07-01","1971-08-01","1971-09-01","1971-10-01","1971-11-01","1971-12-01","1972-01-01","1972-02-01","1972-03-01","1972-04-01","1972-05-01","1972-06-01","1972-07-01","1972-08-01","1972-09-01","1972-10-01","1972-11-01","1972-12-01","1973-01-01","1973-02-01","1973-03-01","1973-04-01","1973-05-01","1973-06-01","1973-07-01","1973-08-01","1973-09-01","1973-10-01","1973-11-01","1973-12-01","1974-01-01","1974-02-01","1974-03-01","1974-04-01","1974-05-01","1974-06-01","1974-07-01","1974-08-01","1974-09-01","1974-10-01","1974-11-01","1974-12-01","1975-01-01","1975-02-01","1975-03-01","1975-04-01","1975-05-01","1975-06-01","1975-07-01","1975-08-01","1975-09-01","1975-10-01","1975-11-01","1975-12-01","1976-01-01","1976-02-01","1976-03-01","1976-04-01","1976-05-01","1976-06-01","1976-07-01","1976-08-01","1976-09-01","1976-10-01","1976-11-01","1976-12-01","1977-01-01","1977-02-01","1977-03-01","1977-04-01","1977-05-01","1977-06-01","1977-07-01","1977-08-01","1977-09-01","1977-10-01","1977-11-01","1977-12-01","1978-01-01","1978-02-01","1978-03-01","1978-04-01","1978-05-01","1978-06-01","1978-07-01","1978-08-01","1978-09-01","1978-10-01","1978-11-01","1978-12-01","1979-01-01","1979-02-01","1979-03-01","1979-04-01","1979-05-01","1979-06-01","1979-07-01","1979-08-01","1979-09-01","1979-10-01","1979-11-01","1979-12-01","1980-01-01","1980-02-01","1980-03-01","1980-04-01","1980-05-01","1980-06-01","1980-07-01","1980-08-01","1980-09-01","1980-10-01","1980-11-01","1980-12-01","1981-01-01","1981-02-01","1981-03-01","1981-04-01","1981-05-01","1981-06-01","1981-07-01","1981-08-01","1981-09-01","1981-10-01","1981-11-01","1981-12-01","1982-01-01","1982-02-01","1982-03-01","1982-04-01","1982-05-01","1982-06-01","1982-07-01","1982-08-01","1982-09-01","1982-10-01","1982-11-01","1982-12-01","1983-01-01","1983-02-01","1983-03-01","1983-04-01","1983-05-01","1983-06-01","1983-07-01","1983-08-01","1983-09-01","1983-10-01","1983-11-01","1983-12-01","1984-01-01","1984-02-01","1984-03-01","1984-04-01","1984-05-01","1984-06-01","1984-07-01","1984-08-01","1984-09-01","1984-10-01","1984-11-01","1984-12-01","1985-01-01","1985-02-01","1985-03-01","1985-04-01","1985-05-01","1985-06-01","1985-07-01","1985-08-01","1985-09-01","1985-10-01","1985-11-01","1985-12-01","1986-01-01","1986-02-01","1986-03-01","1986-04-01","1986-05-01","1986-06-01","1986-07-01","1986-08-01","1986-09-01","1986-10-01","1986-11-01","1986-12-01","1987-01-01","1987-02-01","1987-03-01","1987-04-01","1987-05-01","1987-06-01","1987-07-01","1987-08-01","1987-09-01","1987-10-01","1987-11-01","1987-12-01","1988-01-01","1988-02-01","1988-03-01","1988-04-01","1988-05-01","1988-06-01","1988-07-01","1988-08-01","1988-09-01","1988-10-01","1988-11-01","1988-12-01","1989-01-01","1989-02-01","1989-03-01","1989-04-01","1989-05-01","1989-06-01","1989-07-01","1989-08-01","1989-09-01","1989-10-01","1989-11-01","1989-12-01","1990-01-01","1990-02-01","1990-03-01","1990-04-01","1990-05-01","1990-06-01","1990-07-01","1990-08-01","1990-09-01","1990-10-01","1990-11-01","1990-12-01","1991-01-01","1991-02-01","1991-03-01","1991-04-01","1991-05-01","1991-06-01","1991-07-01","1991-08-01","1991-09-01","1991-10-01","1991-11-01","1991-12-01","1992-01-01","1992-02-01","1992-03-01","1992-04-01","1992-05-01","1992-06-01","1992-07-01","1992-08-01","1992-09-01","1992-10-01","1992-11-01","1992-12-01","1993-01-01","1993-02-01","1993-03-01","1993-04-01","1993-05-01","1993-06-01","1993-07-01","1993-08-01","1993-09-01","1993-10-01","1993-11-01","1993-12-01","1994-01-01","1994-02-01","1994-03-01","1994-04-01","1994-05-01","1994-06-01","1994-07-01","1994-08-01","1994-09-01","1994-10-01","1994-11-01","1994-12-01","1995-01-01","1995-02-01","1995-03-01","1995-04-01","1995-05-01","1995-06-01","1995-07-01","1995-08-01","1995-09-01","1995-10-01","1995-11-01","1995-12-01","1996-01-01","1996-02-01","1996-03-01","1996-04-01","1996-05-01","1996-06-01","1996-07-01","1996-08-01","1996-09-01","1996-10-01","1996-11-01","1996-12-01","1997-01-01","1997-02-01","1997-03-01","1997-04-01","1997-05-01","1997-06-01","1997-07-01","1997-08-01","1997-09-01","1997-10-01","1997-11-01","1997-12-01","1998-01-01","1998-02-01","1998-03-01","1998-04-01","1998-05-01","1998-06-01","1998-07-01","1998-08-01","1998-09-01","1998-10-01","1998-11-01","1998-12-01","1999-01-01","1999-02-01","1999-03-01","1999-04-01","1999-05-01","1999-06-01","1999-07-01","1999-08-01","1999-09-01","1999-10-01","1999-11-01","1999-12-01","2000-01-01","2000-02-01","2000-03-01","2000-04-01","2000-05-01","2000-06-01","2000-07-01","2000-08-01","2000-09-01","2000-10-01","2000-11-01","2000-12-01","2001-01-01","2001-02-01","2001-03-01","2001-04-01","2001-05-01","2001-06-01","2001-07-01","2001-08-01","2001-09-01","2001-10-01","2001-11-01","2001-12-01","2002-01-01","2002-02-01","2002-03-01","2002-04-01","2002-05-01","2002-06-01","2002-07-01","2002-08-01","2002-09-01","2002-10-01","2002-11-01","2002-12-01","2003-01-01","2003-02-01","2003-03-01","2003-04-01","2003-05-01","2003-06-01","2003-07-01","2003-08-01","2003-09-01","2003-10-01","2003-11-01","2003-12-01","2004-01-01","2004-02-01","2004-03-01","2004-04-01","2004-05-01","2004-06-01","2004-07-01","2004-08-01","2004-09-01","2004-10-01","2004-11-01","2004-12-01","2005-01-01","2005-02-01","2005-03-01","2005-04-01","2005-05-01","2005-06-01","2005-07-01","2005-08-01","2005-09-01","2005-10-01","2005-11-01","2005-12-01","2006-01-01","2006-02-01","2006-03-01","2006-04-01","2006-05-01","2006-06-01","2006-07-01","2006-08-01","2006-09-01","2006-10-01","2006-11-01","2006-12-01","2007-01-01","2007-02-01","2007-03-01","2007-04-01","2007-05-01","2007-06-01","2007-07-01","2007-08-01","2007-09-01","2007-10-01","2007-11-01","2007-12-01","2008-01-01","2008-02-01","2008-03-01","2008-04-01","2008-05-01","2008-06-01","2008-07-01","2008-08-01","2008-09-01","2008-10-01","2008-11-01","2008-12-01","2009-01-01","2009-02-01","2009-03-01","2009-04-01","2009-05-01","2009-06-01","2009-07-01","2009-08-01","2009-09-01","2009-10-01","2009-11-01","2009-12-01","2010-01-01","2010-02-01","2010-03-01","2010-04-01","2010-05-01","2010-06-01","2010-07-01","2010-08-01","2010-09-01","2010-10-01","2010-11-01","2010-12-01","2011-01-01","2011-02-01","2011-03-01","2011-04-01","2011-05-01","2011-06-01","2011-07-01","2011-08-01","2011-09-01","2011-10-01","2011-11-01","2011-12-01","2012-01-01","2012-02-01","2012-03-01","2012-04-01","2012-05-01","2012-06-01","2012-07-01","2012-08-01","2012-09-01","2012-10-01","2012-11-01","2012-12-01","2013-01-01","2013-02-01","2013-03-01","2013-04-01","2013-05-01","2013-06-01","2013-07-01","2013-08-01","2013-09-01","2013-10-01","2013-11-01","2013-12-01","2014-01-01","2014-02-01","2014-03-01","2014-04-01","2014-05-01","2014-06-01","2014-07-01","2014-08-01","2014-09-01","2014-10-01","2014-11-01","2014-12-01","2015-01-01","2015-02-01","2015-03-01","2015-04-01"],"y":[0.01481541124844,0.0148056165822906,0.0148558858537614,0.0157693253257472,0.0153685751235601,0.015115923809333,0.0144038276745676,0.0150110044017607,0.0143809733274683,0.0135309278350515,0.0136753160545216,0.0146507360274464,0.0143642940420316,0.013778136168603,0.0133568711305602,0.0133588355109543,0.0134762193124398,0.0133170651866621,0.0134714512291832,0.0133345881979978,0.0134242140746351,0.013642591795648,0.013408721352635,0.0139056921489134,0.0141505942953567,0.0140774952311006,0.0149687330740066,0.0149973930408948,0.014034398034398,0.0141598134282558,0.0157027996212883,0.0169258068311047,0.0178050118536805,0.0185762300575829,0.0191537923922446,0.0198750183078651,0.020360688995962,0.0207311429893568,0.0216794784470176,0.0223093669212976,0.0237739292509611,0.0246123410816629,0.0241492545988201,0.0237240404900614,0.0241086751589278,0.0239490015212614,0.0241049889028274,0.0238549710308394,0.0242462474898994,0.0246968217393605,0.02422710629751,0.0237778684393674,0.0247464697561794,0.0246910031618281,0.0240238946567297,0.0235720674826964,0.0240808366632889,0.023683531850267,0.0234889880455272,0.0234735963762069,0.0234068300491672,0.0235106509579912,0.0230599492100933,0.023161455537132,0.021846042837612,0.0215490866659394,0.0205038272863,0.0210875331564987,0.0207996061613035,0.0210907198940498,0.0204606360804813,0.0206048756529049,0.020315324030598,0.0202977952963808,0.0204909345279313,0.0195034709965878,0.0206740220284621,0.0210964118711375,0.0218097796479627,0.0222035536949604,0.0217343382846101,0.0216440680349267,0.0220361289476519,0.0230571960727422,0.0236750306283726,0.023462684893619,0.0253773699392287,0.0257541349772209,0.0286080372743157,0.0308964438360756,0.0348995724209165,0.0349661730174598,0.0370728352493982,0.0381234531211546,0.0391280744978494,0.0380964739905825,0.0376297037129641,0.0366705983024584,0.0366139385285106,0.0364610987732412,0.035954993979822,0.0356979869174991,0.0347037011446602,0.0337216742079365,0.0332595765039263,0.0336968114449634,0.0324000275627627,0.0336085852906211,0.0343522828903616,0.0344494187405205,0.0337850210584142,0.0339821810797461,0.0348209144831242,0.0344511109284677,0.0332148609127699,0.0339330002188343,0.0332886872221007,0.0321325176162124,0.0314337825606411,0.0324204938967307,0.0310072239703231,0.0314118789066398,0.0305906981802364,0.0306151088255532,0.0308219023196704,0.0288563643511385,0.0292987533694244,0.0285070997026562,0.0285718150339056,0.0278389664445856,0.0275772360651015,0.0271068760989122,0.0283442280477121,0.0272884360763897,0.0274598413829897,0.0266357923778726,0.0271929549797071,0.0278445924799928,0.0272887677841556,0.0275515168286075,0.0272437398265213,0.0270408754310767,0.0259980768545888,0.0265029376053513,0.0266423763080136,0.0280521094564904,0.0274443907478264,0.0278829588885789,0.0275984727488309,0.0279566660625964,0.0295119032373449,0.0295690385429903,0.0296629035173177,0.0324053888602622,0.0351329587108528,0.0355921625161523,0.036723957738686,0.0363276640360074,0.0351511486243678,0.0354089231537057,0.0350943957447553,0.0337356138456764,0.035254240249501,0.0351463083498129,0.0348218336648867,0.0343020797461236,0.0356049221387346,0.0352453201370119,0.0341920109929294,0.0349107464800358,0.0357186257660191,0.0374868301819711,0.0391167219762414,0.0401187935356229,0.0406520243816973,0.0419561373550125,0.0427485203266082,0.0442228409851281,0.0445841188219612,0.0454239801372461,0.0467250676176202,0.0468217494578127,0.0482246269330479,0.0495197924541269,0.051237590828909,0.0516855378281009,0.0494338296431541,0.0494489726863492,0.0488328988540877,0.0481989554326485,0.0476825608536179,0.0480356059764734,0.0450178611821243,0.0453004464799724,0.0438089313637351,0.0420889969221862,0.0404078646236568,0.0396667162624609,0.0382692185143488,0.0373248077715082,0.0371104274954917,0.0371524641810727,0.0358315712754192,0.0348323170731707,0.0361204664308562,0.0360136800409218,0.0353395843892549,0.0353664506110323,0.0345675264274179,0.035218864299078,0.0354700422793808,0.0350207489835944,0.0350773139501624,0.0352879361076082,0.0348715105387401,0.0355059386410375,0.0356990095024029,0.0343390076211146,0.0345251948530335,0.03470325745568,0.0339647398529922,0.033982386617504,0.0325282300803712,0.0350392847014863,0.034939648561235,0.0348363557606604,0.0351228404593147,0.0353823312914052,0.0345687323135994,0.0337756483180682,0.0344716013738862,0.0341644769017797,0.0337892962599444,0.03262561046271,0.0326407040995269,0.0325094035464804,0.0324770013094899,0.0311328699040668,0.0312429101199144,0.0304936358240454,0.0299336090014992,0.029879182921008,0.0291995411618145,0.02968625485734,0.0288746875500228,0.0284484986198213,0.0284981207553047,0.0283823029439113,0.0281481420834374,0.0270040295362965,0.0277135031274273,0.0267395406975319,0.0269568730843479,0.0279032784211385,0.0269041488772284,0.0267325483428506,0.026585707081388,0.026489904737133,0.0271378907011502,0.0258099343285521,0.0251693505861356,0.0262158470499066,0.0258195426599597,0.0266152464044935,0.0262591876834504,0.0262993137377662,0.0265923104238629,0.0267266504613673,0.0270862450207628,0.0268312412718982,0.027153652190349,0.0267294144124231,0.0264967150177501,0.0272636839867472,0.0270150061106325,0.0263759310623617,0.0276733884508979,0.0287015999904168,0.0293837312712611,0.0297103844943579,0.0308896899095271,0.0313997758578207,0.0318195713191128,0.0327800583021001,0.0340212067899767,0.0334028649121488,0.0345415221835177,0.0343276449703207,0.033870757772404,0.0341440543405028,0.0343215570228942,0.0347515082437558,0.0350623041952277,0.0360756812726503,0.0363733964437688,0.0370094892111115,0.0369960461942175,0.0367784930779087,0.0380201728544394,0.0391287233669409,0.0383426627324889,0.0380473658020775,0.0379773867395592,0.036445992220615,0.0370525320844325,0.0369834334959928,0.0360485389227575,0.0354666903548986,0.0349447428536149,0.0351176112314678,0.0352318237831177,0.0350857622046214,0.0343125012007454,0.0336306348487523,0.0334039951392855,0.0335039802728564,0.0326747633164387,0.0323952704510192,0.0329491178570474,0.0327441697218481,0.0322848691833872,0.0317213124117107,0.030109138494429,0.0301232747613547,0.0301629238221048,0.0300806904187711,0.0292935682171981,0.0288761677027329,0.0278765795153444,0.0273032129424027,0.0278255685848387,0.0270931503750895,0.0269421269703761,0.0287670975146281,0.0279325408461718,0.0278927404514215,0.0282378628210852,0.028046454282106,0.027991555369228,0.0273988992581957,0.0277384532058346,0.0277036533889671,0.0279357526169956,0.0272503018288593,0.0272454811146894,0.0275801274302314,0.0275837210339343,0.026337671593921,0.0272076301512606,0.0254911547693128,0.0258209882937947,0.0259848252464142,0.0267131328494747,0.0267514983863532,0.0263782429245283,0.0261501923891231,0.0257522413647216,0.0252606741325257,0.0244361868532948,0.0249392932338549,0.0243851497918743,0.0241841331884042,0.0243316651617785,0.0235674744022319,0.0230113159641916,0.0236029915589054,0.0231878991792474,0.0229444254433521,0.0233487367613535,0.0215797808967541,0.021944085584474,0.0225206281993648,0.0226680911938866,0.0223538264512441,0.022767189227867,0.0226712346075674,0.0219996609888307,0.0217349005138257,0.0215126534432485,0.021982646982647,0.0207873528925442,0.0215621419926666,0.0207952869756778,0.021329672653503,0.021572172792209,0.0208796789722534,0.0211323208397195,0.0206207642316463,0.0203800036367396,0.0201377905071318,0.0203149023404134,0.0208328887940538,0.020372482756415,0.0194601158162704,0.0204273495176974,0.0200300574920426,0.0203516475733484,0.0207073691062893,0.0198811021729603,0.0195408914516545,0.0198939506725983,0.0198592859962777,0.0212137221752606,0.0214298032287242,0.0215966238790223,0.0220359054188438,0.0218601874934167,0.0227459289558061,0.0230732293758697,0.024659453023777,0.024985743922363,0.0268928828583213,0.0279491934441802,0.0288166940014656,0.0285297850677155,0.0286242917970411,0.0289146558027786,0.0299202844845284,0.0292014199142628,0.0291561292832727,0.0291213272938686,0.0287973366625052,0.0285879605568606,0.0287568802575553,0.0294701597338001,0.0298638498788509,0.0294282220794562,0.0297465776593468,0.0296228842644812,0.0304765187419216,0.0308494003705923,0.0318875092916334,0.0309848015954886,0.0305628847845207,0.0306225778436845,0.0299478005583488,0.0293892881253705,0.0284820963809211,0.0286455481327346,0.0279339736222842,0.029023007167736,0.0279050068481688,0.0280275907261849,0.0282583562680963,0.0277241083202993,0.0272028707710431,0.0269652448710927,0.0273969343710703,0.0269368959404481,0.0269228419988191,0.0263941352394257,0.0270412226156792,0.0262016275691107,0.0259636535923382,0.0258738468197928,0.0254244160899654,0.0250045579466957,0.024777357981379,0.0254560896776955,0.0250966421076735,0.0254570047138863,0.0244729029590056,0.0237328110143895,0.0241191993392736,0.0237267664228679,0.023870109058237,0.023383897941681,0.0234351725084438,0.0239969765481812,0.0236948770813632,0.0228573145409509,0.0224356729534577,0.0228994914926656,0.02251448358527,0.0236747024027361,0.0230284373109221,0.0223605662063444,0.0227382872924509,0.0224423931034254,0.0231292607187006,0.0236718718957365,0.0233799918614999,0.0236988755428927,0.0238997116975499,0.0238901317916939,0.0252071470257545,0.0253207514843199,0.0246846508687535,0.0257381369958573,0.0251120456929405,0.0275858216434512,0.0281557414728326,0.0293210585371295,0.0309396974216919,0.0310963646666164,0.0329696223908049,0.0344620093791083,0.036881870825218,0.0393784616992371,0.0420950254893898,0.0437916682975198,0.0451551076153814,0.0472304745524197,0.0478734139307175,0.0474923480755532,0.0481466434827827,0.0487390646412033,0.0498135884149012,0.0493453689473377,0.0489189425628497,0.0487188869065158,0.048905111851068,0.0491636805816074,0.0495647982697671,0.0479976903184301,0.0467584236605351,0.0468513367375324,0.0472580499069591,0.0470022764396316,0.0467654702453278,0.0485550730176732,0.0461674537759551,0.0450635665454102,0.0444211051857828,0.0441327555952861,0.0448149518168788,0.0444630462183337,0.0447799648292216,0.0441125967188235,0.0442593288282581,0.044643726197121,0.0434807419511074,0.0425215621352976,0.0418293730131926,0.0408610706387906,0.0408918166602682,0.0405519157520218,0.0403166252101388,0.040340243506583,0.0404181174638259,0.040278693679217,0.0396632120998043,0.0385034965036299,0.0385067391462337,0.0381045954672202,0.0390124437323695,0.0395415707991254,0.0378739686359739,0.0370300882122658,0.0372366533868074,0.0368816546440722,0.0371665916197347,0.0358295732195552,0.0356276086605975,0.0355738071667648,0.0351268697937214,0.0340039312919975,0.0327776351151127,0.0321227808883484,0.0325692322066761,0.0326500104494188,0.0305008766439861,0.0309768917383398,0.0297050702271821,0.0301510343386389,0.0301009314744063,0.0290230831466271,0.0281506024288415,0.0284449877176327,0.0272622510362181,0.0278280801400947,0.026899986422224,0.0265558477566239,0.0266102962839264],"type":"scatter","mode":"markers","marker":{"color":"rgba(31,119,180,1)","line":{"color":"rgba(31,119,180,1)"}},"error_y":{"color":"rgba(31,119,180,1)"},"error_x":{"color":"rgba(31,119,180,1)"},"line":{"color":"rgba(31,119,180,1)"},"xaxis":"x","yaxis":"y","frame":null}],"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
<pre class="r"><code>## -----------------------------------------------------------------------------
data(&quot;volcano&quot;)
volcanoData = list(
  x = 10 * seq_len(nrow(volcano)),
  y = 10 * seq_len(ncol(volcano)),
  z = volcano,
  col = terrain.colors(500)[cut(volcano, breaks = 500)]
)
library(&quot;rgl&quot;)
with(volcanoData, persp3d(x, y, z, color = col))</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
.volcanocut = cut(volcano, breaks = 500)
stopifnot(!any(is.na(.volcanocut)), all(as.integer(.volcanocut) %in% 1:500))</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
par(mai = rep(0, 4)) # A numerical vector of the form c(bottom, left, top, right) which gives the margin size specified in inches.
pie(rep(1, 8), col=1:8)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-247-1.svg" width="576" /></p>
<pre class="r"><code>tibble(u = factor(1:8), v = 1)</code></pre>
<pre><code>## # A tibble: 8 × 2
##   u         v
##   &lt;fct&gt; &lt;dbl&gt;
## 1 1         1
## 2 2         1
## 3 3         1
## 4 4         1
## 5 5         1
## 6 6         1
## 7 7         1
## 8 8         1</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
ggplot(tibble(u = factor(1:8), v = 1), 
       aes(x = &quot;&quot;,  y = v, fill = u)) +
  geom_bar(stat = &quot;identity&quot;, width = 1) + 
  coord_polar(&quot;y&quot;, start = 0) + theme_void()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-249-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
par(mai = c(0, 0.8, 0, 0))
display.brewer.all()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-250-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
head(brewer.pal.info)</code></pre>
<pre><code>##      maxcolors category colorblind
## BrBG        11      div       TRUE
## PiYG        11      div       TRUE
## PRGn        11      div       TRUE
## PuOr        11      div       TRUE
## RdBu        11      div       TRUE
## RdGy        11      div      FALSE</code></pre>
<pre class="r"><code>table(brewer.pal.info$category)</code></pre>
<pre><code>## 
##  div qual  seq 
##    9    8   18</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
brewer.pal(4, &quot;RdYlGn&quot;) #Creates nice looking color palettes especially for thematic maps</code></pre>
<pre><code>## [1] &quot;#D7191C&quot; &quot;#FDAE61&quot; &quot;#A6D96A&quot; &quot;#1A9641&quot;</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
par(mai = rep(0.1, 4))
mypalette  = colorRampPalette(
    c(&quot;darkorange3&quot;, &quot;white&quot;,&quot;darkblue&quot;)
  )(100)
head(mypalette)</code></pre>
<pre><code>## [1] &quot;#CD6600&quot; &quot;#CE6905&quot; &quot;#CF6C0A&quot; &quot;#D06F0F&quot; &quot;#D17214&quot; &quot;#D27519&quot;</code></pre>
<pre class="r"><code>image(matrix(1:100, nrow = 100, ncol = 10), col = mypalette,
        xaxt = &quot;n&quot;, yaxt = &quot;n&quot;, useRaster = TRUE)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-254-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;pheatmap&quot;)
topGenes = order(rowVars(Biobase::exprs(x)), decreasing = TRUE)[1:500]
rowCenter = function(x) { x - rowMeans(x) }
pheatmap( rowCenter(Biobase::exprs(x)[ topGenes, ] ),
  show_rownames = FALSE, show_colnames = FALSE,
  breaks = seq(-5, +5, length = 101),
  annotation_col =
    pData(x)[, c(&quot;sampleGroup&quot;, &quot;Embryonic.day&quot;, &quot;ScanDate&quot;) ],
  annotation_colors = list(
    sampleGroup = groupColor,
    genotype = c(`FGF4-KO` = &quot;chocolate1&quot;, `WT` = &quot;azure2&quot;),
    Embryonic.day = setNames(brewer.pal(9, &quot;Blues&quot;)[c(3, 6, 9)],
                             c(&quot;E3.25&quot;, &quot;E3.5&quot;, &quot;E4.5&quot;)),
    ScanDate = setNames(brewer.pal(nlevels(x$ScanDate), &quot;YlGn&quot;),
                        levels(x$ScanDate))
  ),
  cutree_rows = 4
)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-255-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
groupColor[1]</code></pre>
<pre><code>##     E3.25 
## &quot;#CAB2D6&quot;</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
hexvals = sapply(1:3, function(i) substr(groupColor[1], i*2, i*2+1))
decvals = strtoi(paste0(&quot;0x&quot;, hexvals))</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;colorspace&quot;)</code></pre>
<pre><code>## 
## Attaching package: &#39;colorspace&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:Gviz&#39;:
## 
##     coords</code></pre>
<pre class="r"><code>library(&quot;grid&quot;)

plothcl = function(h, c, l, what, x0 = 0.5, y0 = 0.5, default.units = &quot;npc&quot;, ...) {
  switch(what,
         &quot;c&quot; = {
           stopifnot(length(l)==1)
           n = length(c)
         },
         &quot;l&quot; = {
           stopifnot(length(c)==1)
           n = length(l)
         },
         stop(&quot;Sapperlot&quot;))

  cr = seq(0.1, 0.5, length = n+1)
  dr = 0.05 / n

  for (j in seq_len(n)) {
    r = c(cr[j]+dr, cr[j+1]-dr)
    for(i in 1:(length(h)-1)){
      phi = seq(h[i], h[i+1], by=1)/180*pi
      px = x0 + c(r[1]*cos(phi), r[2]*rev(cos(phi)))
      py = y0 + c(r[1]*sin(phi), r[2]*rev(sin(phi)))
      mycol = switch(what,
        &quot;c&quot; = hcl(h=mean(h[i+(0:1)]), c=c[j], l=l),
        &quot;l&quot; = hcl(h=mean(h[i+(0:1)]), c=c, l=l[j]))
      grid::grid.polygon(px, py, 
                         gp=gpar(col=mycol, fill=mycol),
                         default.units=default.units,...)
    }
  }
}


## -----------------------------------------------------------------------------
plothcl( h = seq(0, 360, by=3), c = seq(5, 75, by=10), l = 75,   what=&quot;c&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-257-1.svg" width="576" /></p>
<pre class="r"><code>grid.newpage()
plothcl( h = seq(0, 360, by=3), c = 55, l = seq(20, 100, by=10), what=&quot;l&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-257-2.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
gg = ggplot(tibble(A = Biobase::exprs(x)[, 1], M = rnorm(length(A))),
            aes(y = M))
gg + geom_point(aes(x = A), size = 0.2)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-258-1.svg" width="576" /></p>
<pre class="r"><code>gg + geom_point(aes(x = rank(A)), size = 0.2)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-259-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
volume = function(rho, nu)
            pi^(nu/2) * rho^nu / gamma(nu/2+1)

ggplot(tibble(nu    = 1:15,
  Omega = volume(1, nu)), aes(x = nu, y = Omega)) +
geom_line() +
xlab(expression(nu)) + ylab(expression(Omega)) +
geom_text(label =
&quot;Omega(rho,nu)==frac(pi^frac(nu,2)~rho^nu, Gamma(frac(nu,2)+1))&quot;,
  parse = TRUE, x = 6, y = 1.5)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-260-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
ggplot(genes, aes( x = value, color = gene)) + stat_ecdf() +
  theme(text = element_text(family = &quot;serif&quot;))</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-261-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
ggplot(genes, aes( x = value, color = gene)) + stat_ecdf() + theme(text = element_text(family = &quot;Bauhaus 93&quot;))</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-262-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;ggbio&quot;)</code></pre>
<pre><code>## Registered S3 method overwritten by &#39;GGally&#39;:
##   method from   
##   +.gg   ggplot2</code></pre>
<pre><code>## Need specific help about ggbio? try mailing 
##  the maintainer or visit https://lawremi.github.io/ggbio/</code></pre>
<pre><code>## 
## Attaching package: &#39;ggbio&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:Hmisc&#39;:
## 
##     zoom</code></pre>
<pre><code>## The following objects are masked from &#39;package:ggplot2&#39;:
## 
##     geom_bar, geom_rect, geom_segment, ggsave, stat_bin, stat_identity,
##     xlim</code></pre>
<pre class="r"><code>data(&quot;hg19IdeogramCyto&quot;, package = &quot;biovizBase&quot;)
plotIdeogram(hg19IdeogramCyto, subchr = &quot;chr1&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-263-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;GenomicRanges&quot;)
data(&quot;darned_hg19_subset500&quot;, package = &quot;biovizBase&quot;)
autoplot(darned_hg19_subset500, layout = &quot;karyogram&quot;,
         aes(color = exReg, fill = exReg))</code></pre>
<pre><code>## Warning in getIdeoGR(data): geom(ideogram) need valid seqlengths information for accurate mapping,
##                  now use reduced information as ideogram...

## Warning in getIdeoGR(data): geom(ideogram) need valid seqlengths information for accurate mapping,
##                  now use reduced information as ideogram...</code></pre>
<pre><code>## Scale for x is already present.
## Adding another scale for x, which will replace the existing scale.
## Scale for x is already present.
## Adding another scale for x, which will replace the existing scale.</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-264-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
data(&quot;ideoCyto&quot;, package = &quot;biovizBase&quot;)
dn = darned_hg19_subset500
seqlengths(dn) = seqlengths(ideoCyto$hg19)[names(seqlengths(dn))]
dn = keepSeqlevels(dn, paste0(&quot;chr&quot;, c(1:22, &quot;X&quot;)))
autoplot(dn, layout = &quot;karyogram&quot;, aes(color = exReg, fill = exReg))</code></pre>
<pre><code>## Scale for x is already present.
## Adding another scale for x, which will replace the existing scale.
## Scale for x is already present.
## Adding another scale for x, which will replace the existing scale.</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-265-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
darned_hg19_subset500[1:2,]</code></pre>
<pre><code>## GRanges object with 2 ranges and 10 metadata columns:
##       seqnames    ranges strand |       inchr       inrna         snp
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;character&gt;
##   [1]     chr5  86618225      - |           A           I        &lt;NA&gt;
##   [2]     chr7  99792382      - |           A           I        &lt;NA&gt;
##              gene      seqReg       exReg      source      ests      esta
##       &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        &lt;NA&gt;           O        &lt;NA&gt;    amygdala         0         0
##   [2]        &lt;NA&gt;           O        &lt;NA&gt;        &lt;NA&gt;         0         0
##            author
##       &lt;character&gt;
##   [1]    15342557
##   [2]    15342557
##   -------
##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
stopifnot(is(darned_hg19_subset500, &quot;GRanges&quot;), identical(start(darned_hg19_subset500),end(darned_hg19_subset500)))

## -----------------------------------------------------------------------------
ggcars = ggplot(mtcars, aes(x = hp, y = mpg)) + geom_point(size = 5, color = &#39;red&#39;)
ggcars</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-267-1.svg" width="576" /></p>
<pre class="r"><code>ggcars + theme_bw()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-267-2.svg" width="576" /></p>
<pre class="r"><code>ggcars + theme_minimal()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-267-3.svg" width="576" /></p>
</div>
<div id="mixture-models" class="section level2">
<h2>4 Mixture Models</h2>
<pre class="r"><code>## -----------------------------------------------------------------------------
coinflips = (runif(10000) &gt; 0.5)
table(coinflips)</code></pre>
<pre><code>## coinflips
## FALSE  TRUE 
##  4992  5008</code></pre>
<pre class="r"><code>oneFlip = function(fl, mean1 = 1, mean2 = 3, sd1 = 0.5, sd2 = 0.5) {
  if (fl) {
   rnorm(1, mean1, sd1)
  } else {
   rnorm(1, mean2, sd2)
  }
}
fairmix = vapply(coinflips, oneFlip, numeric(1))
library(&quot;ggplot2&quot;)
library(&quot;dplyr&quot;)
ggplot(tibble(value = fairmix), aes(x = value)) +
     geom_histogram(fill = &quot;purple&quot;, binwidth = 0.1)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-269-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
means = c(1, 3)
sds   = c(0.5, 0.5)
values = rnorm(length(coinflips),
               mean = ifelse(coinflips, means[1], means[2]),
               sd   = ifelse(coinflips, sds[1],   sds[2]))



## -----------------------------------------------------------------------------
fair = tibble(
  coinflips = (runif(1e6) &gt; 0.5),
  values = rnorm(length(coinflips),
                 mean = ifelse(coinflips, means[1], means[2]),
                 sd   = ifelse(coinflips, sds[1],   sds[2])))
ggplot(fair, aes(x = values)) +
     geom_histogram(fill = &quot;purple&quot;, bins = 200)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-270-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
ggplot(dplyr::filter(fair, coinflips), aes(x = values)) +
  geom_histogram(aes(y = after_stat(density)), fill = &quot;purple&quot;, binwidth = 0.01) +
  stat_function(fun = dnorm, color = &quot;red&quot;,
                args = list(mean = means[1], sd = sds[1]))</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-271-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
fairtheory = tibble(
  x = seq(-1, 5, length.out = 1000),
  f = 0.5 * dnorm(x, mean = means[1], sd = sds[1]) +
      0.5 * dnorm(x, mean = means[2], sd = sds[2]))
ggplot(fairtheory, aes(x = x, y = f)) +
  geom_line(color = &quot;red&quot;, linewidth = 1.5) + ylab(&quot;mixture density&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-272-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
mystery = tibble(
  coinflips = (runif(1e3) &gt; 0.5),
  values = rnorm(length(coinflips),
                 mean = ifelse(coinflips, 1, 2),
                 sd   = ifelse(coinflips, sqrt(.5), sqrt(.5))))
br2 = with(mystery, seq(min(values), max(values), length.out = 30))
ggplot(mystery, aes(x = values)) +
geom_histogram(fill = &quot;purple&quot;, breaks = br2)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-273-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
head(mystery, 3)</code></pre>
<pre><code>## # A tibble: 3 × 2
##   coinflips values
##   &lt;lgl&gt;      &lt;dbl&gt;
## 1 FALSE      2.83 
## 2 TRUE       1.31 
## 3 TRUE       0.404</code></pre>
<pre class="r"><code>br = with(mystery, seq(min(values), max(values), length.out = 30))
ggplot(mystery, aes(x = values)) +
  geom_histogram(data = dplyr::filter(mystery, coinflips),
     fill = &quot;red&quot;, alpha = 0.2, breaks = br) +
  geom_histogram(data = dplyr::filter(mystery, !coinflips),
     fill = &quot;darkblue&quot;, alpha = 0.2, breaks = br) </code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-274-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
stopifnot(identical(br2, br))

## -----------------------------------------------------------------------------
ggplot(mystery, aes(x = values, fill = coinflips)) +
  geom_histogram(data = dplyr::filter(mystery, coinflips),
     fill = &quot;red&quot;, alpha = 0.2, breaks = br) +
  geom_histogram(data = dplyr::filter(mystery, !coinflips),
     fill = &quot;darkblue&quot;, alpha = 0.2, breaks = br) +
  geom_histogram(fill = &quot;purple&quot;, breaks = br, alpha = 0.2)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-275-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
mus = c(-0.5, 1.5)
lambda = 0.5
u = sample(2, size = 100, replace = TRUE, prob = c(lambda, 1-lambda))
x = rnorm(length(u), mean = mus[u])
dux = tibble(u, x)
head(dux)</code></pre>
<pre><code>## # A tibble: 6 × 2
##       u      x
##   &lt;int&gt;  &lt;dbl&gt;
## 1     2  0.599
## 2     1 -2.98 
## 3     2  1.18 
## 4     1 -0.825
## 5     2  1.67 
## 6     2  2.30</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
group_by(dux, u) |&gt; summarise(mu = mean(x), sigma = sd(x))</code></pre>
<pre><code>## # A tibble: 2 × 3
##       u     mu sigma
##   &lt;int&gt;  &lt;dbl&gt; &lt;dbl&gt;
## 1     1 -0.580 0.927
## 2     2  1.47  0.974</code></pre>
<pre class="r"><code>table(dux$u) / nrow(dux)</code></pre>
<pre><code>## 
##    1    2 
## 0.56 0.44</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
.o = options(digits = 3)
library(&quot;mixtools&quot;)</code></pre>
<pre><code>## mixtools package, version 2.0.0, Released 2022-12-04
## This package is based upon work supported by the National Science Foundation under Grant No. SES-0518772 and the Chan Zuckerberg Initiative: Essential Open Source Software for Science (Grant No. 2020-255193).</code></pre>
<pre><code>## 
## Attaching package: &#39;mixtools&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:gtools&#39;:
## 
##     ddirichlet</code></pre>
<pre><code>## The following object is masked from &#39;package:grid&#39;:
## 
##     depth</code></pre>
<pre class="r"><code>y = c(rnorm(100, mean = -0.2, sd = 0.5),
      rnorm( 50, mean =  0.5, sd =   1))
gm = normalmixEM(y, k = 2, 
                    lambda = c(0.5, 0.5),
                    mu = c(-0.01, 0.01), 
                    sigma = c(3, 3))</code></pre>
<pre><code>## number of iterations= 296</code></pre>
<pre class="r"><code>with(gm, c(lambda, mu, sigma, loglik))</code></pre>
<pre><code>## [1]    0.776    0.224   -0.185    0.823    0.534    0.989 -164.826</code></pre>
<pre class="r"><code>options(.o)</code></pre>
<pre class="r"><code>gm$lambda</code></pre>
<pre><code>## [1] 0.775846 0.224154</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
## PROVENANCE: here&#39;s a record of how the data were created
library(&quot;mosaics&quot;)</code></pre>
<pre><code>## Loading required package: Rcpp</code></pre>
<pre><code>## 
## Attaching package: &#39;mosaics&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:plotly&#39;:
## 
##     export</code></pre>
<pre><code>## The following object is masked from &#39;package:BSgenome&#39;:
## 
##     export</code></pre>
<pre><code>## The following object is masked from &#39;package:rtracklayer&#39;:
## 
##     export</code></pre>
<pre class="r"><code>library(&quot;mosaicsExample&quot;)
setwd(&quot;D:/books/R/blogdown/hugo&quot;)
for(f in c(&quot;wgEncodeSydhTfbsGm12878Stat1StdAlnRep1_chr22_sorted.bam&quot;,
           &quot;wgEncodeSydhTfbsGm12878InputStdAlnRep1_chr22_sorted.bam&quot;))
  constructBins(infile = system.file(file.path(&quot;extdata&quot;, f), package=&quot;mosaicsExample&quot;),
    fileFormat = &quot;bam&quot;, outfileLoc = &quot;../data/&quot;,
    byChr = FALSE, useChrfile = FALSE, chrfile = NULL, excludeChr = NULL,
    PET = FALSE, fragLen = 200, binSize = 200, capping = 0)</code></pre>
<pre><code>## ------------------------------------------------------------
## Info: setting summary
## ------------------------------------------------------------
## Name of aligned read file: C:/Users/lidan/R/library/mosaicsExample/extdata/wgEncodeSydhTfbsGm12878Stat1StdAlnRep1_chr22_sorted.bam 
## Aligned read file format: BAM 
## Directory of processed bin-level files: ../data/ 
## Construct bin-level files by chromosome? N 
## Is file for chromosome info provided? N 
## Data type: Single-end tag (SET)
## Average fragment length: 200 
## Bin size: 200 
## ------------------------------------------------------------</code></pre>
<pre><code>## Use the provided BAM index file.</code></pre>
<pre><code>## Chromosome information is extracted from the BAM file.</code></pre>
<pre><code>## Info: reading the aligned read file and processing it into bin-level files...</code></pre>
<pre><code>## Info: done!</code></pre>
<pre><code>## ------------------------------------------------------------
## Info: processing summary
## ------------------------------------------------------------
## Directory of processed bin-level files: ../data/ 
## Processed bin-level file: wgEncodeSydhTfbsGm12878Stat1StdAlnRep1_chr22_sorted.bam_fragL200_bin200.txt
## Sequencing depth: 231822 
## ------------------------------------------------------------
## ------------------------------------------------------------
## Info: setting summary
## ------------------------------------------------------------
## Name of aligned read file: C:/Users/lidan/R/library/mosaicsExample/extdata/wgEncodeSydhTfbsGm12878InputStdAlnRep1_chr22_sorted.bam 
## Aligned read file format: BAM 
## Directory of processed bin-level files: ../data/ 
## Construct bin-level files by chromosome? N 
## Is file for chromosome info provided? N 
## Data type: Single-end tag (SET)
## Average fragment length: 200 
## Bin size: 200 
## ------------------------------------------------------------</code></pre>
<pre><code>## Use the provided BAM index file.</code></pre>
<pre><code>## Chromosome information is extracted from the BAM file.</code></pre>
<pre><code>## Info: reading the aligned read file and processing it into bin-level files...</code></pre>
<pre><code>## Info: done!</code></pre>
<pre><code>## ------------------------------------------------------------
## Info: processing summary
## ------------------------------------------------------------
## Directory of processed bin-level files: ../data/ 
## Processed bin-level file: wgEncodeSydhTfbsGm12878InputStdAlnRep1_chr22_sorted.bam_fragL200_bin200.txt
## Sequencing depth: 182605 
## ------------------------------------------------------------</code></pre>
<pre class="r"><code>datafiles = c(&quot;../data/wgEncodeSydhTfbsGm12878Stat1StdAlnRep1_chr22_sorted.bam_fragL200_bin200.txt&quot;,
              &quot;../data/wgEncodeSydhTfbsGm12878InputStdAlnRep1_chr22_sorted.bam_fragL200_bin200.txt&quot;)
binTFBS = readBins(type = c(&quot;chip&quot;, &quot;input&quot;), fileName = datafiles)</code></pre>
<pre><code>## Info: reading and preprocessing bin-level data...</code></pre>
<pre><code>## Info: data contains only one chromosome.</code></pre>
<pre><code>## Info: done!</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
binTFBS</code></pre>
<pre><code>## Summary: bin-level data (class: BinData)
## ----------------------------------------
## - # of chromosomes in the data: 1
## - total effective tag counts: 462479
##   (sum of ChIP tag counts of all bins)
## - control sample is incorporated
## - mappability score is NOT incorporated
## - GC content score is NOT incorporated
## - uni-reads are assumed
## ----------------------------------------</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
bincts = print(binTFBS)
ggplot(bincts, aes(x = tagCount)) +
  geom_histogram(binwidth = 1, fill = &quot;forestgreen&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-282-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
ggplot(bincts, aes(x = tagCount)) + scale_y_log10() +
   geom_histogram(binwidth = 1, fill = &quot;forestgreen&quot;)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in continuous y-axis</code></pre>
<pre><code>## Warning: Removed 17 rows containing missing values (`geom_bar()`).</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-283-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
masses = c(A =  331, C =  307, G =  347, T =  322)
probs  = c(A = 0.12, C = 0.38, G = 0.36, T = 0.14)
N  = 7000
sd = 3
nuclt   = sample(length(probs), N, replace = TRUE, prob = probs)
quadwts = rnorm(length(nuclt),
                mean = masses[nuclt],
                sd   = sd)
ggplot(tibble(quadwts = quadwts), aes(x = quadwts)) +
  geom_histogram(bins = 100, fill = &quot;purple&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-284-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;HistData&quot;)
ZeaMays$diff</code></pre>
<pre><code>##  [1]  6.125 -8.375  1.000  2.000  0.750  2.875  3.500  5.125  1.750  3.625
## [11]  7.000  3.000  9.375  7.500 -6.000</code></pre>
<pre class="r"><code>ggplot(ZeaMays, aes(x = diff, ymax = 1/15, ymin = 0)) +
  geom_linerange(linewidth = 1, col = &quot;forestgreen&quot;) + ylim(0, 0.1)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-286-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
stopifnot(nrow(ZeaMays) == 15)</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
B = 1000
meds = replicate(B, {
  i = sample(15, 15, replace = TRUE)
  median(ZeaMays$diff[i])
})
ggplot(tibble(medians = meds), aes(x = medians)) +
  geom_histogram(bins = 30, fill = &quot;purple&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-287-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;bootstrap&quot;)
bootstrap(ZeaMays$diff, B, mean)</code></pre>
<pre><code>## $thetastar
##    [1]  1.25833333  1.20833333  4.65000000  3.54166667  4.99166667  4.42500000
##    [7]  2.72500000  3.58333333  1.50833333  2.76666667  3.03333333  1.26666667
##   [13]  2.40833333  3.20000000  1.21666667  3.87500000  3.76666667  2.67500000
##   [19]  2.43333333  2.00833333  1.95000000  3.16666667  2.55833333  1.97500000
##   [25]  1.45000000  1.21666667  2.21666667  2.68333333 -0.70833333  2.12500000
##   [31]  2.87500000  4.55833333  5.39166667  3.76666667  1.94166667  2.76666667
##   [37]  1.30833333  1.85000000  2.95000000  2.76666667  4.50833333  1.97500000
##   [43]  2.50000000  3.17500000  3.01666667 -0.57500000  4.51666667  2.35000000
##   [49]  3.77500000  3.44166667  2.87500000  3.52500000  2.64166667  2.34166667
##   [55]  2.08333333  3.15833333  3.07500000  3.54166667  1.67500000  1.67500000
##   [61]  2.93333333  4.44166667  3.05000000  1.33333333  2.06666667  3.25833333
##   [67]  1.75833333  3.84166667  3.29166667  6.08333333  1.18333333  3.01666667
##   [73]  2.39166667  3.65833333  2.55833333  0.34166667  0.64166667  4.47500000
##   [79]  0.85833333  4.06666667  2.87500000  1.67500000  1.94166667  2.33333333
##   [85]  3.87500000  5.15833333  2.90000000  2.34166667  3.15000000  1.29166667
##   [91]  1.75833333  1.79166667  3.77500000  1.62500000  2.16666667  1.76666667
##   [97]  0.77500000  2.54166667  2.65833333  4.29166667  2.05000000  1.73333333
##  [103]  1.64166667  3.80000000  3.78333333  0.85833333  2.92500000  2.65000000
##  [109]  2.13333333  2.56666667  3.02500000  2.16666667  2.41666667  0.22500000
##  [115]  2.88333333  2.32500000  0.61666667  1.54166667  2.59166667  0.24166667
##  [121]  2.10833333  1.12500000  3.28333333  3.15833333  1.90000000  3.86666667
##  [127]  1.96666667  2.14166667  2.96666667  2.04166667  1.71666667  2.45000000
##  [133]  2.24166667  2.82500000  2.12500000  3.18333333  4.08333333  1.69166667
##  [139]  2.71666667  1.55000000  3.91666667  3.22500000  1.89166667  3.10000000
##  [145]  2.30833333  2.67500000  2.20000000  2.03333333  2.93333333  3.00000000
##  [151]  2.59166667  2.94166667  1.03333333  0.91666667  2.57500000  2.75833333
##  [157]  0.93333333  4.00000000  1.35833333  2.20833333  0.07500000  2.84166667
##  [163]  4.78333333  1.48333333  3.65833333  2.98333333  1.95833333  2.38333333
##  [169]  2.31666667  3.17500000  1.23333333  3.90833333  4.21666667  2.07500000
##  [175]  4.10833333  2.65833333  3.67500000  2.50000000  2.26666667  2.63333333
##  [181]  2.43333333  2.65000000  2.60833333  3.56666667  3.54166667  0.36666667
##  [187]  4.27500000  3.91666667  2.19166667  1.62500000  2.97500000  1.30833333
##  [193]  3.41666667  2.04166667  2.03333333  3.84166667  4.16666667  2.23333333
##  [199]  2.40833333  4.95000000  3.50833333  4.05000000  4.12500000  2.76666667
##  [205]  3.64166667  3.40833333  1.25833333  2.15000000  1.38333333  5.69166667
##  [211]  1.63333333  3.37500000  0.01666667  2.44166667  3.20833333  1.19166667
##  [217]  3.54166667  4.90000000  1.45000000  3.55000000  3.58333333  0.61666667
##  [223]  1.50833333  3.80833333  4.55833333  3.56666667  1.41666667  0.20833333
##  [229]  2.47500000  3.00000000 -0.12500000  4.04166667  2.21666667  2.25000000
##  [235]  3.34166667  3.28333333  3.26666667  4.86666667  2.81666667  4.79166667
##  [241]  3.07500000  3.37500000  2.05833333  0.69166667  4.61666667  2.49166667
##  [247]  2.99166667  2.61666667  2.50833333  1.86666667  3.28333333  2.70833333
##  [253]  3.65833333  3.15000000  2.63333333  3.43333333  1.77500000  3.93333333
##  [259]  1.95833333  4.55000000  3.34166667  3.55000000  3.25000000  2.84166667
##  [265]  2.92500000  1.67500000  2.74166667  4.34166667  4.18333333  4.20833333
##  [271]  3.30833333  3.80833333  5.23333333  0.60833333  3.82500000  0.86666667
##  [277]  2.66666667  5.10000000  4.15833333  1.69166667  1.37500000  3.85000000
##  [283]  4.29166667  3.48333333 -0.65000000  0.22500000  4.25833333  2.15000000
##  [289]  3.50000000  2.70833333  0.27500000  0.61666667  2.40000000  1.64166667
##  [295]  1.21666667  4.58333333  3.03333333  3.62500000  1.61666667  3.00833333
##  [301]  4.40833333  3.41666667  2.27500000  1.97500000  2.71666667  1.70833333
##  [307]  1.96666667  3.05000000 -0.25000000  2.65833333  3.31666667  4.12500000
##  [313]  3.30000000  3.90833333  1.51666667  2.40000000  1.22500000  2.85000000
##  [319]  3.38333333  3.10000000  2.65833333  2.63333333  1.52500000  3.36666667
##  [325]  2.53333333  1.84166667  3.12500000  0.67500000 -0.90833333  1.40833333
##  [331]  3.35000000  5.58333333  3.05000000  2.00833333  0.97500000  1.49166667
##  [337] -0.47500000  2.35000000  3.89166667  3.69166667  3.20833333  3.85000000
##  [343]  5.30000000  2.30833333  5.37500000  4.30000000  0.57500000  1.67500000
##  [349]  2.25833333  5.06666667  3.30000000 -0.85000000  2.00833333  4.43333333
##  [355]  2.55833333  3.45833333  4.34166667  2.35833333  2.60833333  1.59166667
##  [361]  2.55000000  0.44166667  4.05000000  1.53333333  2.53333333  2.90833333
##  [367]  0.12500000  2.46666667  2.52500000  3.09166667  1.35833333  2.09166667
##  [373]  1.68333333  3.40000000  3.79166667  3.95833333  2.60000000  2.70000000
##  [379]  4.67500000  2.94166667  3.00833333  1.39166667  2.32500000  0.94166667
##  [385]  2.29166667  0.52500000  2.99166667  1.29166667  2.10000000  2.66666667
##  [391]  2.75000000  1.93333333  4.59166667  3.38333333  4.09166667  3.37500000
##  [397]  1.99166667  3.46666667  3.50833333  5.14166667  3.26666667  2.88333333
##  [403]  0.76666667  2.50000000  3.63333333  2.95000000  2.48333333  3.17500000
##  [409]  1.51666667  3.01666667  1.26666667  2.46666667  3.22500000  1.15000000
##  [415]  2.65000000  3.34166667  1.51666667  2.41666667  4.85833333  2.99166667
##  [421]  3.53333333  1.25833333 -0.84166667  3.47500000  4.37500000  1.99166667
##  [427]  4.25000000  2.19166667  3.95833333  4.28333333  3.90000000  2.68333333
##  [433]  2.45833333  3.42500000  3.22500000  3.25833333  3.91666667  0.41666667
##  [439]  2.90833333  2.69166667  2.92500000  3.27500000  2.25833333  1.00833333
##  [445]  3.00000000  1.80833333  2.10000000  3.69166667  2.36666667  2.11666667
##  [451]  2.56666667  3.63333333  2.21666667  2.70833333  2.60833333  4.25833333
##  [457]  3.55000000  1.90833333 -0.55000000  2.74166667  2.22500000  1.67500000
##  [463]  1.05000000  0.69166667  1.97500000  2.94166667  5.25833333  1.12500000
##  [469]  1.63333333  1.97500000  3.52500000  2.49166667  4.01666667  1.41666667
##  [475]  3.60000000  3.68333333  3.75833333  2.92500000  1.51666667  2.88333333
##  [481]  3.10833333  2.66666667  3.35000000  2.38333333  3.57500000  3.01666667
##  [487]  2.81666667  4.58333333  2.35833333  2.56666667  4.08333333  1.29166667
##  [493]  2.09166667  0.50000000  1.56666667  2.15000000  3.26666667  0.96666667
##  [499]  2.02500000  4.88333333  3.94166667  2.92500000  2.94166667  3.00833333
##  [505]  0.85833333  2.54166667  2.10833333  2.85833333  0.70833333  3.40833333
##  [511]  1.87500000  2.15000000  1.10000000  3.08333333  1.20000000  3.17500000
##  [517]  2.18333333  3.40000000  4.69166667  3.90833333  1.42500000  3.49166667
##  [523]  4.09166667  2.79166667  2.11666667  3.32500000  4.50000000  2.66666667
##  [529]  3.25000000  4.15000000  3.33333333  0.89166667  2.32500000  1.79166667
##  [535]  3.26666667  1.16666667  1.86666667  1.00000000  4.55000000  3.02500000
##  [541]  2.62500000  2.48333333  3.44166667  3.55833333  0.91666667  3.63333333
##  [547]  2.55833333  1.10833333  1.34166667  1.82500000  2.45000000  2.98333333
##  [553]  1.35000000  4.59166667  1.90833333  3.72500000  3.65833333  1.50000000
##  [559]  2.19166667  2.70000000  3.58333333  3.17500000  1.55833333  0.54166667
##  [565]  3.08333333  2.03333333  0.88333333  2.52500000  2.19166667  3.15000000
##  [571]  3.20833333  2.84166667  3.15833333  3.59166667  1.25833333  4.67500000
##  [577]  2.14166667  3.56666667  0.00000000  1.89166667  3.52500000  2.68333333
##  [583]  2.38333333  3.22500000  2.37500000  1.44166667  2.20833333  4.70833333
##  [589]  2.70833333  4.06666667  4.26666667  3.65833333  1.03333333  2.77500000
##  [595]  4.81666667  2.10833333  0.52500000  3.55833333  1.12500000  1.20000000
##  [601]  3.45833333  4.20833333  4.34166667  2.04166667 -1.26666667  2.60000000
##  [607]  2.59166667  2.42500000  0.75000000  4.22500000  4.20000000  4.00000000
##  [613]  4.33333333  2.53333333  3.58333333  3.93333333  3.60000000  3.17500000
##  [619]  4.30000000  4.45000000  2.95833333  3.42500000  4.23333333  3.53333333
##  [625]  1.77500000  1.10833333  4.49166667  2.44166667  0.25833333  3.30000000
##  [631]  1.74166667  2.03333333  1.49166667  2.36666667  3.25000000  1.79166667
##  [637]  0.10000000  4.47500000  3.29166667  3.50000000  2.07500000  0.61666667
##  [643]  3.77500000  5.78333333  3.40833333  1.35000000  3.32500000  3.08333333
##  [649]  2.10000000  2.27500000  0.65833333  2.29166667  1.84166667  2.51666667
##  [655]  4.40833333  3.86666667  3.22500000  0.32500000  2.17500000  4.22500000
##  [661]  2.74166667  2.69166667  0.13333333  0.12500000  3.56666667  1.45000000
##  [667]  2.26666667  2.93333333  0.90000000  2.33333333  4.63333333  2.66666667
##  [673]  4.13333333  3.27500000  2.57500000  3.19166667  2.96666667  4.02500000
##  [679]  1.94166667  2.60000000  0.63333333  2.00833333  3.68333333  2.49166667
##  [685]  3.82500000  0.88333333  1.97500000  2.42500000  0.47500000  0.53333333
##  [691] -0.08333333  0.13333333  4.39166667  2.42500000 -1.14166667  1.49166667
##  [697]  3.35000000  1.89166667  3.59166667 -0.43333333  3.47500000  3.80000000
##  [703]  0.99166667 -0.22500000  2.13333333  2.25000000  1.20833333  3.01666667
##  [709]  3.32500000  2.74166667  3.06666667  2.40000000  2.31666667  4.16666667
##  [715]  2.68333333  1.60000000  2.82500000  1.97500000  3.42500000  3.00000000
##  [721]  1.23333333  2.12500000  1.20000000  2.21666667  3.90000000  1.87500000
##  [727]  0.67500000  1.81666667  3.66666667  0.95833333  4.50000000  3.23333333
##  [733]  3.20000000  1.30833333  4.86666667 -0.34166667  4.55833333  1.78333333
##  [739]  1.77500000  1.38333333  3.90833333  2.20833333  3.71666667  3.01666667
##  [745]  3.02500000  1.40000000  3.41666667  2.14166667  4.77500000  2.48333333
##  [751]  4.45000000  0.80833333  1.97500000  3.12500000  3.25000000  4.17500000
##  [757]  3.14166667  1.31666667  2.00833333  1.92500000  3.19166667  0.35833333
##  [763]  3.58333333  3.41666667  2.93333333  2.79166667  1.18333333  3.39166667
##  [769]  0.55000000  2.75833333  0.52500000  1.05000000  2.94166667  2.45000000
##  [775]  1.75833333  3.30000000  2.44166667  3.08333333  1.19166667  2.81666667
##  [781]  3.93333333  3.05833333  3.60833333  4.02500000  1.44166667  3.24166667
##  [787]  1.30000000  2.85000000  3.67500000  2.78333333  2.43333333  4.15833333
##  [793]  4.05000000  2.50000000  1.61666667  3.33333333  2.10000000  3.09166667
##  [799]  2.08333333  3.43333333  3.90833333  1.08333333  3.13333333  2.65000000
##  [805]  2.01666667  1.95000000  2.37500000  2.84166667  4.04166667  1.87500000
##  [811]  1.34166667  2.30833333  3.70833333  3.65000000  0.40833333  2.77500000
##  [817]  2.46666667  4.18333333  4.23333333  2.10833333  2.36666667  2.50000000
##  [823]  4.20000000  4.35833333  2.56666667  1.65000000  5.14166667  1.50833333
##  [829]  3.07500000  1.78333333  3.35833333  3.90000000  3.32500000  4.84166667
##  [835]  1.13333333  4.21666667  2.15833333  0.88333333  4.14166667  1.30000000
##  [841]  4.26666667  1.45000000  2.85833333  4.74166667  2.32500000  2.17500000
##  [847]  0.96666667  2.26666667  0.82500000  3.65833333  2.43333333  3.16666667
##  [853]  3.42500000  0.75000000  1.23333333  3.84166667  2.15833333  3.18333333
##  [859]  2.95833333  4.20833333  3.40000000  1.93333333  3.42500000  2.59166667
##  [865]  1.72500000  2.55833333  4.05000000  1.30833333  3.15000000  0.90000000
##  [871]  3.23333333  2.85000000  2.37500000  2.65000000  3.27500000  0.26666667
##  [877]  1.58333333  2.70833333  4.80833333  0.43333333  2.82500000  2.31666667
##  [883]  3.00000000  2.45833333  1.27500000  3.09166667  2.14166667  1.66666667
##  [889]  0.90000000  3.46666667  3.25833333  3.70000000  1.13333333  5.33333333
##  [895]  3.02500000  2.50833333  2.34166667  0.80833333  1.72500000  3.57500000
##  [901]  3.54166667  3.84166667  2.02500000 -0.55000000  2.35000000  0.91666667
##  [907]  3.20000000  3.25000000  3.24166667  4.74166667  3.18333333  2.32500000
##  [913]  2.25000000  3.21666667  2.45833333  3.54166667  2.74166667  3.07500000
##  [919]  1.81666667  1.03333333  3.21666667 -0.52500000  1.75833333  1.61666667
##  [925]  2.15833333  5.04166667  2.04166667  2.92500000  2.85000000  3.14166667
##  [931]  3.21666667  2.18333333  2.37500000  2.94166667  3.62500000  5.34166667
##  [937]  3.23333333  3.82500000  2.50833333  5.04166667  1.78333333 -0.11666667
##  [943]  3.26666667  3.46666667  2.75000000  2.83333333  2.39166667  1.66666667
##  [949]  3.23333333  3.32500000  1.54166667  0.64166667  1.30833333  1.50000000
##  [955]  0.83333333  4.55000000  1.75833333  2.72500000  3.76666667  2.64166667
##  [961]  1.17500000  3.07500000  1.50833333  3.30000000  0.59166667  2.29166667
##  [967]  2.55000000  4.30833333  3.80000000  3.93333333  2.84166667  1.87500000
##  [973]  3.39166667  3.21666667  4.09166667  3.98333333  0.53333333  3.15833333
##  [979]  3.41666667  4.19166667  2.39166667  1.50833333  2.68333333  1.90000000
##  [985]  1.15000000  3.95833333  0.71666667  1.94166667  2.49166667  1.59166667
##  [991]  4.83333333  3.15000000  2.30833333  2.49166667  4.42500000  2.40000000
##  [997]  2.55000000  4.45000000  2.66666667  0.92500000
## 
## $func.thetastar
## NULL
## 
## $jack.boot.val
## NULL
## 
## $jack.boot.se
## NULL
## 
## $call
## bootstrap(x = ZeaMays$diff, nboot = B, theta = mean)</code></pre>
<pre class="r"><code>bootstrap(ZeaMays$diff, B, median)</code></pre>
<pre><code>## $thetastar
##    [1] 2.875 2.875 3.000 5.125 3.000 3.500 3.500 2.875 3.500 2.875 2.000 2.000
##   [13] 3.000 3.000 3.500 2.000 2.875 3.000 1.000 3.500 3.000 3.000 3.000 2.875
##   [25] 2.000 3.625 3.000 2.875 1.750 5.125 3.500 5.125 3.625 2.875 3.625 3.625
##   [37] 2.875 2.000 3.500 6.125 3.500 3.000 3.000 3.625 3.625 2.875 3.500 2.000
##   [49] 3.000 1.750 2.875 3.625 6.125 2.875 3.500 1.000 5.125 2.000 1.750 5.125
##   [61] 3.000 2.875 2.875 3.000 5.125 2.000 3.625 3.000 3.500 3.000 3.000 3.000
##   [73] 3.000 3.000 2.875 2.875 3.000 2.875 2.875 2.875 3.625 3.500 3.000 2.000
##   [85] 2.875 3.500 2.875 3.000 1.750 2.875 3.500 2.000 3.000 2.000 1.000 2.875
##   [97] 6.125 6.125 2.875 1.750 6.125 3.500 2.875 3.000 2.875 6.125 3.625 2.875
##  [109] 2.000 3.000 3.500 2.875 2.000 2.875 2.875 3.625 1.000 2.875 3.625 2.875
##  [121] 1.750 2.000 3.500 3.500 1.750 3.625 7.000 3.500 3.000 2.875 2.000 2.875
##  [133] 2.000 2.875 2.875 3.000 2.000 1.750 2.000 1.750 2.000 1.750 3.500 3.625
##  [145] 2.875 1.750 2.875 3.500 3.000 3.625 3.500 3.500 1.750 5.125 2.875 2.875
##  [157] 3.500 1.750 3.000 2.000 2.000 2.875 3.500 3.000 3.500 3.625 3.000 2.875
##  [169] 5.125 1.750 2.000 3.000 3.000 3.500 3.000 5.125 6.125 3.000 3.625 2.875
##  [181] 2.875 3.000 2.875 3.500 3.500 3.000 3.500 2.875 1.750 5.125 3.500 2.000
##  [193] 2.000 3.000 3.625 2.875 2.875 3.000 3.000 3.000 3.625 2.875 2.000 3.500
##  [205] 2.000 3.000 3.500 2.000 6.125 2.875 2.875 3.625 3.000 2.875 3.000 2.000
##  [217] 3.500 3.625 2.000 3.625 3.625 3.000 3.500 3.500 0.750 3.000 5.125 5.125
##  [229] 3.000 3.625 2.000 1.750 6.125 3.500 2.875 3.000 1.750 3.625 2.875 2.875
##  [241] 2.000 3.500 3.000 2.875 1.000 2.875 3.625 2.000 2.875 3.000 2.875 3.500
##  [253] 2.875 3.500 3.000 1.750 5.125 3.625 2.000 3.500 2.875 3.000 2.875 3.000
##  [265] 5.125 2.875 1.000 5.125 3.000 3.000 2.875 3.000 3.000 2.875 3.500 2.875
##  [277] 3.500 3.500 3.000 1.000 2.875 2.000 3.500 1.750 5.125 3.500 2.875 3.625
##  [289] 0.750 1.750 6.125 3.000 3.000 3.625 2.875 3.000 2.000 3.000 3.000 3.000
##  [301] 2.875 3.000 3.625 3.500 2.000 2.000 3.000 1.750 2.000 3.000 3.500 3.500
##  [313] 3.625 3.500 6.125 3.000 5.125 5.125 3.500 2.875 3.500 3.625 2.875 3.000
##  [325] 2.000 7.000 2.875 3.000 2.000 3.625 3.625 2.000 2.000 3.500 2.875 2.000
##  [337] 2.875 2.000 2.875 3.000 2.000 3.000 3.000 2.875 3.500 2.875 3.000 2.875
##  [349] 2.000 5.125 3.625 1.750 3.000 1.000 2.875 1.000 2.000 3.500 3.500 3.000
##  [361] 1.000 3.500 2.000 2.875 3.625 1.750 3.625 2.875 3.000 3.500 3.500 3.500
##  [373] 3.500 3.000 3.000 2.000 3.500 3.000 6.125 5.125 2.875 3.000 3.625 3.625
##  [385] 3.500 3.000 1.750 3.000 3.625 3.000 2.000 3.000 2.875 5.125 1.750 2.875
##  [397] 3.500 3.500 3.000 2.000 3.000 2.000 2.875 2.875 3.500 2.000 2.875 2.875
##  [409] 3.000 3.500 2.875 2.000 3.000 3.500 3.000 3.500 2.875 5.125 2.875 2.875
##  [421] 5.125 3.625 3.500 1.000 5.125 3.625 3.000 1.750 2.000 3.625 2.000 2.000
##  [433] 3.625 2.875 3.000 3.000 3.000 3.625 1.750 3.000 3.625 3.000 2.875 3.500
##  [445] 3.625 5.125 3.000 1.000 2.875 3.625 3.000 5.125 1.000 6.125 3.000 3.500
##  [457] 1.750 3.500 1.000 1.750 1.000 6.125 3.500 2.875 5.125 6.125 2.000 3.500
##  [469] 5.125 2.000 3.500 3.625 2.875 5.125 2.875 1.750 2.875 2.875 3.625 2.000
##  [481] 3.500 1.000 3.000 1.000 1.750 6.125 3.500 3.625 3.500 2.875 3.500 2.000
##  [493] 2.000 2.875 2.000 2.875 1.750 2.875 3.625 3.500 3.000 3.000 5.125 3.625
##  [505] 3.625 3.000 3.625 1.750 2.000 3.000 3.500 2.000 3.000 2.875 2.875 2.000
##  [517] 3.000 3.625 2.000 3.500 2.000 2.000 3.500 2.875 2.000 3.625 3.625 5.125
##  [529] 3.000 6.125 2.875 3.625 3.000 2.000 5.125 3.500 3.000 3.000 2.875 1.000
##  [541] 3.625 3.500 3.500 3.500 2.875 2.875 3.000 3.500 3.500 3.500 5.125 2.000
##  [553] 3.625 2.875 3.500 3.500 3.500 3.500 3.000 2.000 2.000 1.000 3.500 3.625
##  [565] 2.875 2.875 5.125 5.125 3.000 1.000 3.000 1.000 3.000 2.875 2.875 2.875
##  [577] 6.125 3.500 3.500 3.000 3.500 3.000 3.500 1.750 3.625 2.000 2.000 3.000
##  [589] 3.625 2.000 2.875 3.500 3.500 3.625 3.625 3.500 2.875 3.000 3.000 6.125
##  [601] 3.000 2.875 1.750 5.125 2.875 3.625 3.625 3.000 1.000 3.500 3.000 3.500
##  [613] 2.875 3.625 3.500 2.000 2.875 1.750 5.125 3.625 6.125 3.000 1.750 2.875
##  [625] 3.500 1.750 3.500 3.500 3.000 3.500 3.625 3.500 3.000 3.500 2.875 1.750
##  [637] 1.000 2.875 1.000 3.625 5.125 5.125 2.000 3.500 3.625 1.750 2.000 3.500
##  [649] 3.500 2.000 2.875 2.875 2.875 2.875 3.500 3.500 3.000 2.875 2.875 3.500
##  [661] 1.750 2.875 5.125 3.625 6.125 2.875 3.500 2.875 3.000 2.875 2.875 3.625
##  [673] 3.000 5.125 3.500 5.125 2.000 2.875 3.000 2.000 2.000 3.000 1.750 2.875
##  [685] 3.000 3.625 3.625 3.000 3.500 3.500 2.875 3.000 1.750 3.000 3.625 2.875
##  [697] 2.875 2.875 3.000 3.000 3.000 3.625 3.000 5.125 2.875 3.500 2.875 6.125
##  [709] 1.750 2.000 2.875 5.125 3.500 2.000 2.875 3.000 5.125 3.000 2.875 3.500
##  [721] 3.625 3.500 3.500 2.000 2.875 3.000 3.000 3.625 2.875 5.125 3.000 3.625
##  [733] 2.000 2.875 7.000 3.625 2.000 2.000 2.875 3.000 3.625 2.000 2.000 6.125
##  [745] 2.000 1.750 2.875 3.500 2.000 6.125 2.000 3.000 2.875 5.125 2.000 2.000
##  [757] 3.000 1.750 3.500 2.000 3.000 1.750 2.000 5.125 3.000 1.750 5.125 3.500
##  [769] 3.000 5.125 3.625 5.125 2.875 2.875 0.750 2.000 3.000 3.625 2.000 2.875
##  [781] 3.000 3.500 2.000 3.625 1.750 3.625 3.000 6.125 3.000 3.500 3.000 1.750
##  [793] 5.125 2.875 3.500 3.500 2.875 3.500 1.750 2.000 3.625 3.000 3.000 3.625
##  [805] 2.875 2.000 3.000 6.125 2.875 3.000 3.625 2.875 1.750 2.000 2.875 3.625
##  [817] 3.625 3.500 3.500 1.000 2.875 2.875 3.000 3.000 1.750 2.875 3.000 2.000
##  [829] 3.500 3.625 5.125 1.750 3.500 2.875 3.625 1.000 3.000 1.750 2.875 3.625
##  [841] 2.875 2.875 3.500 5.125 5.125 3.000 2.875 2.875 3.500 2.000 2.000 3.625
##  [853] 2.875 2.000 2.000 3.000 3.000 3.500 3.625 2.000 3.000 2.000 3.000 2.875
##  [865] 1.000 5.125 2.875 2.000 3.000 3.500 3.000 2.875 3.000 5.125 3.500 1.000
##  [877] 3.500 1.750 3.000 2.875 5.125 3.625 2.875 3.500 3.000 5.125 3.500 2.875
##  [889] 2.000 3.625 2.875 3.500 3.500 2.875 3.000 3.000 2.000 3.500 2.875 3.000
##  [901] 3.000 3.000 2.875 2.875 2.875 3.625 3.000 1.000 3.000 2.875 2.875 3.625
##  [913] 1.750 3.625 2.875 2.875 1.750 3.500 3.000 3.000 2.000 3.000 3.000 3.000
##  [925] 3.000 6.125 3.000 2.000 3.500 3.000 2.000 5.125 6.125 3.000 6.125 3.000
##  [937] 3.500 3.625 3.500 3.625 3.625 3.500 3.625 1.750 2.000 3.000 2.875 5.125
##  [949] 2.000 2.875 2.000 3.500 2.000 3.500 1.750 2.000 3.500 3.000 3.000 2.000
##  [961] 2.000 3.000 1.000 3.000 3.500 2.875 1.750 2.875 2.875 1.000 2.875 3.625
##  [973] 3.500 3.000 5.125 5.125 3.625 2.875 2.875 2.000 3.500 2.875 3.500 2.875
##  [985] 3.500 3.625 3.000 5.125 3.500 5.125 2.000 3.000 3.500 3.500 2.875 3.000
##  [997] 2.000 2.875 2.000 2.875
## 
## $func.thetastar
## NULL
## 
## $jack.boot.val
## NULL
## 
## $jack.boot.se
## NULL
## 
## $call
## bootstrap(x = ZeaMays$diff, nboot = B, theta = median)</code></pre>
<blockquote>
<p>Laplace distribution</p>
</blockquote>
<p>Probability density function :</p>
<p><span class="math display">\[f(x\mid \mu ,b)={\frac  {1}{2b}}\exp \left(-{\frac  {|x-\mu |}{b}}\right)\,\!\]</span></p>
<p>Laplace is a function of normal and exponential random variables:</p>
<p><span class="math display">\[X=\sqrt{W}\cdot Z, \quad W\sim Exp(1), \quad Z\sim N(0,1)\]</span>
The Laplace distribution with mean <span class="math inline">\(\mu \in \mathbb{R}\)</span> and scale <span class="math inline">\(b &gt; 0\)</span> has the probability density function <span class="math display">\[f_X(x)=\frac{1}{2b}\exp\left(-\frac{\left|x\right|}{b}\right)\]</span>
Note that if we can show that the Laplace distribution with mean <span class="math inline">\(0\)</span> is a mixture of normals, then by shifting all these normals by <span class="math inline">\(\mu\)</span>, it follows that the Laplace distribution with mean <span class="math inline">\(\mu\)</span> is also a mixture of normals.</p>
<p>Fix <span class="math inline">\(b &gt; 0\)</span>. Let <span class="math inline">\(W \sim \text{Exp}(\frac{1}{2b^2})\)</span>, i.e. <span class="math display">\[f_W(w) = \dfrac{1}{2b^2} \exp \left(-\dfrac{x}{2b^2} \right)\]</span> for <span class="math inline">\(w \geq 0\)</span>, and let <span class="math inline">\(X \mid W = w \sim \mathcal{N}(0, w)\)</span>. We claim that <span class="math inline">\(X\)</span> has the Laplace distribution with mean <span class="math inline">\(0\)</span> and scale <span class="math inline">\(b\)</span>. This is equivalent to showing that for any <span class="math inline">\(x\)</span>,
<span class="math display">\[\begin{align}
f_X(x)=\frac{1}{2b}\exp\left(-\frac{\left|x\right|}{b}\right)
\iff
\int_{0}^{\infty} f_{X\mid W=w}(x)f_{W}(w)dw = \frac{1}{2b}\exp\left(-\frac{\left|x\right|}{b}\right)
\end{align}\]</span>
<span class="math display">\[\begin{align}
\int_{0}^{\infty} f_{X\mid W=w}(x)f_{W}(w)dw &amp;= \int_{0}^{\infty} \frac{1}{\sqrt{2\pi w}}\exp\left(-\frac{x^2}{2w}\right)\frac{1}{2b^2}\exp\left(-\frac{w}{2b^2}\right)dw\\
&amp;= \frac{1}{2b^2}\int_{0}^{\infty}\frac{1}{\sqrt{2\pi w}}\exp\left(-\frac{w^2+x^2b^2}{2b^2w}\right)dw\\
&amp;= \frac{1}{2b^2}\int_{0}^{\infty}\frac{1}{\sqrt{2\pi w}}\exp\left(-\frac{(w-|x|b)^2}{2b^2w}-\frac{2w|x|b}{2b^2w}\right)dw\\
&amp;= \frac{1}{2b^2}e^{-|x|/b}\int_{0}^{\infty}\frac{b}{\sqrt{2\pi b^2 w}}\exp\left(-\frac{(w-|x|b)^2}{2b^2w}\right)dw\\
&amp;=\frac{1}{2b}\exp\left(-\frac{\left|x\right|}{b}\right)
\end{align}\]</span></p>
<p>The Mean of <span class="math display">\[\frac{1}{2b}\exp\left(-\frac{\left|x-\mu\right|}{b}\right)\]</span> is <span class="math inline">\(\mu\)</span> and Variance is <span class="math inline">\(2b^2\)</span>.</p>
<p>The Laplace distribution is an example of where the consideration of the generative process indicates how the variance and mean are linked. The expectation value and variance of an asymmetric Laplace distribution <span class="math inline">\(AL(\theta, \mu, \sigma)\)</span> are <span class="math display">\[\mathbb E(X)=\theta+ \mu\]</span> and <span class="math display">\[\text{var} (X)=\mu^2 +\sigma^2\]</span> Note the variance is dependent on the mean, unless
(the case of the symmetric Laplace Distribution). This is the feature of the distribution that makes it useful. Having a mean-variance dependence is very common for physical measurements, be they microarray fluorescence intensities, peak heights from a mass spectrometer, or reads counts from high-throughput sequencing, as we’ll see in the next section.</p>
<pre class="r"><code>## -----------------------------------------------------------------------------
c(N3 = choose(5, 3), N15 = choose(29, 15))</code></pre>
<pre><code>##       N3      N15 
##       10 77558760</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
w = rexp(10000, rate = 1)

## -----------------------------------------------------------------------------
mu  = 0.3
lps = rnorm(length(w), mean = mu, sd = sqrt(w))
ggplot(data.frame(lps), aes(x = lps)) +
  geom_histogram(fill = &quot;purple&quot;, binwidth = 0.1)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-289-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
mu = 0.3; sigma = 0.4; theta = -1
w  = rexp(10000, 1)
alps = rnorm(length(w), theta + mu * w, sigma * sqrt(w))
ggplot(tibble(alps), aes(x = alps)) +
  geom_histogram(fill = &quot;purple&quot;, binwidth = 0.1)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-290-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
ggplot(tibble(x = rgamma(10000, shape = 2, rate = 1/3)),
   aes(x = x)) + geom_histogram(bins = 100, fill= &quot;purple&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-291-1.svg" width="576" /></p>
<pre class="r"><code>ggplot(tibble(x = rgamma(10000, shape = 10, rate = 3/2)),
   aes(x = x)) + geom_histogram(bins = 100, fill= &quot;purple&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-291-2.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
lambda = rgamma(10000, shape = 10, rate = 3/2)
gp = rpois(length(lambda), lambda = lambda)
ggplot(tibble(x = gp), aes(x = x)) +
  geom_histogram(bins = 100, fill= &quot;purple&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-292-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;vcd&quot;)
ofit = goodfit(gp, &quot;nbinomial&quot;)
plot(ofit, xlab = &quot;&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-293-1.svg" width="576" /></p>
<pre class="r"><code>ofit$par</code></pre>
<pre><code>## $size
## [1] 10.23463
## 
## $prob
## [1] 0.6037692</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
x    = 0:95
mu   = 50
vtot = 80
v1   = vtot - mu
scale = v1/mu    # 0.6
shape = mu^2/v1  # 83.3
p1   = dgamma(x = x, scale = 0.6, shape = 80)
p2   = dpois(x = x, lambda = mu*1.2)
p3   = dnbinom(x = x, mu = mu, size = mu^2/vtot)</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;RColorBrewer&quot;)
cols = brewer.pal(8, &quot;Paired&quot;)
par(mfrow=c(3,1), mai=c(0.5, 0.5, 0.01, 0.01))
xlim = x[c(1, length(x))]
plot(NA, NA, xlim=xlim, ylim=c(0,0.07), type=&quot;n&quot;, ylab=&quot;&quot;, xlab=&quot;&quot;)
polygon(x, p1, col=cols[1])
abline(v=mu, col=&quot;black&quot;, lwd=3)
abline(v=mu*1.2, col=cols[2], lty=2, lwd=3)
plot(x, p2, col=cols[3], xlim=xlim, ylab=&quot;&quot;, xlab=&quot;&quot;, type=&quot;h&quot;, lwd=2)
abline(v=mu*1.2, col=cols[2], lwd=2)
abline(v=mu*1.1, col=cols[4], lty=2, lwd=3)
plot(x, p3, col=cols[4], xlim=xlim, type=&quot;h&quot;, lwd=2, ylab=&quot;&quot;, xlab=&quot;&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-295-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
lambdas = seq(100, 900, by = 100)
simdat = lapply(lambdas, function(l)
    tibble(y = rpois(n = 40, lambda=l), lambda = l)
  ) %&gt;% bind_rows
simdat</code></pre>
<pre><code>## # A tibble: 360 × 2
##        y lambda
##    &lt;int&gt;  &lt;dbl&gt;
##  1   105    100
##  2   102    100
##  3    97    100
##  4    94    100
##  5   100    100
##  6   100    100
##  7   102    100
##  8    97    100
##  9    90    100
## 10   107    100
## # ℹ 350 more rows</code></pre>
<pre class="r"><code>library(&quot;ggbeeswarm&quot;)
ggplot(simdat, aes(x = lambda, y = y)) +
  geom_beeswarm(alpha = 0.6, color = &quot;purple&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-297-1.svg" width="576" /></p>
<pre class="r"><code>ggplot(simdat, aes(x = lambda, y = sqrt(y))) +
  geom_beeswarm(alpha = 0.6, color = &quot;purple&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-298-1.svg" width="576" /></p>
<p><a href="https://math.stackexchange.com/questions/1536459/expected-value-of-square-root-of-poisson-random-variable">Expected Value of Square Root of Poisson Random Variable</a></p>
<p>In general, for smooth <span class="math inline">\(g(X)\)</span> you can do a Taylor expansion around the mean <span class="math inline">\(\mu=E(X)\)</span> :</p>
<p><span class="math display">\[g(X)=g(\mu)+g′(\mu)(X−\mu)+\frac{g′′(\mu)}{2!}(X−\mu)^2+\frac{g′′′(\mu)}{3!}(X−\mu)^3+⋯\]</span>
So</p>
<p><span class="math display">\[E[g(X)]=g(\mu)+\frac{g′′(\mu)}{2!}m_2+\frac{g′′′(\mu)}{3!}m_3+⋯\]</span>
where <span class="math inline">\(m_i\)</span> is the <span class="math inline">\(i^{th}\)</span> centered moment. In our case <span class="math inline">\(m_2=m_3=\lambda\)</span>, so:</p>
<p><span class="math display">\[E[g(X)]=\sqrt{\lambda}−\frac{\lambda^{−1/2}}{8}+\frac{\lambda^{−3/2}}{16}+⋯\]</span>
This approximation is useful only if <span class="math inline">\(\lambda \gg 1\)</span></p>
<p>The mean and variance of any Poisson process is given as</p>
<p><span class="math display">\[E[P(\lambda_i)] = Var[P(\lambda_i)] = \lambda_i\]</span></p>
<p><a href="https://stats.stackexchange.com/questions/505298/square-root-transformation-of-poisson-process-how-small-var-sqrtp-lambda">Square root transformation of Poisson process</a></p>
<p><span class="math display">\[\operatorname{var}(\sqrt{P(\lambda)})=E[P(\lambda)]-E[\sqrt{P(\lambda)}]^2=\lambda-E[\sqrt{P(\lambda)}]^2\]</span></p>
<p>The second term can be approximated <em>better</em> as follows:</p>
<p><span class="math display">\[E[\sqrt{P(\lambda)}]\approx \sqrt{\lambda}-\frac{\lambda^{-1/2}}{8}+\frac{\lambda^{-3/2}}{16}+...\]</span></p>
<p>Which is why the expected value is also approximated by <span class="math inline">\(\sqrt \lambda\)</span>. Square of it will be
<span class="math display">\[E[\sqrt{P(\lambda)}]^2\approx \lambda -\frac{1}{4}+\frac{9}{64\lambda}+...\]</span></p>
<p>So, the variance will be approximately
<span class="math display">\[\operatorname{var}(\sqrt{P(\lambda)})\approx \frac{1}{4}-\frac{9}{64\lambda}+...\]</span></p>
<p>which is approximately <span class="math inline">\(1/4\)</span> for large <span class="math inline">\(\lambda\)</span>, and the standard deviation is approximately <span class="math inline">\(1/2\)</span>.</p>
<pre class="r"><code>## -----------------------------------------------------------------------------
.o = options(digits = 3)
summarise(group_by(simdat, lambda), sd(y), sd(2*sqrt(y)))</code></pre>
<pre><code>## # A tibble: 9 × 3
##   lambda `sd(y)` `sd(2 * sqrt(y))`
##    &lt;dbl&gt;   &lt;dbl&gt;             &lt;dbl&gt;
## 1    100    10.3             1.03 
## 2    200    14.7             1.05 
## 3    300    17.9             1.04 
## 4    400    19.2             0.959
## 5    500    23.1             1.03 
## 6    600    25.5             1.05 
## 7    700    26.8             1.02 
## 8    800    26.7             0.945
## 9    900    37.0             1.24</code></pre>
<pre class="r"><code>options(.o)</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
muvalues = 2^seq(0, 10, by = 1)
simgp = lapply(muvalues, function(mu) {
  u = rnbinom(n = 1e4, mu = mu, size = 4) # size:   target for number of successful trials, or dispersion parameter (the shape parameter of the gamma mixing distribution). Must be strictly positive, need not be integer.
  tibble(mean = mean(u), sd = sd(u),
         lower = quantile(u, 0.025),
         upper = quantile(u, 0.975),
         mu = mu)
  } ) %&gt;% bind_rows
simgp</code></pre>
<pre><code>## # A tibble: 11 × 5
##       mean     sd lower upper    mu
##      &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1    1.00   1.11     0    4      1
##  2    1.98   1.72     0    6      2
##  3    4.03   2.86     0   11      4
##  4    8.03   4.93     1   20      8
##  5   16.1    8.96     3   38     16
##  6   32.0   16.7      8   71     32
##  7   64.1   33.2     16  143     64
##  8  127.    64.9     32  282    128
##  9  257.   129.      69  556    256
## 10  507.   252.     133 1094.   512
## 11 1020.   508.     272 2239.  1024</code></pre>
<pre class="r"><code>head(as.data.frame(simgp), 2)</code></pre>
<pre><code>##     mean       sd lower upper mu
## 1 1.0019 1.110234     0     4  1
## 2 1.9824 1.723133     0     6  2</code></pre>
<pre class="r"><code>ggplot(simgp, aes(x = mu, y = mean, ymin = lower, ymax = upper)) +
  geom_point() + geom_errorbar()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-302-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
simgp = mutate(simgp,
  slopes = 1 / sd,
  trsf   = cumsum(slopes * mean))
ggplot(simgp, aes(x = mean, y = trsf)) +
  geom_point() + geom_line() + xlab(&quot;&quot;)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-303-1.svg" width="576" /></p>
<p>Call our transformation function <span class="math inline">\(g\)</span>, and assume it’s differentiable. Also call our random variables <span class="math inline">\(X_i\)</span> , with means <span class="math inline">\(\mu_i\)</span> and variances <span class="math inline">\(v_i\)</span> , and we assume that
<span class="math inline">\(\mu_i\)</span> and <span class="math inline">\(v_i\)</span> are related by a functional relationship <span class="math inline">\(v_i = \nu(\mu_i)\)</span> . Then, for values of <span class="math inline">\(X_i\)</span> in the neighborhood of its mean <span class="math inline">\(\mu_i\)</span>,
<span class="math display">\[g(X_i)=g(\mu_i)+g&#39;(\mu_i)(X_i-\mu_i)+\cdots\]</span></p>
<p>where the dots stand for higher order terms that we can neglect. The variances of the transformed values are then
<span class="math display">\[\text{Var }(g(X_i))\simeq g&#39;(\mu_i)^2\]</span>
<span class="math display">\[\text{Var }(X_i)\simeq g&#39;(\mu_i)^2\nu(\mu_i)\]</span></p>
<p>where we have used the rules <span class="math inline">\(\text{Var}(X-c)=\text{Var}(X)\)</span> and <span class="math inline">\(\text{Var}(cX)=c^2\text{Var}(X)\)</span> that hold whenever <span class="math inline">\(c\)</span>
is a constant number. Requiring that this be constant leads to the differential equation <span class="math display">\[g&#39;(x)=\frac{1}{\sqrt{\nu(x)}}\]</span> For a given mean-variance relationship <span class="math inline">\(\nu(\mu)\)</span>, we can solve this for the function <span class="math inline">\(g\)</span> . Let’s check this for some simple cases:</p>
<blockquote>
<ul>
<li>if <span class="math inline">\(\nu(\mu)=\mu\)</span>(Poisson), we recover <span class="math inline">\(g(x)=\sqrt{x}\)</span>, the square root transformation.</li>
</ul>
</blockquote>
<blockquote>
<ul>
<li>If <span class="math inline">\(\nu(\mu)=\alpha \mu^2\)</span>, solving the differential equation <span class="math display">\[g&#39;(x)=\frac{1}{\sqrt{\nu(x)}}\]</span> gives <span class="math inline">\(g(x)=\log(x)\)</span>. This explains why the logarithm transformation is so popular in many data analysis applications: it acts as a variance stabilizing transformation whenever the data have a constant coefficient of variation, that is, when the standard deviation is proportional to the mean.</li>
</ul>
</blockquote>
<pre class="r"><code>## -----------------------------------------------------------------------------
f = function(x, a) 
  ifelse (a==0, 
    sqrt(x), 
    log(2*sqrt(a) * sqrt(x*(a*x+1)) + 2*a*x+1) / (2*sqrt(a)))
x  = seq(0, 24, by = 0.1)
df = lapply(c(0, 0.05*2^(0:5)), function(a) 
  tibble(x = x, a = a, y = f(x, a))) %&gt;% bind_rows()
ggplot(df, aes(x = x, y = y, col = factor(a))) + 
  geom_line() + labs(col = expression(alpha))</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-304-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
f2 = function(x, a) ifelse (a==0, sqrt(x), acosh(2*a*x + 1) / (2*sqrt(a)))  
with(df, max(abs(f2(x,a) - y)))</code></pre>
<pre><code>## [1] 8.881784e-16</code></pre>
<pre class="r"><code>stopifnot(with(df, max(abs(f2(x,a) - y))) &lt; 1e-10)</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
a = c(0.2, 0.5, 1)
f(1e6, a) </code></pre>
<pre><code>## [1] 15.196731 10.259171  7.600903</code></pre>
<pre class="r"><code>1/(2*sqrt(a)) * (log(1e6) + log(4*a))</code></pre>
<pre><code>## [1] 15.196728 10.259170  7.600902</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
setwd(&quot;D:/books/R/blogdown/hugo&quot;)
mx = readRDS(&quot;../data/Myst.rds&quot;)$yvar
str(mx)</code></pre>
<pre><code>##  num [1:1800] 0.3038 0.0596 -0.0204 0.1849 0.2842 ...</code></pre>
<pre class="r"><code>ggplot(tibble(mx), aes(x = mx)) + geom_histogram(binwidth = 0.025)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-308-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
wA = runif(length(mx))
wB = 1 - wA

## -----------------------------------------------------------------------------
iter      = 0
loglik    = -Inf
delta     = +Inf
tolerance = 1e-3
miniter   = 50
maxiter   = 1000</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
while((delta &gt; tolerance) &amp;&amp; (iter &lt;= maxiter) || (iter &lt; miniter)) {
  lambda = mean(wA)
  muA = weighted.mean(mx, wA)
  muB = weighted.mean(mx, wB)
  sdA = sqrt(weighted.mean((mx - muA)^2, wA))
  sdB = sqrt(weighted.mean((mx - muB)^2, wB))

  pA   =    lambda    * dnorm(mx, mean = muA, sd = sdA)
  pB   = (1 - lambda) * dnorm(mx, mean = muB, sd = sdB)
  ptot = pA + pB
  wA   = pA / ptot
  wB   = pB / ptot

  loglikOld = loglik
  loglik = sum(log(pA)) + sum(log(pB))
  delta = abs(loglikOld - loglik)
  iter = iter + 1
}
iter</code></pre>
<pre><code>## [1] 315</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
.o = options(digits = 3)
c(lambda, muA, muB, sdA, sdB)</code></pre>
<pre><code>## [1]  0.5244  0.1473 -0.1694  0.1498  0.0983</code></pre>
<pre class="r"><code>options(.o)

## -----------------------------------------------------------------------------
.o = options(digits = 3)
gm = mixtools::normalmixEM(mx, k = 2)</code></pre>
<pre><code>## number of iterations= 276</code></pre>
<pre class="r"><code>with(gm, c(lambda[1], mu, sigma))</code></pre>
<pre><code>## [1]  0.4756 -0.1694  0.1473  0.0983  0.1498</code></pre>
<pre class="r"><code>options(.o)</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
library(&quot;flexmix&quot;)</code></pre>
<pre><code>## 
## Attaching package: &#39;flexmix&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:boot&#39;:
## 
##     boot</code></pre>
<pre><code>## The following object is masked from &#39;package:Gviz&#39;:
## 
##     group</code></pre>
<pre class="r"><code>data(&quot;NPreg&quot;)
NPreg</code></pre>
<pre><code>##              x         yn yp yb class id1 id2
## 1   4.17663259 22.3803787  4  0     1   1   1
## 2   1.20163055  5.1115746  3  0     1   1   1
## 3   2.29500635 13.2510576  9  0     1   2   1
## 4   5.96586786 30.2852404  3  1     1   2   1
## 5   2.35808338 14.7645078  4  0     1   3   2
## 6   7.63706104 42.8337601  1  1     1   3   2
## 7   6.45246460 34.2667327  2  1     1   4   2
## 8   9.40163061 42.8138796  2  1     1   4   2
## 9   8.29927115 49.0659954  2  1     1   5   3
## 10  4.33053110 19.3005980  4  0     1   5   3
## 11  4.61978652 29.8703835  2  0     1   6   3
## 12  9.71513912 46.4716001  0  1     1   6   3
## 13  2.27872578  4.6923751  9  1     1   7   4
## 14  6.92897805 34.3449694  0  1     1   7   4
## 15  7.25327696 32.9045352  0  1     1   8   4
## 16  0.40434891  1.2872814  6  1     1   8   4
## 17  9.36491266 49.8382514  2  1     1   9   5
## 18  2.57714695 11.8085276  4  0     1   9   5
## 19  7.74066373 33.0757635  1  1     1  10   5
## 20  9.89474383 51.6069735  4  1     1  10   5
## 21  0.55744518  4.6421303 10  0     1  11   6
## 22  1.54234713  7.9628999 10  0     1  11   6
## 23  8.09633770 45.4036015  2  1     1  12   6
## 24  1.08087069  4.1806895  7  0     1  12   6
## 25  3.13088989 10.1874921  2  0     1  13   7
## 26  7.36439054 37.6754354  2  1     1  13   7
## 27  8.85874283 45.7929465  2  1     1  14   7
## 28  5.66849986 33.4717028  2  1     1  14   7
## 29  6.50576695 32.7460204  1  1     1  15   8
## 30  1.53762205  4.8097527  6  0     1  15   8
## 31  6.73142951 34.4618322  1  1     1  16   8
## 32  0.54238057  6.7040248  8  0     1  16   8
## 33  0.34155727  1.5408095  6  0     1  17   9
## 34  4.75640223 22.0913005  5  0     1  17   9
## 35  8.73919645 42.1726314  0  1     1  18   9
## 36  2.70205396 10.7961692  4  0     1  18   9
## 37  8.30704714 40.8369753  0  1     1  19  10
## 38  3.74932799 13.5857934  6  0     1  19  10
## 39  4.28106160 17.0062023  5  0     1  20  10
## 40  0.44036390  4.8016176  4  0     1  20  10
## 41  7.20383449 35.2574679  1  1     1  21  11
## 42  0.82404352  3.6779612  7  0     1  21  11
## 43  4.42757293 21.4447769  3  1     1  22  11
## 44  0.33236471  3.9858411  4  0     1  22  11
## 45  8.98413221 41.3208325  1  1     1  23  12
## 46  4.09455624 22.0601242  2  0     1  23  12
## 47  1.95838182  7.5923929  4  0     1  24  12
## 48  9.15891990 41.5530343  0  1     1  24  12
## 49  8.89296908 45.8451285  1  0     1  25  13
## 50  2.45296791  7.2185934  5  0     1  25  13
## 51  0.06295317 -0.7382478  4  0     1  26  13
## 52  2.85040085 11.5908587  3  0     1  26  13
## 53  7.96088626 36.2613458  3  1     1  27  14
## 54  5.77690206 28.5847337  3  1     1  27  14
## 55  4.13608936 25.3425513  3  0     1  28  14
## 56  0.42128309 -1.4128968  7  0     1  28  14
## 57  9.84276558 51.1781358  1  1     1  29  15
## 58  0.20367758  3.0164145  8  0     1  29  15
## 59  1.22520362 10.0824988  3  0     1  30  15
## 60  7.71576158 40.6650416  2  1     1  30  15
## 61  5.68957164 33.5874400  1  1     1  31  16
## 62  1.64931152  7.6441527  8  1     1  31  16
## 63  6.21120097 31.0314027  2  0     1  32  16
## 64  1.26452997 -1.4815158  5  0     1  32  16
## 65  6.92773212 30.1226659  2  0     1  33  17
## 66  5.03979465 24.0594062  0  1     1  33  17
## 67  7.31309950 44.6842204  0  1     1  34  17
## 68  4.63261999 27.0774981  4  0     1  34  17
## 69  0.51121986  5.3782324  4  0     1  35  18
## 70  1.07330848  5.5521353  4  0     1  35  18
## 71  7.96409829 34.2089461  2  1     1  36  18
## 72  1.99945993  5.6326370  3  0     1  36  18
## 73  1.87378620  6.1134515  7  0     1  37  19
## 74  4.74580563 19.8444540  2  0     1  37  19
## 75  5.76907872 26.7631365  3  1     1  38  19
## 76  0.52463379  4.3079712  2  0     1  38  19
## 77  3.46584146 14.5557600  6  0     1  39  20
## 78  7.29252838 39.4454486  1  1     1  39  20
## 79  5.40030459 26.7301592  3  0     1  40  20
## 80  2.79238097 14.0218112  8  0     1  40  20
## 81  3.90283673 15.9778277  2  0     1  41  21
## 82  5.04182776 26.2447923  2  0     1  41  21
## 83  6.42973634 30.8354474  1  1     1  42  21
## 84  7.48850271 35.4915284  5  1     1  42  21
## 85  8.63020112 48.2392871  2  1     1  43  22
## 86  7.15776147 39.7617173  3  1     1  43  22
## 87  8.97324030 43.3963793  0  1     1  44  22
## 88  2.60951572 13.8106373  3  0     1  44  22
## 89  2.96134369 12.6836033  5  1     1  45  23
## 90  8.83580907 46.3313585  1  1     1  45  23
## 91  6.14941012 23.8764264  4  1     1  46  23
## 92  1.95363952 10.5716207  4  0     1  46  23
## 93  4.39655041 18.5927401  3  0     1  47  24
## 94  3.13603434 15.2807182  4  0     1  47  24
## 95  5.99525632 29.7251730  1  1     1  48  24
## 96  8.79126861 53.4866896  2  1     1  48  24
## 97  9.68095973 49.4019432  2  1     1  49  25
## 98  7.40750214 37.4921016  2  1     1  49  25
## 99  4.11756789 26.1089606  2  1     1  50  25
## 100 8.77639766 43.8204884  0  1     1  50  25
## 101 8.39039786 30.0024180  9  0     2  51  26
## 102 2.11470492 30.7784942  3  1     2  51  26
## 103 9.04111851 25.4453871  4  0     2  52  26
## 104 3.39129320 37.8679635  2  1     2  52  26
## 105 9.04211228 26.6354192  6  0     2  53  27
## 106 3.99784800 43.6441463  5  1     2  53  27
## 107 6.35936341 40.1565408  6  0     2  54  27
## 108 7.99747483 26.8208712  7  0     2  54  27
## 109 4.40447242 47.2149865  6  0     2  55  28
## 110 8.03273979 28.4504319  6  0     2  55  28
## 111 9.03479311 30.4918954  5  0     2  56  28
## 112 3.21633814 34.7144548  4  1     2  56  28
## 113 1.16274254 18.5742013  2  1     2  57  29
## 114 3.38076958 37.0781720  2  1     2  57  29
## 115 1.25251094 22.5944761  3  1     2  58  29
## 116 5.87154603 38.5059444  4  0     2  58  29
## 117 1.77085263 32.5862954  3  0     2  59  30
## 118 3.43636802 36.4778478  4  1     2  59  30
## 119 7.16614059 29.6802797  3  0     2  60  30
## 120 3.76918128 40.6183397  5  1     2  60  30
## 121 1.52081588 29.7501823  1  1     2  61  31
## 122 5.37764869 40.1085457  7  0     2  61  31
## 123 0.68309405 26.2862360  2  1     2  62  31
## 124 5.80866236 38.1224013  6  0     2  62  31
## 125 5.32623267 34.4266149  4  1     2  63  32
## 126 1.55523806 28.9870979  4  1     2  63  32
## 127 1.77223938 31.0807937  3  1     2  64  32
## 128 9.47912345 25.0666566  7  0     2  64  32
## 129 7.34257603 34.7295232  3  0     2  65  33
## 130 7.56841214 30.5249015 12  0     2  65  33
## 131 6.38570772 38.8844988  4  0     2  66  33
## 132 3.23916002 40.8915645  6  1     2  66  33
## 133 1.55249906 27.9477605  3  1     2  67  34
## 134 0.44023583 17.5178401  3  1     2  67  34
## 135 5.10267347 38.4661073  4  1     2  68  34
## 136 6.25680008 35.7063530  5  1     2  68  34
## 137 3.03429356 35.4377378  7  1     2  69  35
## 138 5.58708904 34.4944798  8  0     2  69  35
## 139 6.21969731 34.1132328  3  0     2  70  35
## 140 9.35484102 23.6351578  4  0     2  70  35
## 141 9.51248483 18.8757761  8  0     2  71  36
## 142 6.13313907 38.2737395  8  0     2  71  36
## 143 9.33058083 20.5529819  9  0     2  72  36
## 144 0.99726202 26.3021062  4  1     2  72  36
## 145 5.05373145 36.3972843  3  0     2  73  37
## 146 3.98479898 40.5567099  4  0     2  73  37
## 147 1.77266492 27.3847921  3  1     2  74  37
## 148 2.24338040 28.1594832  2  1     2  74  37
## 149 3.46701554 39.0302418  6  1     2  75  38
## 150 4.06068979 34.0714501  4  1     2  75  38
## 151 5.82560360 38.2653651  4  1     2  76  38
## 152 8.81287538 22.8008357  6  0     2  76  38
## 153 4.57472311 36.2760541  4  1     2  77  39
## 154 8.24189374 29.1903484  3  0     2  77  39
## 155 4.76454221 44.6066641  5  1     2  78  39
## 156 7.98348782 27.5794882  2  0     2  78  39
## 157 5.24510010 41.9042338  2  0     2  79  40
## 158 6.69475207 39.1258420  4  0     2  79  40
## 159 2.70818324 38.7040567  9  1     2  80  40
## 160 7.52318812 35.7197554  5  0     2  80  40
## 161 3.69377164 43.4333492  6  1     2  81  41
## 162 4.69006520 39.3015355  0  1     2  81  41
## 163 0.60398059 20.6504112  3  1     2  82  41
## 164 0.43911146 11.3941301  3  1     2  82  41
## 165 5.81689226 34.8166924  5  1     2  83  42
## 166 6.09789830 37.6550523  7  0     2  83  42
## 167 9.81102144 24.9727956 11  0     2  84  42
## 168 3.13948817 40.4528939  2  1     2  84  42
## 169 9.74593014 20.2982802  4  0     2  85  43
## 170 6.17080719 38.8148035  7  0     2  85  43
## 171 8.92470426 18.9851511 10  0     2  86  43
## 172 6.44711653 33.5411911  6  0     2  86  43
## 173 3.63741036 34.8878700  2  1     2  87  44
## 174 7.36664722 30.5144068  5  0     2  87  44
## 175 6.63565478 35.2423764  8  0     2  88  44
## 176 5.79653555 41.0503333  1  0     2  88  44
## 177 3.20066856 33.9889591  7  1     2  89  45
## 178 8.30276330 32.0745613  4  0     2  89  45
## 179 2.59769300 33.9575573  2  1     2  90  45
## 180 7.17121302 35.3457404  3  0     2  90  45
## 181 3.66614757 34.6844817  3  1     2  91  46
## 182 2.94551030 36.8147255  3  1     2  91  46
## 183 7.63484824 31.7443404  5  0     2  92  46
## 184 8.38360701 26.6002184  6  0     2  92  46
## 185 7.95656429 36.3470091  3  0     2  93  47
## 186 7.21477790 39.0676688  4  0     2  93  47
## 187 1.03266907 22.7904631  3  1     2  94  47
## 188 6.26142937 39.1718546  9  0     2  94  47
## 189 3.75780662 36.3338405  5  1     2  95  48
## 190 5.93019124 41.2870574  9  0     2  95  48
## 191 1.27990952 19.2903026  3  1     2  96  48
## 192 8.97071640 25.0368343  5  0     2  96  48
## 193 1.64046864 25.3235371  2  0     2  97  49
## 194 5.27091961 39.5271491  8  1     2  97  49
## 195 3.24612225 36.6728042  4  0     2  98  49
## 196 3.67903721 47.7854039  4  0     2  98  49
## 197 0.73162182 22.7780922  2  1     2  99  50
## 198 6.69055712 37.5966075  4  0     2  99  50
## 199 4.08498638 44.6838712  1  0     2 100  50
## 200 9.18692898 22.4081259  7  0     2 100  50</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
m1 = flexmix(yn ~ x + I(x^2), data = NPreg, k = 2)

## -----------------------------------------------------------------------------
ggplot(NPreg, aes(x = x, y = yn)) + geom_point()</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-313-1.svg" width="576" /></p>
<pre class="r"><code>## -----------------------------------------------------------------------------
modeltools::parameters(m1, component = 1)</code></pre>
<pre><code>##                       Comp.1
## coef.(Intercept) -0.20801986
## coef.x            4.81563289
## coef.I(x^2)       0.03636977
## sigma             3.47480999</code></pre>
<pre class="r"><code>modeltools::parameters(m1, component = 2)</code></pre>
<pre><code>##                      Comp.2
## coef.(Intercept) 14.7181608
## coef.x            9.8452394
## coef.I(x^2)      -0.9682084
## sigma             3.4808821</code></pre>
<pre class="r"><code>modeltools::clusters(m1)</code></pre>
<pre><code>##   [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
##  [38] 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1
##  [75] 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
## [112] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 2 2 2
## [149] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 1 2 2 1 2 2 2 2 2 2 2 2 2 1
## [186] 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
table(NPreg$class, modeltools::clusters(m1))</code></pre>
<pre><code>##    
##      1  2
##   1 95  5
##   2  5 95</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
summary(m1)</code></pre>
<pre><code>## 
## Call:
## flexmix(formula = yn ~ x + I(x^2), data = NPreg, k = 2)
## 
##        prior size post&gt;0 ratio
## Comp.1 0.494  100    145 0.690
## Comp.2 0.506  100    141 0.709
## 
## &#39;log Lik.&#39; -642.5451 (df=9)
## AIC: 1303.09   BIC: 1332.775</code></pre>
<pre class="r"><code>## -----------------------------------------------------------------------------
NPreg = mutate(NPreg, gr = factor(class))
ggplot(NPreg, aes(x = x, y = yn, group = gr)) +
   geom_point(aes(colour = gr, shape = gr)) +
   scale_colour_hue(l = 40, c = 180)</code></pre>
<p><img src="../../../../2023/04/28/modern-statistics-for-modern-biology-1/index_files/figure-html/unnamed-chunk-319-1.svg" width="576" /></p>
</div>

    </div>
  </article>

  


</main>

      <footer class="footer">
        <ul class="footer-links">
          <li>
          </li>
          <li>
          </li>
        </ul>
      </footer>

    </div>
    



<script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>



<script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
<script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/yaml.min.js"></script>
<script>hljs.configure({languages: []}); hljs.initHighlightingOnLoad();</script>



    
<script src="../../../../js/math-code.js"></script>
<script async src="//cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"></script>


    
  </body>
</html>

